---
title: "Cell type mapping with NNLS"
date: "Last compiled: `r format(Sys.time(), '%d %B %Y')`"
format: html
editor: visual
editor_options: 
  chunk_output_type: inline
---

Cell type mapping with SRT data refers to a set of methods which allows you to infer the quantity of cells from SRT expression profiles. `semla` offers a quick method based on *Non-Negative Least Squares* to infer cell type proportions directly from Visium spot expression profiles. We utilize the NNLS method implemented in the [RcppML](https://github.com/zdebruine/RcppML) *v0.3.7* package available on CRAN: <https://cran.r-project.org/web/packages/RcppML/index.html>. Please ensure you have the correct version installed before proceeding with this tutorial.

In this tutorial, we'll take a look at two examples. First, we have two tissue sections of a [mouse brain](#mbrain) and a single cell RNA-seq dataset from the Allen Brain Atlas. In the second example, we have a [mouse kidney](#mkidney) section and single cell RNA-seq data from the Tabula Muris Senis.

```{r include=FALSE, eval=FALSE}
install.packages(url = "https://cran.r-project.org/src/contrib/RcppML_0.3.7.tar.gz", 
                 repos=NULL, type="source")
```

```{r, message=FALSE, warning=FALSE}
library(semla)
library(tibble)
library(dplyr)
if (!requireNamespace('TabulaMurisSenisData', quietly = TRUE)) {
  BiocManager::install("TabulaMurisSenisData")
}

library(TabulaMurisSenisData)
library(SingleCellExperiment)
library(Seurat)
library(purrr)
library(patchwork)
library(ggplot2)
library(RcppML)
```

## Mouse brain {#mbrain}

For this tutorial, we will use a single-cell data set from the Allen Brain Atlas. You can check the code chunk below or follow [this](https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1) link to download it manually.

```{r eval=FALSE}
#| code-fold: true
options(timeout=200)
tmpdir <- "." # Set current wd or change to tmpdir()
dir.create(paste0(tmpdir, "/mousebrain"))
targetdir <- paste0(tmpdir, "/mousebrain")
dir.create(paste0(targetdir, "/single-cell"))
destfile <- paste0(targetdir, "/single-cell/allen_brain.rds")
download.file("https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1", destfile = destfile)

```

We will also need to download the 10x Visium data from 10x Genomics website. You can download the files directly with R by following the code chunk below or download the data directly from [here](https://www.10xgenomics.com/resources/datasets).

```{r eval=FALSE}
#| code-fold: true
dir.create(paste0(targetdir, "/visium"))

# Download section 1
dir.create(paste0(targetdir, "/visium/S1"))
download.file(url = "https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.h5", 
              destfile = paste0(targetdir, "/visium/S1/filtered_feature_bc_matrix.h5"))
download.file(url = "https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz",
              destfile = paste0(targetdir, "/visium/S1/spatial.tar.gz"))
untar(tarfile = paste0(targetdir, "/visium/S1/spatial.tar.gz"), 
      exdir =  paste0(targetdir, "/visium/S1/"))
file.remove(paste0(targetdir, "/visium/S1/spatial.tar.gz"))

# Download section 2
dir.create(paste0(targetdir, "/visium/S2"))
download.file(url = "https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Posterior/V1_Mouse_Brain_Sagittal_Posterior_filtered_feature_bc_matrix.h5", 
              destfile = paste0(targetdir, "/visium/S2/filtered_feature_bc_matrix.h5"))
download.file(url = "https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Posterior/V1_Mouse_Brain_Sagittal_Posterior_spatial.tar.gz",
              destfile = paste0(targetdir, "/visium/S2/spatial.tar.gz"))
untar(tarfile = paste0(targetdir, "/visium/S2/spatial.tar.gz"), 
      exdir =  paste0(targetdir, "/visium/S2/"))
file.remove(paste0(targetdir, "/visium/S2/spatial.tar.gz"))
```

Load Visium data with semla into a Seurat object. The following steps assumes that the mouse brain 10x Visium data is located in `./mousebrain/visium/`.

```{r eval=FALSE}
# Assemble spaceranger output files
samples <- Sys.glob("./mousebrain/visium/*/filtered_feature_bc_matrix.h5")
imgs <- Sys.glob("./mousebrain/visium/*/spatial/tissue_hires_image.png")
spotfiles <- Sys.glob("./mousebrain/visium/*/spatial/tissue_positions_list.csv")
json <- Sys.glob("./mousebrain/visium/*/spatial/scalefactors_json.json")

infoTable <- tibble(samples, imgs, spotfiles, json, 
                    section_id = paste0("section_", 1:2))

# Create Seurat object with 1 Sagittal Anterior section and 1 Sagittal Posterior section
se_brain_spatial <- ReadVisiumData(infoTable)

```

Load single-cell data

```{r eval=FALSE}

se_allen <- readRDS("./mousebrain/single-cell/allen_brain.rds")

```

### Normalize data

Here we will apply the same log-normalization procedure to both the 10x Visium data (`se_brain_spatial`) and to the single-cell data (`se_brain_singlecell`). We set the number of variable features quite high because later on we will use the intersect between the variable features in the single-cell data and the variable features in the 10x Visium data for NNLS. The NNLS method is quite fast so there is actually no need to select only a subset of features. Instead, we can just use all genes that are shared across the single-cell and 10x Visium data.

```{r eval=FALSE}

# Normalize data and find variable features for Visium data
se_brain_spatial <- se_brain_spatial |>
  NormalizeData() |>
  FindVariableFeatures(nfeatures = 10000)

# Normalize data and run vanilla analysis to create UMAP embedding
se_allen <- se_allen |>
  NormalizeData() |>
  FindVariableFeatures() |> 
  ScaleData() |> 
  RunPCA() |> 
  RunUMAP(reduction = "pca", dims = 1:30)

# Rerun FindVariableFeatures to increase the number before cell type deconvolution
se_allen <- se_allen |> 
  FindVariableFeatures(nfeatures = 10000)

```

```{r eval=FALSE, include=FALSE}
saveRDS(se_brain_spatial, "R_objects/se_brain_spatial")
```

```{r include=FALSE}

se_allen <- readRDS("R_objects/allen_brain")
se_brain_spatial <- readRDS("R_objects/se_brain_spatial")

```

We can visualize the available cell types in our UMAP embedding of the cells. We have access to 23 annotated cell types, including L2-L6 layer neurons which have a distinct spatial distribution in the tissue.

```{r}
DimPlot(se_allen, group.by = "subclass")
```

### Run NNLS

The `RunNNLS()` method requires a normalized `Seurat` object with 10x Visium data and a a normalized `Seurat` object with single-cell data. The `groups` argument defines where the cell type labels should be taken from in the single-cell `Seurat` object. In our single-cell `Seurat` object, the labels are stored in the "subclass" column.

```{r}
DefaultAssay(se_brain_spatial) <- "Spatial"

ti <- Sys.time()
se_brain_spatial <- RunNNLS(object = se_brain_spatial, 
                            singlecell_object = se_allen, 
                            groups = "subclass")
sprintf("RunNNLS completed in %s seconds", round(Sys.time() - ti, digits = 2))

# Check available cell types
rownames(se_brain_spatial)
```

NB: the CR cell type was discarded because the number of cell was lower than 10. 10 is the lower limit for the allowed number of cells per cell type but this can be overridden with `minCells_per_celltype`.

The plots below show the spatial distributions of proportions for a selected set of cell types.

```{r class.source = 'fold-hide'}
# Plot selected cell types
DefaultAssay(se_brain_spatial) <- "celltypeprops"

selected_celltypes <- c("L2/3 IT", "L4", "L5 IT", 
                       "L5 PT", "L6 CT", "L6 IT", "L6b", 
                       "Oligo", "Pvalb", "Meis2", "Astro",
                       "VLMC", "SMC")
```

```{r}
se_brain_spatial <- LoadImages(se_brain_spatial, image_height = 1e3)
plots <- lapply(seq_along(selected_celltypes), function(i) {
  MapFeatures(se_brain_spatial, pt_size = 1.3,
            features = selected_celltypes[i], image_use = "raw",
            arrange_features = "row", scale = "shared", 
            override_plot_dims = TRUE,
            colors = RColorBrewer::brewer.pal(n = 9, name = "Spectral") |> rev(), 
            scale_alpha = TRUE)  +
  plot_layout(guides = "collect") & 
  theme(legend.position = "right", legend.margin = margin(b = 50),
        legend.text = element_text(angle = 0),
        plot.title = element_blank())
}) |> setNames(nm = selected_celltypes)
```

### Mapped mouse brain cell types {.tabset}

```{r results='asis', echo=FALSE}
for(i in 1:length(selected_celltypes)){
  cat("#### ", selected_celltypes[i],"\n")
  print(plots[[i]])
  cat("\n\n")
}
```

### Visualize multiple cell types

We can also visualize some of these cell types in one single plot with `MapMultipleFeatures()`.

```{r message=FALSE, fig.asp=0.4, fig.width=10}

# Load H&E images
se_brain_spatial <- se_brain_spatial |> 
  LoadImages()

# Plot multiple features
MapMultipleFeatures(se_brain_spatial, 
                    image_use = "raw", 
                    pt_size = 2, max_cutoff = 0.99,
                    override_plot_dims = TRUE, 
                    colors = c("#332288", "#88CCEE", "#44AA99", "#117733", "#DDCC77", "#CC6677","#AA4499"),
                    features = selected_celltypes[1:7]) +
  plot_layout(guides = "collect")

```

<br>

### Cell type co-localization

By computing the pair-wise correlation between cell types across spots, we can get an idea of which cell types often appear together in the same spots.

```{r fig.asp=1, fig.width=4}
cor_matrix <- FetchData(se_brain_spatial, selected_celltypes) |> 
  mutate_all(~ if_else(.x<0.1, 0, .x)) |>  # Filter lowest values (-> set as 0)
  cor()

diag(cor_matrix) <- NA
max_val <- max(cor_matrix, na.rm = T)
cols <- RColorBrewer::brewer.pal(7, "RdYlBu") |> rev(); cols[4] <- "white"
pheatmap::pheatmap(cor_matrix, 
                   breaks = seq(-max_val, max_val, length.out = 100),
                   color=colorRampPalette(cols)(100),
                   cellwidth = 14, cellheight = 14, 
                   treeheight_col = 10, treeheight_row = 10, 
                   main = "Cell type correlation\nwithin spots")
```

<br>

We can also use the predicted cell type proportions to compute factors using NMF, and thereby summarizing the cell types presence within the same location into a set of predefined factors. Each factor can be viewed as a niche of a certain cell type composition.

```{r fig.asp=1, fig.width=4}
nmf_data <- FetchData(se_brain_spatial, selected_celltypes) |> 
  RcppML::nmf(k = 10, verbose = F)

nmf_data_h <- nmf_data$h |> as.data.frame()
rownames(nmf_data_h) <- paste0("Factor_", 1:10)
colnames(nmf_data_h) <- selected_celltypes

nmf_data_h <- nmf_data_h |> 
  mutate_at(colnames(nmf_data_h), 
            ~(scale(., center = FALSE, scale = max(., na.rm = TRUE)/1)))

nmf_data_h$Factor <- rownames(nmf_data_h) |> 
  factor(levels = paste0("Factor_", 1:10))
nmf_data_h_df <- nmf_data_h |> 
  tidyr::pivot_longer(cols = all_of(selected_celltypes), 
                      names_to = "Cell", 
                      values_to = "Weight")

ggplot(nmf_data_h_df, aes(x=Factor, y=Cell, size=Weight, color=Weight)) +
  geom_point() +
  labs(title="Cell type contribution", x="Factor", y = "Cell type", 
       color = "", size = "Scaled weight") +
  scale_color_viridis_c(direction = -1, option = "magma") +
  theme_bw() +
  theme(axis.text.x = element_text(angle=45, hjust=1), 
        panel.grid = element_blank())
```

## Mouse kidney {#mkidney}

In the second example, we'll look at data from mouse kidney. We can obtain the single-cell data with the `TabulaMurisSenisData` R package from bioconductor. Let's load the data and create a `Seurat` object from it.

```{r eval=FALSE}
#| code-fold: true
sce <- TabulaMurisSenisDroplet(tissues = "Kidney")$Kidney
umis <- as(counts(sce), "dgCMatrix")
se_kidney_singlecell <- CreateSeuratObject(counts = umis, meta.data = colData(sce) |> as.data.frame())

```

The 10x Visium mouse kidney data can be downloaded from the 10x genomics website.

```{r eval=FALSE}
#| code-fold: true

dir.create(paste0(tmpdir, "/kidney"))
targetdir <- paste0(tmpdir, "/kidney")
dir.create(paste0(targetdir, "/visium"))

# Download section 1
download.file(url = "https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Kidney/V1_Mouse_Kidney_filtered_feature_bc_matrix.h5", 
              destfile = paste0(targetdir, "/visium/filtered_feature_bc_matrix.h5"))
download.file(url = "https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Kidney/V1_Mouse_Kidney_spatial.tar.gz",
              destfile = paste0(targetdir, "/visium/spatial.tar.gz"))
untar(tarfile = paste0(targetdir, "/visium/spatial.tar.gz"), 
      exdir =  paste0(targetdir, "/visium/"))
file.remove(paste0(targetdir, "/visium/spatial.tar.gz"))

samples <- "./kidney/visium/filtered_feature_bc_matrix.h5"
imgs <- "./kidney/visium/spatial/tissue_hires_image.png"
spotfiles <- "./kidney/visium/spatial/tissue_positions_list.csv"
json <- "./kidney/visium/spatial/scalefactors_json.json"

infoTable <- tibble::tibble(samples, imgs, spotfiles, json)

se_kidney_spatial <- ReadVisiumData(infoTable)

```

### Normalize data

We apply the same normalization procedure to `se_kidney_spatial` and `se_kidney_singlecell` and run `FindVariableFeatures()` to detect the top most variable genes.

```{r eval=FALSE}

se_kidney_spatial <- se_kidney_spatial |>
  NormalizeData() |>
  FindVariableFeatures(nfeatures = 10000)

```

For the single-cell kidney data, we'll also filter the data prior to normalization to include cells collected at age 18m and remove cells with labels "nan" and "CD45". This leaves us with 17 cell types.

```{r eval=FALSE}

keep_cells <- colnames(se_kidney_singlecell)[se_kidney_singlecell$age == "18m" & (!se_kidney_singlecell$free_annotation %in% c("nan", "CD45"))]
se_kidney_singlecell <- subset(se_kidney_singlecell, cells = keep_cells)

```

```{r eval=FALSE}

se_kidney_singlecell <- se_kidney_singlecell |>
  NormalizeData() |>
  FindVariableFeatures() |> 
  ScaleData() |> 
  RunPCA() |> 
  RunUMAP(reduction = "pca", dims = 1:30)

se_kidney_singlecell <- se_kidney_singlecell |> 
  FindVariableFeatures(nfeatures = 10000)

```

```{r include=FALSE}
# Load prepared kidney data
se_kidney_singlecell <- readRDS("~/10x_data/R_objects/se_kidney_singlecell")
se_kidney_spatial <- readRDS("~/10x_data/R_objects/se_kidney_spatial")

```

### Run NNLS

Again, the `RunNNLS()` method requires a single-cell `Seurat` object and a 10x Visium `Seurat` object. The cell type annotations are stored in the "free_annotation" column.

```{r}
  
ti <- Sys.time()
DefaultAssay(se_kidney_spatial) <- "Spatial"
se_kidney_spatial <- RunNNLS(object = se_kidney_spatial, 
                      singlecell_object = se_kidney_singlecell, 
                      groups = "free_annotation")
sprintf("RunNNLS finished in %s seconds", round(Sys.time() - ti, digits = 2))

# Check available cell types
rownames(se_kidney_spatial)

```

NB: two cell types were discarded as they didn't pass the minimum allowed cells per cell type threshold. The cell types discarded were NK cells and plasma cells.

```{r class.source = 'fold-hide'}
# Plot selected cell types
DefaultAssay(se_kidney_spatial) <- "celltypeprops"

selected_celltypes <- c("Epcam     kidney distal convoluted tubule epithelial cell",
                        "Epcam    brush cell",
                        "Epcam    kidney collecting duct principal cell",
                        "Epcam    kidney proximal convoluted tubule epithelial cell",
                        "Epcam    podocyte",
                        "Epcam    proximal tube epithelial cell",
                        "Epcam    thick ascending tube S epithelial cell",
                        "Pecam    fenestrated capillary endothelial",
                        "Pecam    kidney capillary endothelial cell",
                        "Stroma    kidney mesangial cell")
plots <- lapply(seq_along(selected_celltypes), function(i) {
  MapFeatures(se_kidney_spatial, pt_size = 1.3,
            features = selected_celltypes[i], 
            override_plot_dims = TRUE) & 
  theme(plot.title = element_blank())
}) |> setNames(nm = selected_celltypes)

```

### Mapped kidney cell types {.tabset}

```{r results='asis', echo=FALSE}
for(i in 1:length(selected_celltypes)){
  cat("#### ", selected_celltypes[i], "\n")
  cat("\n")
  print(plots[[i]])
  cat("\n\n")
}
```

```{r eval=FALSE, include=FALSE}
saveRDS(se_kidney_singlecell, file = "~/10x_data/R_objects/se_kidney_singlecell")
saveRDS(se_kidney_spatial, file = "~/10x_data/R_objects/se_kidney_spatial")
```

### Visualize multiple cell types

We can select some of these cell types and visualize them in one plot

```{r message=FALSE, fig.asp=0.4, fig.width=10}

# Load H&E images
se_kidney_spatial <- se_kidney_spatial |> 
  LoadImages()

# Plot multiple features
MapMultipleFeatures(se_kidney_spatial, 
                    image_use = "raw", 
                    pt_size = 2, max_cutoff = 0.95,
                    override_plot_dims = TRUE, 
                    colors = c("#332288", "#88CCEE", "#117733", "#DDCC77", "#CC6677","#AA4499"),
                    features = selected_celltypes[c(1, 2, 3, 4, 6, 7)]) +
  plot_layout(guides = "collect")

```

Or zoom in on a region of interest

```{r message=FALSE, fig.asp=0.8, fig.width=10}

# Reload H&E image in higher resolution
se_kidney_spatial <- LoadImages(se_kidney_spatial, image_height = 1500)

# Plot H&E image
rst <- GetImages(se_kidney_spatial)[[1]]
p1 <- ggplot()  +
  ggtitle("H&E image of zoomed in area") + 
  theme(plot.title = element_text(hjust = 0.2)) +
  inset_element(p = rst[(0.5*nrow(rst)):(0.7*nrow(rst)), (0.5*ncol(rst)):(0.7*ncol(rst))], 
                left = 0, bottom = 0, right = 1, top = 1)

# Plot multiple features with zoom
p2 <- MapMultipleFeatures(se_kidney_spatial, 
                    image_use = "raw", 
                    pt_size = 4.5, max_cutoff = 0.95,
                    crop_area = c(0.5, 0.5, 0.7, 0.7),
                    colors = c("#332288", "#88CCEE", "#117733", "#DDCC77", "#CC6677","#AA4499"),
                    features = selected_celltypes[c(1, 2, 3, 4, 6, 7)]) +
  plot_layout(guides = "collect") &
  theme(plot.title = element_blank())

p1 / p2

```

<br>

------------------------------------------------------------------------

<details open>

<summary>**Package versions**</summary>

-   `semla`: `r packageVersion("semla")`

-   `RcppML`: `r packageVersion("RcppML")`

</details>

<details>

<summary>**Session info**</summary>

```{r}
sessionInfo()
```

</details>

<br>
