---
title: "Digital unrolling"
date: "Last compiled: `r format(Sys.time(), '%d %B %Y')`"
format:
  html:
    default-image-extension: png
editor: visual
---

In this tutorial, we will look at how to perform digital unrolling of folded "swiss roll" tissue sections.

The data set comes from a "swiss roll" of a mouse colon which was published by M. Parigi et al. (including yours truly) [*The spatial transcriptomic landscape of the healing mouse intestine following damage*](https://www.nature.com/articles/s41467-022-28497-0).

```{r warning=FALSE, message=FALSE}
library(semla)
```

## Mouse colon data

Let's load and have a quick look at the data.

```{r}
se_mcolon <- readRDS(system.file("extdata/mousecolon", 
                                 "se_mcolon", 
                                 package = "semla"))

se_mcolon <- se_mcolon |> 
  LoadImages()

ImagePlot(se_mcolon)
```

In the H&E image, we can clearly see the organization of large part of the lower intestine. The outermost part of the roll is the ascending colon and the center of the roll is the rectum.

We can identify genes with a regionalized pattern pretty easily but it would we great if we could somehow unroll the tissue so that we can investigate gene expression patterns along the proximal-distal axis.

Below is a spatial map of the expression of Fxyd4 which is up-regulated closer to the distal part of the colon.

```{r}
MapFeatures(se_mcolon, features = "Fxyd4", 
            image_use = "raw", override_plot_dims = TRUE,
            colors = c("darkblue", "cyan", "yellow", "red", "darkred"))
```

## Pre-processing

The goal of this tutorial is to show how `semla` can be used to find an "unrolled" coordinate system. To achieve this, we can start a shiny application with `CutSpatialNetwork()` which will allow us to do some interesting stuff. More about that later.

Before we get started, we need to do a little bit of pre-processing. First, we will tile the H&E image using `TileImage()`. Tiling is necessary to make it possible to visualize the H&E images in a more interactive way. The path to the raw H&E image can be found with `GetStaffli(se_mcolon)@imgs`.

NB: If the image file has been moved, `TileImage` the path might no longer be valid. Make sure that the path is valid before proceeding. To get decent resolution, we you should use the "tissue_hires_image.png" output by spaceranger.

```{r}
im <- GetStaffli(se_mcolon)@imgs[1]
sprintf("File exists: %s", file.exists(im))
```

Now we can load the image with `magick`, tile the image and export it to a local directory. The function returns the path to the directory the tiles were exported to. We will save this to a variable called `tilepath`

```{r eval=FALSE}
library(magick)
im <- image_read(im)
tilepath <- TileImage(im = im, outpath = "~/Downloads/")
```

The `CutSpatialNetwork()` function requires another file to work properly. This file holds information about the "spatial network" (see `GetSpatialNetwork()` for details) which will be used in the shiny application to define how spots in the Visium data are connected to each other.

To demonstrate what this means, we can generate a spatial network and visualize it.

```{r}
spatnet <- GetSpatialNetwork(se_mcolon)[[1]]

library(ggplot2)

ggplot() +
  geom_segment(data = spatnet, aes(x = x_start, xend = x_end, y = y_start, yend = y_end)) +
  scale_y_reverse() +
  coord_fixed()
```

In the plot above, you can see the edges that connects neighboring spots to each other.

Now let's export this spatial network to make it available for our shiny application.

```{r eval=FALSE}
export_graph(se_mcolon, sampleID = 1, outdir = tilepath$datapath)
```

## Prepare network

Now we are ready to run the app! Once the app loads, you should see the H&E image with spots and edges. The goal is to disconnect the layer of the roll by cutting the edges between spots in different layers. Cutting can be done by holding the `SHIFT` key while move the cursor across edges which will make them red. You can also "repair" edges by holding the `CTRL` key.

The image below illustrates how edges between two layers have been sliced.

![*Sliced edges are colored in red*](cut.png)

Once all edges has been cut, you have to press the `Quit & Save` button. If you just close the app without pressing "Quit & Save", the results will not be returned back to R.

NB: We need to make sure to save the output to a variable. Here, we will save the output to `tidy_network`.

```{r include=FALSE}
tidy_network <- readRDS("~/semla/repo/tests/digital_unrolling_mcolon")
#saveRDS(tidy_network, "~/semla/repo/tests/digital_unrolling_mcolon")
#export_tidygraph(tidy_network, outdir = tilepath)
```

```{r eval=FALSE}
tidy_network <- CutSpatialNetwork(se_mcolon, datadir = tilepath$datapath)
```

For downstream steps, it is *crucial* that the layers have been separated appropriately. Even a single edge between two layer will make it impossible to run down stream steps reliably. The image below demonstrates a situation where the separation will fail.

![*The arrow points at two edges (colored in black) which still connects two separate layers*](wrong.png)

Luckily, any changes you made in the app are by default written to the spatial network file, so if you miss something, you can simply run the app again and fix the issue.

Let's plot our full network and our filtered network:

```{r warning=FALSE, message=FALSE, fig.asp=0.6}
library(tidygraph)
library(patchwork)
library(dplyr)

edges_full <- tidy_network |> 
  activate(edges) |> 
  as_tibble()
edges_filtered <- tidy_network |> 
  activate(edges) |> 
  as_tibble() |> 
  filter(keep)

p1 <- ggplot(edges_full, 
       aes(x, xend = x_end, y, yend = y_end)) +
  geom_segment() +
  scale_y_reverse() +
  coord_fixed() +
  ggtitle("Full network")

p2 <- ggplot(edges_filtered, 
       aes(x, xend = x_end, y, yend = y_end)) +
  geom_segment() +
  scale_y_reverse() +
  coord_fixed() +
  ggtitle("Filtered network")

p1 + p2
```

And just like that, we have separated the the layers!

## Unrolling the network

Now comes the part where we do the actual "unrolling" using `AdjustTissueCoordinates()`. There are a few important things to note here. `AdjustTissueCoordinates` applies an algorithm to sort nodes in the spatial graph between the end points. Distances between nodes are used as a proxy for the actual distances in the tissue. These distances between nodes are the shortest paths, i.e. the minimum number of edges needed to visit (geodesic). For obvious reasons, distances cannot be calculated between nodes that are not connected, and for this reason, the function only handles connected graphs.

Unfortunately, our roll is not fully connected, so the function will only deal with the largest sub graph.

```{r}
all_networks <- to_components(tidy_network |> 
                                activate(edges) |> 
                                filter(keep))
edges_labeled <- do.call(bind_rows, lapply(seq_along(all_networks), function(i) {
  all_networks[[i]] |> 
    activate(edges) |> 
    as_tibble() |> 
    mutate(graph = paste0("sub graph", i))
}))

ggplot(edges_labeled, aes(x, xend = x_end, y, yend = y_end)) +
  geom_segment() +
  scale_y_reverse() +
  coord_fixed() +
  facet_grid(~graph)
```

The second, smaller sub graph will be discarded.

Another thing to note is that this function relies on a few assumptions about the shape of the folded tissue. If you were to try it on somethings completely different, let's say a tissue folded in a zig-zag shape, it might not work as well.

```{r}
tidy_network_adjusted <- AdjustTissueCoordinates(full_graph = tidy_network)
```

Our `tidy_network_adjusted` variable holds a `tibble` with the node coordinates and the new `x_dist` and `y_dist` coordinates. As a sanity check, we can color our original spatial map by the new distance values to see if they make sense.

```{r fig.asp=0.6, fig.width=9}
p1 <- ggplot(tidy_network_adjusted, aes(x, y, color = x_dist)) +
  geom_point() +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 9, name = "Spectral") |> 
                          rev()) +
  coord_fixed() +
  scale_y_reverse()

p2 <- ggplot(tidy_network_adjusted, aes(x, y, color = y_dist)) +
  geom_point() +
  scale_color_gradientn(colours = RColorBrewer::brewer.pal(n = 9, name = "Spectral") |> 
                          rev()) +
  coord_fixed() +
  scale_y_reverse()

p1 + p2
```

The `x_dist` values, which is arguably the most relevant for regionalization patterns, looks quite good. `y_dist` corresponds to the geodesic distances from the nodes along the shortest path, where negative values indicate that the spots are located outside the shortest path and positive values indicate that the spots are located inside the shortest path.

Now that we have our new coordinate system, we can start looking at regionalization of gene expression along the proximal-distal axis.

```{r warning=FALSE, message=FALSE, fig.asp=0.6, fig.width=8}
library(tibble)
library(tidyr)

selected_genes <- c("Fxyd4", "Slc37a2", "Hmgcs2", "Cyp4b1")

se_mcolon <- ScaleData(se_mcolon)
gg <- FetchData(se_mcolon, vars = selected_genes, slot = "scale.data") |> 
  rownames_to_column(var = "name") |> 
  left_join(y = tidy_network_adjusted, by = "name") |> 
  pivot_longer(all_of(selected_genes), names_to = "variable", values_to = "value")

ggplot(gg, aes(x_dist, value, color = variable)) +
  geom_smooth()
```

With these coordinates in our hands, it is quite straightforward to extract genes whose expression depend on distance along the proximal-distal axis. This will not be part of this tutorial, but below is a rather simple way of sorting genes by expression in a heatmap. Here, it becomes quite clear what genes are regionalized. There is one part of the heatmap that stands out, just on the right side of the center, which are the genes that are expressed in a lymphoid structure located in that part of the colon.

```{r warning=FALSE, message=FALSE, fig.asp=3}
sorted_spots <- tidy_network_adjusted |> 
  arrange(x_dist) |> 
  pull(name)

# Get scale data
scale_data <- GetAssayData(se_mcolon, 
                           slot = "scale.data", 
                           assay = "Spatial")[VariableFeatures(se_mcolon), sorted_spots] |> 
  as.matrix()

# Bin data
bins <- cut(1:ncol(scale_data), breaks = 50)
binMat <- do.call(cbind, lapply(levels(bins), function(lvl) {
  rowMeans(scale_data[, bins == lvl])
}))

pheatmap::pheatmap(binMat, cluster_cols = FALSE, border_color = FALSE,
                   breaks = seq(-1, 1, length.out = 50), 
                   color = scico::scico(palette = "bam", n = 51))
```

## Other applications

The tools described here could also be used for other applications. Maybe we are interested in a specific part of a tissue where we want to define an axis. Let's make a manual selection of the outermost layer of the mouse brain and subset the `Seurat` object to contain spots from this layer.

```{r eval=FALSE}
se_mbrain <- readRDS(system.file("extdata/mousebrain", "se_mbrain", package = "semla"))
se_mbrain <- LoadImages(se_mbrain)
se_mbrain$region <- "NA"
se_mbrain <- FeatureViewer(se_mbrain)

se_mbrain_subset <- SubsetSTData(se_mbrain, expression = region == "outer_layer")
```

```{r fig.asp=1, fig.width=6, eval=FALSE}
MapLabels(se_mbrain_subset, column_name = "region", image_use = "raw")
```

```{r include=FALSE, eval=FALSE}
jpeg(filename = "digital_unrolling/mBrain.jpg", width = 1e3, height = 1e3, res = 200)
MapLabels(se_mbrain_subset, column_name = "region", image_use = "raw")
dev.off()
```

![](digital_unrolling/mBrain.jpg)

Then we can create a spatial network from this layer and modify it a bit to make it compatible with `AdjustTissueCoordinates`:

```{r eval=FALSE}
spatial_network <- GetSpatialNetwork(se_mbrain_subset)[[1]]
nodes <- spatial_network |> select(from, x_start, y_start) |> group_by(from) |> 
  slice_head(n = 1) |> rename(name = from, x = x_start, y = y_start)
edges <- spatial_network |> select(from, to) |> mutate(keep = TRUE)
spatial_network <- tidygraph::tbl_graph(nodes = nodes, edges = edges, directed = FALSE)
```

Now that we have the spatial network, we can run `AdjustTissueCoordinates` to compute distances between the end points of the selected region:

```{r eval=FALSE}
# Run adjust tissue coordinates
test <- AdjustTissueCoordinates(full_graph = spatial_network)
se_mbrain_subset$x_dist <- test[match(test$name, colnames(se_mbrain_subset)), "x_dist", drop = TRUE]

# Plot x_dist
MapFeatures(se_mbrain_subset, features = "x_dist", image_use = "raw", 
            colors = RColorBrewer::brewer.pal(n = 9, name = "Spectral"), 
            override_plot_dims = TRUE, pt_size = 2.5) &
  theme(plot.title = element_blank())

```

```{r include=FALSE, eval=FALSE}
jpeg(filename = "digital_unrolling/mBrain_x_dist.jpg", width = 1e3, height = 1e3, res = 200)
MapFeatures(se_mbrain_subset, features = "x_dist", image_use = "raw", 
            colors = RColorBrewer::brewer.pal(n = 9, name = "Spectral"), 
            override_plot_dims = TRUE, pt_size = 2.5) &
  theme(plot.title = element_blank())
dev.off()
```

![](digital_unrolling/mBrain_x_dist.jpg)

<br>

------------------------------------------------------------------------

<details open>

<summary>**Package version**</summary>

-   `semla`: `r packageVersion("semla")`

</details>

<details>

<summary>**Session info**</summary>

```{r}
sessionInfo()
```

</details>

<br>
