<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>semla - analysis of SRT data • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="semla - analysis of SRT data">
<meta name="description" content="Spatially Resolved Transcriptomics data exploration and analysis. semla provides interactive tools for image alignment and data exploration together with several analysis and visualization methods.">
<meta property="og:description" content="Spatially Resolved Transcriptomics data exploration and analysis. semla provides interactive tools for image alignment and data exploration together with several analysis and visualization methods.">
<meta property="og:image" content="https://ludvigla.github.io/semla/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/semla.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><h6 class="dropdown-header" data-toc-skip>Intro</h6></li>
    <li><a class="dropdown-item" href="articles/getting_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="articles/subset_and_merge.html">Subset/merge</a></li>
    <li><a class="dropdown-item" href="articles/visiumHD.html">VisiumHD</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6></li>
    <li><a class="dropdown-item" href="articles/numeric_features.html">Map numeric features</a></li>
    <li><a class="dropdown-item" href="articles/categorical_features.html">Map categorical features</a></li>
    <li><a class="dropdown-item" href="articles/advanced_visualization.html">Advanced visualization</a></li>
    <li><a class="dropdown-item" href="articles/feature_viewer.html">Feature viewer</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial methods</h6></li>
    <li><a class="dropdown-item" href="articles/radial_distances.html">Radial distances</a></li>
    <li><a class="dropdown-item" href="articles/region_neighbors.html">Region neighbors</a></li>
    <li><a class="dropdown-item" href="articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a></li>
    <li><a class="dropdown-item" href="articles/spatial_autocorrelation.html">Spatial autocorrelation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Analysis</h6></li>
    <li><a class="dropdown-item" href="articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a></li>
    <li><a class="dropdown-item" href="articles/digital_unrolling.html">Digital unrolling</a></li>
    <li><a class="dropdown-item" href="articles/NNMF.html">Non-negative Matrix Factorization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Image processing</h6></li>
    <li><a class="dropdown-item" href="articles/images_and_coordinates.html">Images and spot coordinates</a></li>
    <li><a class="dropdown-item" href="articles/mask_images.html">Image masking</a></li>
    <li><a class="dropdown-item" href="articles/image_alignment.html">Image alignment</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/create_object.html">Create a Staffli object</a></li>
    <li><a class="dropdown-item" href="articles/compare_cell_type_mapping_NNLS.html">Cell type mapping comparison</a></li>
    <li><a class="dropdown-item" href="articles/cell_type_mapping_benchmark.html">Cell type mapping benchmark</a></li>
    <li><a class="dropdown-item" href="articles/slide-seq.html">Slide-Seq data</a></li>
    <li><a class="dropdown-item" href="articles/IF_data.html">IF data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="semla-">semla <a class="anchor" aria-label="anchor" href="#semla-"></a>
</h1>
</div>
<!-- badges: start -->

<!-- badges: end -->
<p><br></p>
<p><code>semla</code> is an R package that collects useful tools for Spatially Resolved Transcriptomics data analysis and visualization.</p>
<p>If you are visiting our Github page, please find more information at our package <a href="https://ludvigla.github.io/semla/">website</a>. Here, you can find documentation of functions together with examples on how to use them, as well as tutorials showing how to use <code>semla</code> for analysis and visualization of 10x Visium data.</p>
<p>For a more scientific description of the package, check out our associated publication “<i>Semla: a versatile toolkit for spatially resolved transcriptomics analysis and visualization</i>” in Bioinformatics (<a href="https://doi.org/10.1093/bioinformatics/btad626" class="external-link">DOI:10.1093/bioinformatics/btad626</a>). Please cite this article if you use <code>semla</code> in your studies, and we’d be very grateful!</p>
<p><br></p>
<blockquote>
<p>[!NOTE]<br>
From version <code>1.4.0</code> onwards, <code>semla</code> has a strict dependency on <a href="https://satijalab.org/seurat/" class="external-link">Seurat v5</a>. Please open an issue if you run into any problems.</p>
</blockquote>
<p><br></p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The dev version of the package can be installed through GitHub using;</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ludvigla/semla"</span><span class="op">)</span></span></code></pre>
<p><br></p>
<div class="section level3">
<h3 id="setting-up-a-conda-environment">Setting up a conda environment<a class="anchor" aria-label="anchor" href="#setting-up-a-conda-environment"></a>
</h3>
<p>If you are using a <a href="https://docs.conda.io/en/latest/miniconda.html" class="external-link">conda</a>, you can follow the steps below:</p>
<pre><code>conda create -n r-semla r-essentials r-base</code></pre>
<p>Activate the environment:</p>
<pre><code>conda activate r-semla</code></pre>
<p>Then you can open RStudio from the environment. This should make sure that RStudio uses the R version and packages that are located in the conda environment. On Mac OS, you can open RStudio by typing the path to the RStudio application in the terminal:</p>
<pre><code>/Applications/RStudio.app/Contents/MacOS/RStudio</code></pre>
<p>Note that <code>semla</code> requires R v4.1 or higher, so you should double check that you have the correct version installed.</p>
<p>Now you should be able to install <code>semla</code> in your conda environment:</p>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"ludvigla/semla"</span><span class="op">)</span></span></code></pre>
<p>The <code>magick</code> R package might fail to install in your conda environment and in that case you can install it with conda instead:</p>
<pre><code>conda install -c conda-forge r-magick</code></pre>
<p><br></p>
</div>
<div class="section level3">
<h3 id="renv">Renv<a class="anchor" aria-label="anchor" href="#renv"></a>
</h3>
<p>If you are familiar with <a href="https://rstudio.github.io/renv/articles/renv.html" class="external-link">renv</a>, you can install all necessary R packages (with exact versions) using <code>renv::restore()</code> with the <code>renv.lock</code> file provided in our GitHub repo.</p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="docker">Docker<a class="anchor" aria-label="anchor" href="#docker"></a>
</h3>
<p>As an additional option, we provide a docker image on <a href="https://hub.docker.com/r/ludlar/semla" class="external-link">Docker Hub</a> that you can download to run a container from. The image is based on the <a href="https://hub.docker.com/r/rocker/rstudio" class="external-link">rocker</a> RStudio Server image and comes with RStudio Server pre-installed.</p>
<p>To access the server we need to publish a port for our container, we’ll use the –publish flag (-p for short) on the docker run command. The format of the –publish command is [host port]:[container port]. So, if we wanted to expose port 8000 inside the container to port 8080 outside the container, we would pass 8080:8000 to the –publish flag. We can also expose multiple ports which will be useful to use some of the interactive features in <code>semla</code> (see our <code>Interactive Viewer</code> section <a href="https://ludvigla.github.io/semla/articles/feature_viewer.html">here</a>).</p>
<p><code>&lt;YOURPASSWORD&gt;</code> can be any password you want to use for RStudio Server and the default username is <code>rstudio</code>. <code>&lt;SOURCEPATH&gt;</code> should be a path on your local machine which will be accessible from the docker container at <code>/home/rstudio</code>. Use <code>source="$(pwd)"</code> for current working directory.</p>
<pre><code>sudo docker run -d -p 1337:8787 -p 3030:3030 --name semla -e PASSWORD=&lt;YOURPASSWORD&gt; --memory=16g --mount type=bind,source="&lt;SOURCEPATH&gt;",target=/home/rstudio -e ROOT=TRUE ludlar/semla:latest</code></pre>
<p>Visit the docker <a href="https://docs.docker.com/language/java/run-containers/" class="external-link">docs</a> for more information on how to run containers.</p>
<p><br></p>
</div>
</div>
<div class="section level2">
<h2 id="citing-semla">Citing <em>semla</em>
<a class="anchor" aria-label="anchor" href="#citing-semla"></a>
</h2>
<p>Larsson L, Franzén L, Ståhl PL, Lundeberg J. Semla: a versatile toolkit for spatially resolved transcriptomics analysis and visualization. Bioinformatics. 2023 Oct 3;39(10):btad626. doi: 10.1093/bioinformatics/btad626. PMID: 37846051; PMCID: PMC10597621. <a href="https://academic.oup.com/bioinformatics/article/39/10/btad626/7319366" class="external-link uri">https://academic.oup.com/bioinformatics/article/39/10/btad626/7319366</a></p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="what-is-semla">What is semla?<a class="anchor" aria-label="anchor" href="#what-is-semla"></a>
</h2>
<p>A semla is a delicious pastry, traditionally consumed around a specific day, “Semmeldagen” or “Fettisdagen” (“Fat Tuesday”), in Sweden. It is made up of a light cardamom flavored wheat bun, fluffy whipped cream, sweet crunchy almond paste, and topped with dusted icing sugar. If that’s not reason enough for us to name this package after this fantastic creation, let’s also say <em>semla</em> is an abbreviation for “<strong>S</strong>patially r<strong>E</strong>solved transcripto<strong>M</strong>ics too<strong>L</strong>s for <strong>A</strong>nalysis”.</p>
<p><br></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing semla</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-4209-2911" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-3755-718X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://github.com/jemorlanes" class="external-link">Javier Escudero Morlanes</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0009-0004-2885-2456" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>
<a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a> <br><small class="roles"> Funder </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://github.com/jemorlanes" class="external-link">Javier Escudero Morlanes</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
