<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>semla • semla</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png"><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet"><meta property="og:title" content="semla"><meta property="og:image" content="https://ludvigla.github.io/semla/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="articles/semla.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Intro</h6>
    <a class="dropdown-item" href="articles/getting_started.html">Get started</a>
    <a class="dropdown-item" href="articles/subset_and_merge.html">Subset/merge</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive</h6>
    <a class="dropdown-item" href="articles/feature_viewer.html">Feature viewer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6>
    <a class="dropdown-item" href="articles/numeric_features.html">Map numeric features</a>
    <a class="dropdown-item" href="articles/categorical_features.html">Map categorical features</a>
    <a class="dropdown-item" href="articles/advanced_visualization.html">Advanced visualization</a>
    <a class="dropdown-item" href="articles/scalebar.html">Scalebar</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial methods</h6>
    <a class="dropdown-item" href="articles/radial_distances.html">Radial distances</a>
    <a class="dropdown-item" href="articles/region_neighbors.html">Region neighbors</a>
    <a class="dropdown-item" href="articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a>
    <a class="dropdown-item" href="articles/spatial_autocorrelation.html">Spatial autocorrelation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Analysis</h6>
    <a class="dropdown-item" href="articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a>
    <a class="dropdown-item" href="articles/digital_unrolling.html">Digital unrolling</a>
    <a class="dropdown-item" href="articles/NNMF.html">Non-negative Matrix Factorization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image processing</h6>
    <a class="dropdown-item" href="articles/images_and_coordinates.html">Images and spot coordinates</a>
    <a class="dropdown-item" href="articles/mask_images.html">Image masking</a>
    <a class="dropdown-item" href="articles/image_alignment.html">Image alignment</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="logo.png" class="logo" alt=""><h1>semla</h1>
      
    </div>

<div id="semla" class="section level1">

<div class="section level2">
<h2 id="todo">TODO<a class="anchor" aria-label="anchor" href="#todo"></a></h2>
<div class="section level3">
<h3 id="external-data">External data<a class="anchor" aria-label="anchor" href="#external-data"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>: An external data resource with mouse brain data. Contains spaceranger output files and a prepared Seurat object. These files can be accessed with <code>system.file("extdata/mousebrain", "path_to_file", package = "semla")</code>
</li>
<li>
<input type="checkbox" disabled checked>: An external data resource with mouse colon data. Contains spaceranger output files and a prepared Seurat object. These files can be accessed with <code>system.file("extdata/mousecolon", "path_to_file", package = "semla")</code>
</li>
</ul></div>
<div class="section level3">
<h3 id="data-loading">Data loading<a class="anchor" aria-label="anchor" href="#data-loading"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><code>LoadAndMergeMatrices</code> : load and merge gene expression matrices</li>
<li>
<input type="checkbox" disabled checked><code>LoadSpatialCoordinates</code> : load and merge spatial coordinates</li>
<li>
<input type="checkbox" disabled checked><code>ReadVisiumData</code> : wrapper for <code>LoadAndMergeMatrices</code> and <code>LoadSpatialCoordinates</code> to load all data into a Seurat object. Dropped compatibility with “1k” and “2k” arrays. The function is similar to <code>InputFromTable</code>.</li>
<li>
<input type="checkbox" disabled checked><code>LoadImages</code> : load H&amp;E images into Seurat object</li>
</ul></div>
<div class="section level3">
<h3 id="subset-and-merge">Subset and Merge<a class="anchor" aria-label="anchor" href="#subset-and-merge"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><code>SubsetSTData</code> : for subsetting <code>Seurat</code> objects created with <code>semla</code>
</li>
<li>
<input type="checkbox" disabled checked><code>MergeSTData</code> : for merging <code>Seurat</code> objects created with <code>semla</code>
</li>
</ul></div>
<div class="section level3">
<h3 id="on-load">On load<a class="anchor" aria-label="anchor" href="#on-load"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><code>.onAttach</code> : on load message when loading R package</li>
</ul></div>
<div class="section level3">
<h3 id="external-data-1">External data<a class="anchor" aria-label="anchor" href="#external-data-1"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><em>Visium mouse brain dataset</em> : mouse brain test data</li>
<li>
<input type="checkbox" disabled checked><em>Visium mouse colon dataset</em> : mouse colon test data</li>
</ul></div>
<div class="section level3">
<h3 id="transformations">Transformations<a class="anchor" aria-label="anchor" href="#transformations"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><code>CoordMirror</code> : mirror coordinates at predefined center</li>
<li>
<input type="checkbox" disabled checked><code>CoordTransform</code> : apply rotations and translations to coordinates around predefined center</li>
<li>
<input type="checkbox" disabled checked><code>ImageTranslate</code> : apply translations to image</li>
<li>
<input type="checkbox" disabled checked><code>ImageTransform</code> : apply rotation and translation to image</li>
<li>
<input type="checkbox" disabled checked><code>CoordAndImageTransform</code> : wrapper for <code>ImageTransform</code> and <code>CoordTransform</code> to apply transformations to paired image and coordinates. Can do rotations, translations and mirroring</li>
</ul></div>
<div class="section level3">
<h3 id="image-processing">Image processing<a class="anchor" aria-label="anchor" href="#image-processing"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><code>MaskImages</code> uses a simple method to identify the tissue section and mask the background implemented with <code>magick</code>
</li>
<li>
<input type="checkbox" disabled checked><code>RigidTransformImages</code> takes a tibble with transformations and applies these transformations to the H&amp;E images together with the spot coordinates.</li>
</ul></div>
<div class="section level3">
<h3 id="tissue-viewer">Tissue viewer<a class="anchor" aria-label="anchor" href="#tissue-viewer"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>make OSD viewer from tile sourse</li>
<li>
<input type="checkbox" disabled checked>draw spots on top of tile source</li>
<li>
<input type="checkbox" disabled checked>lasso selection tool for manual annotation with paper JS</li>
<li>
<input type="checkbox" disabled checked>annotate spots and save selection</li>
<li>
<input type="checkbox" disabled checked>make tile function <code>TileImage</code> compatible with OSD</li>
<li>
<input type="checkbox" disabled checked>send data to OSD viewer from R or read from R on selection?</li>
</ul></div>
<div class="section level3">
<h3 id="tissue-aligner">Tissue aligner<a class="anchor" aria-label="anchor" href="#tissue-aligner"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>make paper JS tool for image tranformations</li>
<li>
<input type="checkbox" disabled checked>communicate transformations to R</li>
<li>
<input type="checkbox" disabled checked>apply transformations to images in R object with <code>CoordAndImageTransform</code>
</li>
<li>
<input type="checkbox" disabled checked>add side bar for image selection</li>
</ul></div>
<div class="section level3">
<h3 id="digital-unrolling">Digital unrolling<a class="anchor" aria-label="anchor" href="#digital-unrolling"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><code>CutSpatialNetwork</code> : Open a shiny/react application that can be used for “digital unrolling”</li>
</ul></div>
<div class="section level3">
<h3 id="visualization-tools">Visualization tools<a class="anchor" aria-label="anchor" href="#visualization-tools"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><code>FeatureOverlay</code> : this function has now been removed and instead we use <code>MapFeatures</code> which now takes an argument <code>image_use</code> to inject images under the plot.</li>
<li>
<input type="checkbox" disabled checked><code>DimOverlay</code> : this function has now been removed and instead we use <code>MapFeatures</code> which now takes an argument <code>image_use</code> to inject images under the plot.</li>
<li>
<input type="checkbox" disabled checked><code>MapFeatures</code> : similar to <code>ST.FeaturePlot</code> and <code>ST.DimPlot</code> but only works with numeric features such as gene expression, PCs, QC metrics etc.</li>
<li>
<input type="checkbox" disabled checked><code>MapLabels</code> : similar to <code>ST.FeaturePlot</code> but only works with categorical features such as clusters, conditions etc.</li>
<li>
<input type="checkbox" disabled checked><code>FactorGeneLoadingPlot</code> : Not sure if needed with the singlet R package available</li>
<li>
<input type="checkbox" disabled checked><code>ImagePlot</code> : plots H&amp;E images stored in a Seurat object</li>
<li>
<input type="checkbox" disabled checked><code>MapFeaturesSummary</code> : create spatial maps with <code>MapFeatures</code> and add a summary plot next to it</li>
<li>
<input type="checkbox" disabled checked><code>MapLabelsSummary</code> : create spatial maps with <code>MapLabels</code> and add a summary plot next to it</li>
<li>
<input type="checkbox" disabled checked><code>AnglePlot</code> : draw a set of arrows on top of a selected region of spots. Mean to be used as a helper for selecting angle intervals in <code>RadialDistance</code>
</li>
</ul></div>
<div class="section level3">
<h3 id="matrix-factorization">Matrix factorization<a class="anchor" aria-label="anchor" href="#matrix-factorization"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><code>RunNMF</code> : no longer needed with the singlet R package available</li>
<li>
<input type="checkbox" disabled checked><code>SeededNMF</code> : seed matrix factorization with a priori information fro single-cell data</li>
</ul></div>
<div class="section level3">
<h3 id="spatial-autocorelation">Spatial autocorelation<a class="anchor" aria-label="anchor" href="#spatial-autocorelation"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>migrate <code>GetSpatNet</code> from STUtility</li>
<li>
<input type="checkbox" disabled checked>migrate <code>CorSpatialGenes</code> from STUtility and rename it to <code>CorSpatialFeatures</code>. This is a more general function that takes any type of feature as input, not just genes. This could for example be dimensionality reduction results instead.</li>
</ul></div>
<div class="section level3">
<h3 id="spatial-methods">Spatial methods<a class="anchor" aria-label="anchor" href="#spatial-methods"></a></h3>
<ul><li>
<input type="checkbox" disabled checked>migrate <code>RegionNeighbours</code> from STUtility</li>
<li>
<input type="checkbox" disabled checked><code>RadialDistance</code> : calculates the radial distances from the borders of a region of interest</li>
<li>
<input type="checkbox" disabled checked><code>DisconnectRegions</code> : separates spatially disconnected regions</li>
<li>
<input type="checkbox" disabled checked><code>NeighbourhoodEnrichmentTest</code> : Pairwise testing of cluster co-occurance. Start with checking nearest neighbors. Later, add possibility of measuring by distance(?).</li>
<li>
<input type="checkbox" disabled checked><code>LabelAssortativityTest</code> : Inspired by Newman’s Assortativity. Check whether selection of spots follow a dispersed or clustered spatial pattern.</li>
<li>[] Add <code>CentralityScore</code> method?</li>
</ul></div>
<div class="section level3">
<h3 id="staffli">Staffli<a class="anchor" aria-label="anchor" href="#staffli"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked>migrate and modify STUtility S4 class object <code>Staffli</code>
</li>
</ul></div>
<div class="section level3">
<h3 id="semla-website">semla website<a class="anchor" aria-label="anchor" href="#semla-website"></a></h3>
<ul class="task-list"><li>
<input type="checkbox" disabled checked><em>getting started</em>
</li>
<li>
<input type="checkbox" disabled checked><em>visualization</em>
</li>
<li>
<input type="checkbox" disabled checked><em>spatial methods</em>
</li>
</ul></div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

