<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="semla">
<title>Cell type mapping with NNLS • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Cell type mapping with NNLS">
<meta property="og:description" content="semla">
<meta property="og:image" content="https://ludvigla.github.io/semla/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/semla.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Intro</h6>
    <a class="dropdown-item" href="../articles/getting_started.html">Get started</a>
    <a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive</h6>
    <a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6>
    <a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a>
    <a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a>
    <a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a>
    <a class="dropdown-item" href="../articles/scalebar.html">Scalebar</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial methods</h6>
    <a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a>
    <a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a>
    <a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a>
    <a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Analysis</h6>
    <a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a>
    <a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a>
    <a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image processing</h6>
    <a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a>
    <a class="dropdown-item" href="../articles/mask_images.html">Image masking</a>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Cell type mapping with NNLS</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 15 March
2023</h4>
      
      
      <div class="d-none name"><code>cell_type_mapping_with_NNLS.rmd</code></div>
    </div>

    
    
<p>Cell type mapping with SRT data refers to a set of methods which
allows you to infer the quantity of cells from SRT expression profiles.
<code>semla</code> offers a quick method based on <em>Non-Negative Least
Squares</em> to infer cell type proportions directly from Visium spot
expression profiles. We utilize the NNLS method implemented in the <a href="https://github.com/zdebruine/RcppML" class="external-link">RcppML</a> <em>v0.3.7</em>
package available on CRAN: <a href="https://cran.r-project.org/web/packages/RcppML/index.html" class="external-link uri">https://cran.r-project.org/web/packages/RcppML/index.html</a>.
Please ensure you have the correct version installed before proceeding
with this tutorial.</p>
<p>In this tutorial, we’ll take a look at two examples. First, we have
two tissue sections of a <a href="#mbrain">mouse brain</a> and a single
cell RNA-seq dataset from the Allen Brain Atlas. In the second example,
we have a <a href="#mkidney">mouse kidney</a> section and single cell
RNA-seq data from the Tabula Muris Senis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ludvigla.github.io/semla/">semla</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">'TabulaMurisSenisData'</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"TabulaMurisSenisData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmicompbio/TabulaMurisSenisData" class="external-link">TabulaMurisSenisData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="mbrain">Mouse brain<a class="anchor" aria-label="anchor" href="#mbrain"></a>
</h2>
<p>For this tutorial, we will use a single-cell data set from the Allen
Brain Atlas. You can check the code chunk below or follow <a href="https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1" class="external-link">this</a>
link to download it manually.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>timeout<span class="op">=</span><span class="fl">200</span><span class="op">)</span></span>
<span><span class="va">tmpdir</span> <span class="op">&lt;-</span> <span class="st">"."</span> <span class="co"># Set current wd or change to tmpdir()</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">tmpdir</span>, <span class="st">"/mousebrain"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">targetdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">tmpdir</span>, <span class="st">"/mousebrain"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/single-cell"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">destfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/single-cell/allen_brain.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1"</span>, destfile <span class="op">=</span> <span class="va">destfile</span><span class="op">)</span></span></code></pre></div>
<p>We will also need to download the 10x Visium data from 10x Genomics
website. You can download the files directly with R by following the
code chunk below or download the data directly from <a href="https://www.10xgenomics.com/resources/datasets" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download section 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S1"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.h5"</span>, </span>
<span>              destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S1/filtered_feature_bc_matrix.h5"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S1/spatial.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span>tarfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S1/spatial.tar.gz"</span><span class="op">)</span>, </span>
<span>      exdir <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S1/"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S1/spatial.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download section 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S2"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Posterior/V1_Mouse_Brain_Sagittal_Posterior_filtered_feature_bc_matrix.h5"</span>, </span>
<span>              destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S2/filtered_feature_bc_matrix.h5"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Posterior/V1_Mouse_Brain_Sagittal_Posterior_spatial.tar.gz"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S2/spatial.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span>tarfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S2/spatial.tar.gz"</span><span class="op">)</span>, </span>
<span>      exdir <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S2/"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/S2/spatial.tar.gz"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Load Visium data with semla into a Seurat object. The following steps
assumes that the mouse brain 10x Visium data is located in
<code>./mousebrain/visium/</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Assemble spaceranger output files</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span><span class="st">"./mousebrain/visium/*/filtered_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span><span class="va">imgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span><span class="st">"./mousebrain/visium/*/spatial/tissue_hires_image.png"</span><span class="op">)</span></span>
<span><span class="va">spotfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span><span class="st">"./mousebrain/visium/*/spatial/tissue_positions_list.csv"</span><span class="op">)</span></span>
<span><span class="va">json</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span><span class="st">"./mousebrain/visium/*/spatial/scalefactors_json.json"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">infoTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">imgs</span>, <span class="va">spotfiles</span>, <span class="va">json</span>, </span>
<span>                    section_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"section_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create Seurat object with 1 Sagittal Anterior section and 1 Sagittal Posterior section</span></span>
<span><span class="va">se_brain_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReadVisiumData.html">ReadVisiumData</a></span><span class="op">(</span><span class="va">infoTable</span><span class="op">)</span></span></code></pre></div>
<p>Load single-cell data</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_allen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"./mousebrain/single-cell/allen_brain.rds"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="normalize-data">Normalize data<a class="anchor" aria-label="anchor" href="#normalize-data"></a>
</h3>
<p>Here we will apply the same log-normalization procedure to both the
10x Visium data (<code>se_brain_spatial</code>) and to the single-cell
data (<code>se_brain_singlecell</code>). We set the number of variable
features quite high because later on we will use the intersect between
the variable features in the single-cell data and the variable features
in the 10x Visium data for NNLS. The NNLS method is quite fast so there
is actually no need to select only a subset of features. Instead, we can
just use all genes that are shared across the single-cell and 10x Visium
data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Normalize data and find variable features for Visium data</span></span>
<span><span class="va">se_brain_spatial</span> <span class="op">&lt;-</span> <span class="va">se_brain_spatial</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Normalize data and run vanilla analysis to create UMAP embedding</span></span>
<span><span class="va">se_allen</span> <span class="op">&lt;-</span> <span class="va">se_allen</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Rerun FindVariableFeatures to increase the number before cell type deconvolution</span></span>
<span><span class="va">se_allen</span> <span class="op">&lt;-</span> <span class="va">se_allen</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
<p>We can visualize the available cell types in our UMAP embedding of
the cells. We have access to 23 annotated cell types, including L2-L6
layer neurons which have a distinct spatial distribution in the
tissue.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">se_allen</span>, group.by <span class="op">=</span> <span class="st">"subclass"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="run-nnls">Run NNLS<a class="anchor" aria-label="anchor" href="#run-nnls"></a>
</h3>
<p>The <code><a href="../reference/celltype-prediction.html">RunNNLS()</a></code> method requires a normalized
<code>Seurat</code> object with 10x Visium data and a a normalized
<code>Seurat</code> object with single-cell data. The
<code>groups</code> argument defines where the cell type labels should
be taken from in the single-cell <code>Seurat</code> object. In our
single-cell <code>Seurat</code> object, the labels are stored in the
“subclass” column.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">se_brain_spatial</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Spatial"</span></span>
<span></span>
<span><span class="va">ti</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">se_brain_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/celltype-prediction.html">RunNNLS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">se_brain_spatial</span>, </span>
<span>                            singlecell_object <span class="op">=</span> <span class="va">se_allen</span>, </span>
<span>                            groups <span class="op">=</span> <span class="st">"subclass"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## ── <span style="font-weight: bold;">Predicting cell type proportions</span> ──</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Fetching data from Seurat objects</span></span></code></pre>
<pre><code><span><span class="co">## →   Filtering out features that are only present in one data set</span></span></code></pre>
<pre><code><span><span class="co">## →   Kept 3378 features for deconvolution</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: RcppML</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Preparing data for NNLS</span></span></code></pre>
<pre><code><span><span class="co">## →   Downsampling scRNA-seq data to include a maximum of 50 cells per cell type</span></span></code></pre>
<pre><code><span><span class="co">## → <span style="color: #FF55FF;">  Cell type(s) CR have too few cells (&lt;10) and will be excluded</span></span></span></code></pre>
<pre><code><span><span class="co">## →   Kept 22 cell types after filtering</span></span></code></pre>
<pre><code><span><span class="co">## →   Calculating cell type expression profiles</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Predicting cell type proportions with NNLS for 22 cell types</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Returning results in a new 'Assay' named 'celltypeprops'</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Setting default assay to 'celltypeprops'</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Finished</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"RunNNLS completed in %s seconds"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ti</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "RunNNLS completed in 5.85 seconds"</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check available cell types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">se_brain_spatial</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Astro"      "Endo"       "L2/3 IT"    "L4"         "L5 IT"     </span></span>
<span><span class="co">##  [6] "L5 PT"      "L6 CT"      "L6 IT"      "L6b"        "Lamp5"     </span></span>
<span><span class="co">## [11] "Macrophage" "Meis2"      "NP"         "Oligo"      "Peri"      </span></span>
<span><span class="co">## [16] "Pvalb"      "SMC"        "Serpinf1"   "Sncg"       "Sst"       </span></span>
<span><span class="co">## [21] "VLMC"       "Vip"</span></span></code></pre>
<p>NB: the CR cell type was discarded because the number of cell was
lower than 10. 10 is the lower limit for the allowed number of cells per
cell type but this can be overridden with
<code>minCells_per_celltype</code>.</p>
<p>The plots below show the spatial distributions of proportions for a
selected set of cell types.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot selected cell types</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">se_brain_spatial</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltypeprops"</span></span>
<span></span>
<span><span class="va">selected_celltypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"L2/3 IT"</span>, <span class="st">"L4"</span>, <span class="st">"L5 IT"</span>, </span>
<span>                       <span class="st">"L5 PT"</span>, <span class="st">"L6 CT"</span>, <span class="st">"L6 IT"</span>, <span class="st">"L6b"</span>, </span>
<span>                       <span class="st">"Oligo"</span>, <span class="st">"Pvalb"</span>, <span class="st">"Meis2"</span>, <span class="st">"Astro"</span>,</span>
<span>                       <span class="st">"VLMC"</span>, <span class="st">"SMC"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_brain_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="va">se_brain_spatial</span>, image_height <span class="op">=</span> <span class="fl">1e3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## ── <span style="font-weight: bold;">Loading H&amp;E images</span> ──</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Loading image from ./mousebrain/visium/S1/spatial/tissue_hires_image.png</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Scaled image from 1998x2000 to 1000x1001 pixels</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Loading image from ./mousebrain/visium/S2/spatial/tissue_hires_image.png</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Scaled image from 1998x2000 to 1000x1001 pixels</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Saving loaded H&amp;E images as 'rasters' in Seurat object</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">selected_celltypes</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se_brain_spatial</span>, pt_size <span class="op">=</span> <span class="fl">1.3</span>,</span>
<span>            features <span class="op">=</span> <span class="va">selected_celltypes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, image_use <span class="op">=</span> <span class="st">"raw"</span>,</span>
<span>            arrange_features <span class="op">=</span> <span class="st">"row"</span>, scale <span class="op">=</span> <span class="st">"shared"</span>, </span>
<span>            override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colors <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">9</span>, name <span class="op">=</span> <span class="st">"Spectral"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>            scale_alpha <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span>, legend.margin <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">margin</a></span><span class="op">(</span>b <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>,</span>
<span>        legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>        plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span>nm <span class="op">=</span> <span class="va">selected_celltypes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Returning more (or less) than 1 row per `summarise()` group was deprecated in</span></span>
<span><span class="co">## dplyr 1.1.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use `reframe()` instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> When switching from `summarise()` to `reframe()`, remember that `reframe()`</span></span>
<span><span class="co">##   always returns an ungrouped data frame and adjust accordingly.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">semla</span> package.</span></span>
<span><span class="co">##   Please report the issue to the authors.</span></span></code></pre>
</div>
<div class="section level3 tabset">
<h3 class="tabset" id="mapped-mouse-brain-cell-types">Mapped mouse brain cell types<a class="anchor" aria-label="anchor" href="#mapped-mouse-brain-cell-types"></a>
</h3>













<ul class="nav nav-tabs" id="mapped-mouse-brain-cell-types" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l23-it" id="l23-it-tab" type="button" role="tab" aria-controls="l23-it" aria-selected="true" class="active nav-link">L2/3 IT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l4" id="l4-tab" type="button" role="tab" aria-controls="l4" aria-selected="false" class="nav-link">L4</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l5-it" id="l5-it-tab" type="button" role="tab" aria-controls="l5-it" aria-selected="false" class="nav-link">L5 IT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l5-pt" id="l5-pt-tab" type="button" role="tab" aria-controls="l5-pt" aria-selected="false" class="nav-link">L5 PT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l6-ct" id="l6-ct-tab" type="button" role="tab" aria-controls="l6-ct" aria-selected="false" class="nav-link">L6 CT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l6-it" id="l6-it-tab" type="button" role="tab" aria-controls="l6-it" aria-selected="false" class="nav-link">L6 IT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l6b" id="l6b-tab" type="button" role="tab" aria-controls="l6b" aria-selected="false" class="nav-link">L6b</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#oligo" id="oligo-tab" type="button" role="tab" aria-controls="oligo" aria-selected="false" class="nav-link">Oligo</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#pvalb" id="pvalb-tab" type="button" role="tab" aria-controls="pvalb" aria-selected="false" class="nav-link">Pvalb</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#meis2" id="meis2-tab" type="button" role="tab" aria-controls="meis2" aria-selected="false" class="nav-link">Meis2</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#astro" id="astro-tab" type="button" role="tab" aria-controls="astro" aria-selected="false" class="nav-link">Astro</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#vlmc" id="vlmc-tab" type="button" role="tab" aria-controls="vlmc" aria-selected="false" class="nav-link">VLMC</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#smc" id="smc-tab" type="button" role="tab" aria-controls="smc" aria-selected="false" class="nav-link">SMC</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="l23-it" role="tabpanel" aria-labelledby="l23-it-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
</div>
<div class="tab-pane" id="l4" role="tabpanel" aria-labelledby="l4-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-2.png" width="700"></p>
</div>
<div class="tab-pane" id="l5-it" role="tabpanel" aria-labelledby="l5-it-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-3.png" width="700"></p>
</div>
<div class="tab-pane" id="l5-pt" role="tabpanel" aria-labelledby="l5-pt-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-4.png" width="700"></p>
</div>
<div class="tab-pane" id="l6-ct" role="tabpanel" aria-labelledby="l6-ct-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-5.png" width="700"></p>
</div>
<div class="tab-pane" id="l6-it" role="tabpanel" aria-labelledby="l6-it-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-6.png" width="700"></p>
</div>
<div class="tab-pane" id="l6b" role="tabpanel" aria-labelledby="l6b-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-7.png" width="700"></p>
</div>
<div class="tab-pane" id="oligo" role="tabpanel" aria-labelledby="oligo-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-8.png" width="700"></p>
</div>
<div class="tab-pane" id="pvalb" role="tabpanel" aria-labelledby="pvalb-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-9.png" width="700"></p>
</div>
<div class="tab-pane" id="meis2" role="tabpanel" aria-labelledby="meis2-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-10.png" width="700"></p>
</div>
<div class="tab-pane" id="astro" role="tabpanel" aria-labelledby="astro-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-11.png" width="700"></p>
</div>
<div class="tab-pane" id="vlmc" role="tabpanel" aria-labelledby="vlmc-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-12.png" width="700"></p>
</div>
<div class="tab-pane" id="smc" role="tabpanel" aria-labelledby="smc-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-13.png" width="700"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="visualize-multiple-cell-types">Visualize multiple cell types<a class="anchor" aria-label="anchor" href="#visualize-multiple-cell-types"></a>
</h3>
<p>We can also visualize some of these cell types in one single plot
with <code><a href="../reference/visualize-multiple-features.html">MapMultipleFeatures()</a></code>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load H&amp;E images</span></span>
<span><span class="va">se_brain_spatial</span> <span class="op">&lt;-</span> <span class="va">se_brain_spatial</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot multiple features</span></span>
<span><span class="fu"><a href="../reference/visualize-multiple-features.html">MapMultipleFeatures</a></span><span class="op">(</span><span class="va">se_brain_spatial</span>, </span>
<span>                    image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                    pt_size <span class="op">=</span> <span class="fl">2</span>, max_cutoff <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span>                    override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                    colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#332288"</span>, <span class="st">"#88CCEE"</span>, <span class="st">"#44AA99"</span>, <span class="st">"#117733"</span>, <span class="st">"#DDCC77"</span>, <span class="st">"#CC6677"</span>,<span class="st">"#AA4499"</span><span class="op">)</span>,</span>
<span>                    features <span class="op">=</span> <span class="va">selected_celltypes</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-15-1.png" width="960"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="cell-type-co-localization">Cell type co-localization<a class="anchor" aria-label="anchor" href="#cell-type-co-localization"></a>
</h3>
<p>By computing the pair-wise correlation between cell types across
spots, we can get an idea of which cell types often appear together in
the same spots.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html" class="external-link">FetchData</a></span><span class="op">(</span><span class="va">se_brain_spatial</span>, <span class="va">selected_celltypes</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_all</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">.x</span><span class="op">&lt;</span><span class="fl">0.1</span>, <span class="fl">0</span>, <span class="va">.x</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span>  <span class="co"># Filter lowest values (-&gt; set as 0)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">cor_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NA</span></span>
<span><span class="va">max_val</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">cor_matrix</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">7</span>, <span class="st">"RdYlBu"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span>; <span class="va">cols</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"white"</span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">cor_matrix</span>, </span>
<span>                   breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="va">max_val</span>, <span class="va">max_val</span>, length.out <span class="op">=</span> <span class="fl">100</span><span class="op">)</span>,</span>
<span>                   color<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>,</span>
<span>                   cellwidth <span class="op">=</span> <span class="fl">14</span>, cellheight <span class="op">=</span> <span class="fl">14</span>, </span>
<span>                   treeheight_col <span class="op">=</span> <span class="fl">10</span>, treeheight_row <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                   main <span class="op">=</span> <span class="st">"Cell type correlation\nwithin spots"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-16-1.png" width="384"></p>
<p><br></p>
<p>We can also use the predicted cell type proportions to compute
factors using NMF, and thereby summarizing the cell types presence
within the same location into a set of predefined factors. Each factor
can be viewed as a niche of a certain cell type composition.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nmf_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html" class="external-link">FetchData</a></span><span class="op">(</span><span class="va">se_brain_spatial</span>, <span class="va">selected_celltypes</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">RcppML</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RcppML/man/nmf.html" class="external-link">nmf</a></span><span class="op">(</span>k <span class="op">=</span> <span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nmf_data_h</span> <span class="op">&lt;-</span> <span class="va">nmf_data</span><span class="op">$</span><span class="va">h</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">nmf_data_h</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Factor_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">nmf_data_h</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">selected_celltypes</span></span>
<span></span>
<span><span class="va">nmf_data_h</span> <span class="op">&lt;-</span> <span class="va">nmf_data_h</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html" class="external-link">mutate_at</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">nmf_data_h</span><span class="op">)</span>, </span>
<span>            <span class="op">~</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/scale.html" class="external-link">scale</a></span><span class="op">(</span><span class="va">.</span>, center <span class="op">=</span> <span class="cn">FALSE</span>, scale <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">.</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">/</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nmf_data_h</span><span class="op">$</span><span class="va">Factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">nmf_data_h</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span>levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Factor_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">nmf_data_h_df</span> <span class="op">&lt;-</span> <span class="va">nmf_data_h</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="va">selected_celltypes</span><span class="op">)</span>, </span>
<span>                      names_to <span class="op">=</span> <span class="st">"Cell"</span>, </span>
<span>                      values_to <span class="op">=</span> <span class="st">"Weight"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">nmf_data_h_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Factor</span>, y<span class="op">=</span><span class="va">Cell</span>, size<span class="op">=</span><span class="va">Weight</span>, color<span class="op">=</span><span class="va">Weight</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title<span class="op">=</span><span class="st">"Cell type contribution"</span>, x<span class="op">=</span><span class="st">"Factor"</span>, y <span class="op">=</span> <span class="st">"Cell type"</span>, </span>
<span>       color <span class="op">=</span> <span class="st">""</span>, size <span class="op">=</span> <span class="st">"Scaled weight"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_color_viridis_c</a></span><span class="op">(</span>direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, option <span class="op">=</span> <span class="st">"magma"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle<span class="op">=</span><span class="fl">45</span>, hjust<span class="op">=</span><span class="fl">1</span><span class="op">)</span>, </span>
<span>        panel.grid <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-17-1.png" width="384"></p>
</div>
</div>
<div class="section level2">
<h2 id="mkidney">Mouse kidney<a class="anchor" aria-label="anchor" href="#mkidney"></a>
</h2>
<p>In the second example, we’ll look at data from mouse kidney. We can
obtain the single-cell data with the <code>TabulaMurisSenisData</code> R
package from bioconductor. Let’s load the data and create a
<code>Seurat</code> object from it.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TabulaMurisSenisData/man/TabulaMurisSenisDroplet.html" class="external-link">TabulaMurisSenisDroplet</a></span><span class="op">(</span>tissues <span class="op">=</span> <span class="st">"Kidney"</span><span class="op">)</span><span class="op">$</span><span class="va">Kidney</span></span>
<span><span class="va">umis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="st">"dgCMatrix"</span><span class="op">)</span></span>
<span><span class="va">se_kidney_singlecell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">umis</span>, meta.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The 10x Visium mouse kidney data can be downloaded from the 10x
genomics website.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">tmpdir</span>, <span class="st">"/kidney"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">targetdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">tmpdir</span>, <span class="st">"/kidney"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Download section 1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Kidney/V1_Mouse_Kidney_filtered_feature_bc_matrix.h5"</span>, </span>
<span>              destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/filtered_feature_bc_matrix.h5"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Kidney/V1_Mouse_Kidney_spatial.tar.gz"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/spatial.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span>tarfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/spatial.tar.gz"</span><span class="op">)</span>, </span>
<span>      exdir <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">targetdir</span>, <span class="st">"/visium/spatial.tar.gz"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="st">"./kidney/visium/filtered_feature_bc_matrix.h5"</span></span>
<span><span class="va">imgs</span> <span class="op">&lt;-</span> <span class="st">"./kidney/visium/spatial/tissue_hires_image.png"</span></span>
<span><span class="va">spotfiles</span> <span class="op">&lt;-</span> <span class="st">"./kidney/visium/spatial/tissue_positions_list.csv"</span></span>
<span><span class="va">json</span> <span class="op">&lt;-</span> <span class="st">"./kidney/visium/spatial/scalefactors_json.json"</span></span>
<span></span>
<span><span class="va">infoTable</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">imgs</span>, <span class="va">spotfiles</span>, <span class="va">json</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_kidney_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReadVisiumData.html">ReadVisiumData</a></span><span class="op">(</span><span class="va">infoTable</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="normalize-data-1">Normalize data<a class="anchor" aria-label="anchor" href="#normalize-data-1"></a>
</h3>
<p>We apply the same normalization procedure to
<code>se_kidney_spatial</code> and <code>se_kidney_singlecell</code> and
run <code><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures()</a></code> to detect the top most variable
genes.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_kidney_spatial</span> <span class="op">&lt;-</span> <span class="va">se_kidney_spatial</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
<p>For the single-cell kidney data, we’ll also filter the data prior to
normalization to include cells collected at age 18m and remove cells
with labels “nan” and “CD45”. This leaves us with 17 cell types.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se_kidney_singlecell</span><span class="op">)</span><span class="op">[</span><span class="va">se_kidney_singlecell</span><span class="op">$</span><span class="va">age</span> <span class="op">==</span> <span class="st">"18m"</span> <span class="op">&amp;</span> <span class="op">(</span><span class="op">!</span><span class="va">se_kidney_singlecell</span><span class="op">$</span><span class="va">free_annotation</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nan"</span>, <span class="st">"CD45"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">se_kidney_singlecell</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">se_kidney_singlecell</span>, cells <span class="op">=</span> <span class="va">keep_cells</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_kidney_singlecell</span> <span class="op">&lt;-</span> <span class="va">se_kidney_singlecell</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_kidney_singlecell</span> <span class="op">&lt;-</span> <span class="va">se_kidney_singlecell</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-nnls-1">Run NNLS<a class="anchor" aria-label="anchor" href="#run-nnls-1"></a>
</h3>
<p>Again, the <code><a href="../reference/celltype-prediction.html">RunNNLS()</a></code> method requires a single-cell
<code>Seurat</code> object and a 10x Visium <code>Seurat</code> object.
The cell type annotations are stored in the “free_annotation”
column.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ti</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">se_kidney_spatial</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"Spatial"</span></span>
<span><span class="va">se_kidney_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/celltype-prediction.html">RunNNLS</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">se_kidney_spatial</span>, </span>
<span>                      singlecell_object <span class="op">=</span> <span class="va">se_kidney_singlecell</span>, </span>
<span>                      groups <span class="op">=</span> <span class="st">"free_annotation"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## ── <span style="font-weight: bold;">Predicting cell type proportions</span> ──</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Fetching data from Seurat objects</span></span></code></pre>
<pre><code><span><span class="co">## →   Filtering out features that are only present in one data set</span></span></code></pre>
<pre><code><span><span class="co">## →   Kept 4993 features for deconvolution</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Preparing data for NNLS</span></span></code></pre>
<pre><code><span><span class="co">## →   Downsampling scRNA-seq data to include a maximum of 50 cells per cell type</span></span></code></pre>
<pre><code><span><span class="co">## → <span style="color: #FF55FF;">  Cell type(s) CD45    NK cell,CD45    plasma cell have too few cells (&lt;10) and will be excluded</span></span></span></code></pre>
<pre><code><span><span class="co">## →   Kept 15 cell types after filtering</span></span></code></pre>
<pre><code><span><span class="co">## →   Calculating cell type expression profiles</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Predicting cell type proportions with NNLS for 15 cell types</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Returning results in a new 'Assay' named 'celltypeprops'</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Setting default assay to 'celltypeprops'</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Finished</span></span></code></pre>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"RunNNLS finished in %s seconds"</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="va">ti</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "RunNNLS finished in 0.55 seconds"</span></span></code></pre>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check available cell types</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">se_kidney_spatial</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "CD45    B cell"                                            </span></span>
<span><span class="co">##  [2] "CD45    T cell"                                            </span></span>
<span><span class="co">##  [3] "CD45    macrophage"                                        </span></span>
<span><span class="co">##  [4] "Epcam     kidney distal convoluted tubule epithelial cell" </span></span>
<span><span class="co">##  [5] "Epcam    brush cell"                                       </span></span>
<span><span class="co">##  [6] "Epcam    kidney collecting duct principal cell"            </span></span>
<span><span class="co">##  [7] "Epcam    kidney proximal convoluted tubule epithelial cell"</span></span>
<span><span class="co">##  [8] "Epcam    podocyte"                                         </span></span>
<span><span class="co">##  [9] "Epcam    proximal tube epithelial cell"                    </span></span>
<span><span class="co">## [10] "Epcam    thick ascending tube S epithelial cell"           </span></span>
<span><span class="co">## [11] "Pecam    Kidney cortex artery cell"                        </span></span>
<span><span class="co">## [12] "Pecam    fenestrated capillary endothelial"                </span></span>
<span><span class="co">## [13] "Pecam    kidney capillary endothelial cell"                </span></span>
<span><span class="co">## [14] "Stroma    fibroblast"                                      </span></span>
<span><span class="co">## [15] "Stroma    kidney mesangial cell"</span></span></code></pre>
<p>NB: two cell types were discarded as they didn’t pass the minimum
allowed cells per cell type threshold. The cell types discarded were NK
cells and plasma cells.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot selected cell types</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">se_kidney_spatial</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltypeprops"</span></span>
<span></span>
<span><span class="va">selected_celltypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Epcam     kidney distal convoluted tubule epithelial cell"</span>,</span>
<span>                        <span class="st">"Epcam    brush cell"</span>,</span>
<span>                        <span class="st">"Epcam    kidney collecting duct principal cell"</span>,</span>
<span>                        <span class="st">"Epcam    kidney proximal convoluted tubule epithelial cell"</span>,</span>
<span>                        <span class="st">"Epcam    podocyte"</span>,</span>
<span>                        <span class="st">"Epcam    proximal tube epithelial cell"</span>,</span>
<span>                        <span class="st">"Epcam    thick ascending tube S epithelial cell"</span>,</span>
<span>                        <span class="st">"Pecam    fenestrated capillary endothelial"</span>,</span>
<span>                        <span class="st">"Pecam    kidney capillary endothelial cell"</span>,</span>
<span>                        <span class="st">"Stroma    kidney mesangial cell"</span><span class="op">)</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">selected_celltypes</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se_kidney_spatial</span>, pt_size <span class="op">=</span> <span class="fl">1.3</span>,</span>
<span>            features <span class="op">=</span> <span class="va">selected_celltypes</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>, </span>
<span>            override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span>nm <span class="op">=</span> <span class="va">selected_celltypes</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3 tabset">
<h3 class="tabset" id="mapped-kidney-cell-types">Mapped kidney cell types<a class="anchor" aria-label="anchor" href="#mapped-kidney-cell-types"></a>
</h3>










<ul class="nav nav-tabs" id="mapped-kidney-cell-types" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#epcam-kidney-distal-convoluted-tubule-epithelial-cell" id="epcam-kidney-distal-convoluted-tubule-epithelial-cell-tab" type="button" role="tab" aria-controls="epcam-kidney-distal-convoluted-tubule-epithelial-cell" aria-selected="true" class="active nav-link">Epcam kidney distal convoluted tubule epithelial cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#epcam-brush-cell" id="epcam-brush-cell-tab" type="button" role="tab" aria-controls="epcam-brush-cell" aria-selected="false" class="nav-link">Epcam brush cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#epcam-kidney-collecting-duct-principal-cell" id="epcam-kidney-collecting-duct-principal-cell-tab" type="button" role="tab" aria-controls="epcam-kidney-collecting-duct-principal-cell" aria-selected="false" class="nav-link">Epcam kidney collecting duct principal cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#epcam-kidney-proximal-convoluted-tubule-epithelial-cell" id="epcam-kidney-proximal-convoluted-tubule-epithelial-cell-tab" type="button" role="tab" aria-controls="epcam-kidney-proximal-convoluted-tubule-epithelial-cell" aria-selected="false" class="nav-link">Epcam kidney proximal convoluted tubule epithelial cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#epcam-podocyte" id="epcam-podocyte-tab" type="button" role="tab" aria-controls="epcam-podocyte" aria-selected="false" class="nav-link">Epcam podocyte</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#epcam-proximal-tube-epithelial-cell" id="epcam-proximal-tube-epithelial-cell-tab" type="button" role="tab" aria-controls="epcam-proximal-tube-epithelial-cell" aria-selected="false" class="nav-link">Epcam proximal tube epithelial cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#epcam-thick-ascending-tube-s-epithelial-cell" id="epcam-thick-ascending-tube-s-epithelial-cell-tab" type="button" role="tab" aria-controls="epcam-thick-ascending-tube-s-epithelial-cell" aria-selected="false" class="nav-link">Epcam thick ascending tube S epithelial cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#pecam-fenestrated-capillary-endothelial" id="pecam-fenestrated-capillary-endothelial-tab" type="button" role="tab" aria-controls="pecam-fenestrated-capillary-endothelial" aria-selected="false" class="nav-link">Pecam fenestrated capillary endothelial</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#pecam-kidney-capillary-endothelial-cell" id="pecam-kidney-capillary-endothelial-cell-tab" type="button" role="tab" aria-controls="pecam-kidney-capillary-endothelial-cell" aria-selected="false" class="nav-link">Pecam kidney capillary endothelial cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#stroma-kidney-mesangial-cell" id="stroma-kidney-mesangial-cell-tab" type="button" role="tab" aria-controls="stroma-kidney-mesangial-cell" aria-selected="false" class="nav-link">Stroma kidney mesangial cell</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="epcam-kidney-distal-convoluted-tubule-epithelial-cell" role="tabpanel" aria-labelledby="epcam-kidney-distal-convoluted-tubule-epithelial-cell-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
</div>
<div class="tab-pane" id="epcam-brush-cell" role="tabpanel" aria-labelledby="epcam-brush-cell-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-2.png" width="700"></p>
</div>
<div class="tab-pane" id="epcam-kidney-collecting-duct-principal-cell" role="tabpanel" aria-labelledby="epcam-kidney-collecting-duct-principal-cell-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-3.png" width="700"></p>
</div>
<div class="tab-pane" id="epcam-kidney-proximal-convoluted-tubule-epithelial-cell" role="tabpanel" aria-labelledby="epcam-kidney-proximal-convoluted-tubule-epithelial-cell-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-4.png" width="700"></p>
</div>
<div class="tab-pane" id="epcam-podocyte" role="tabpanel" aria-labelledby="epcam-podocyte-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-5.png" width="700"></p>
</div>
<div class="tab-pane" id="epcam-proximal-tube-epithelial-cell" role="tabpanel" aria-labelledby="epcam-proximal-tube-epithelial-cell-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-6.png" width="700"></p>
</div>
<div class="tab-pane" id="epcam-thick-ascending-tube-s-epithelial-cell" role="tabpanel" aria-labelledby="epcam-thick-ascending-tube-s-epithelial-cell-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-7.png" width="700"></p>
</div>
<div class="tab-pane" id="pecam-fenestrated-capillary-endothelial" role="tabpanel" aria-labelledby="pecam-fenestrated-capillary-endothelial-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-8.png" width="700"></p>
</div>
<div class="tab-pane" id="pecam-kidney-capillary-endothelial-cell" role="tabpanel" aria-labelledby="pecam-kidney-capillary-endothelial-cell-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-9.png" width="700"></p>
</div>
<div class="tab-pane" id="stroma-kidney-mesangial-cell" role="tabpanel" aria-labelledby="stroma-kidney-mesangial-cell-tab">

<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-10.png" width="700"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="visualize-multiple-cell-types-1">Visualize multiple cell types<a class="anchor" aria-label="anchor" href="#visualize-multiple-cell-types-1"></a>
</h3>
<p>We can select some of these cell types and visualize them in one
plot</p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load H&amp;E images</span></span>
<span><span class="va">se_kidney_spatial</span> <span class="op">&lt;-</span> <span class="va">se_kidney_spatial</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot multiple features</span></span>
<span><span class="fu"><a href="../reference/visualize-multiple-features.html">MapMultipleFeatures</a></span><span class="op">(</span><span class="va">se_kidney_spatial</span>, </span>
<span>                    image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                    pt_size <span class="op">=</span> <span class="fl">2</span>, max_cutoff <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>                    override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                    colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#332288"</span>, <span class="st">"#88CCEE"</span>, <span class="st">"#117733"</span>, <span class="st">"#DDCC77"</span>, <span class="st">"#CC6677"</span>,<span class="st">"#AA4499"</span><span class="op">)</span>,</span>
<span>                    features <span class="op">=</span> <span class="va">selected_celltypes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">6</span>, <span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-28-1.png" width="960"></p>
<p>Or zoom in on a region of interest</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Reload H&amp;E image in higher resolution</span></span>
<span><span class="va">se_kidney_spatial</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="va">se_kidney_spatial</span>, image_height <span class="op">=</span> <span class="fl">1500</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot H&amp;E image</span></span>
<span><span class="va">rst</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetImages.html">GetImages</a></span><span class="op">(</span><span class="va">se_kidney_spatial</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"H&amp;E image of zoomed in area"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/inset_element.html" class="external-link">inset_element</a></span><span class="op">(</span>p <span class="op">=</span> <span class="va">rst</span><span class="op">[</span><span class="op">(</span><span class="fl">0.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">rst</span><span class="op">)</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="fl">0.7</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">rst</span><span class="op">)</span><span class="op">)</span>, <span class="op">(</span><span class="fl">0.5</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">rst</span><span class="op">)</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="fl">0.7</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">rst</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                left <span class="op">=</span> <span class="fl">0</span>, bottom <span class="op">=</span> <span class="fl">0</span>, right <span class="op">=</span> <span class="fl">1</span>, top <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot multiple features with zoom</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-multiple-features.html">MapMultipleFeatures</a></span><span class="op">(</span><span class="va">se_kidney_spatial</span>, </span>
<span>                    image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                    pt_size <span class="op">=</span> <span class="fl">4.5</span>, max_cutoff <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>                    crop_area <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.7</span>, <span class="fl">0.7</span><span class="op">)</span>,</span>
<span>                    colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#332288"</span>, <span class="st">"#88CCEE"</span>, <span class="st">"#117733"</span>, <span class="st">"#DDCC77"</span>, <span class="st">"#CC6677"</span>,<span class="st">"#AA4499"</span><span class="op">)</span>,</span>
<span>                    features <span class="op">=</span> <span class="va">selected_celltypes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">6</span>, <span class="fl">7</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">/</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-29-1.png" width="960"></p>
<p><br></p>
<hr>
<details open><summary><strong>Package versions</strong>
</summary><ul>
<li><p><code>semla</code>: 0.1.0</p></li>
<li><p><code>RcppML</code>: 0.3.7</p></li>
</ul></details><details><summary><strong>Session info</strong>
</summary><div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin13.4.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /Users/ludviglarsson/miniconda3/envs/R4.2/lib/libopenblasp-r0.3.21.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] patchwork_1.1.2             purrr_1.0.1                </span></span>
<span><span class="co">##  [3] SingleCellExperiment_1.18.0 SummarizedExperiment_1.26.1</span></span>
<span><span class="co">##  [5] Biobase_2.56.0              GenomicRanges_1.48.0       </span></span>
<span><span class="co">##  [7] GenomeInfoDb_1.32.4         IRanges_2.30.1             </span></span>
<span><span class="co">##  [9] S4Vectors_0.34.0            BiocGenerics_0.42.0        </span></span>
<span><span class="co">## [11] MatrixGenerics_1.8.1        matrixStats_0.62.0         </span></span>
<span><span class="co">## [13] TabulaMurisSenisData_1.2.0  tibble_3.1.8               </span></span>
<span><span class="co">## [15] semla_0.1.0                 ggplot2_3.4.1              </span></span>
<span><span class="co">## [17] dplyr_1.1.0                 SeuratObject_4.1.3         </span></span>
<span><span class="co">## [19] Seurat_4.3.0               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] utf8_1.2.2                    spatstat.explore_3.0-5       </span></span>
<span><span class="co">##   [3] reticulate_1.26               tidyselect_1.2.0             </span></span>
<span><span class="co">##   [5] RSQLite_2.2.17                AnnotationDbi_1.58.0         </span></span>
<span><span class="co">##   [7] htmlwidgets_1.5.4             grid_4.2.1                   </span></span>
<span><span class="co">##   [9] Rtsne_0.16                    munsell_0.5.0                </span></span>
<span><span class="co">##  [11] codetools_0.2-18              ragg_1.2.2                   </span></span>
<span><span class="co">##  [13] ica_1.0-3                     future_1.27.0                </span></span>
<span><span class="co">##  [15] miniUI_0.1.1.1                withr_2.5.0                  </span></span>
<span><span class="co">##  [17] spatstat.random_3.0-1         colorspace_2.0-3             </span></span>
<span><span class="co">##  [19] progressr_0.10.1              filelock_1.0.2               </span></span>
<span><span class="co">##  [21] highr_0.9                     knitr_1.39                   </span></span>
<span><span class="co">##  [23] rstudioapi_0.14               ROCR_1.0-11                  </span></span>
<span><span class="co">##  [25] tensor_1.5                    listenv_0.8.0                </span></span>
<span><span class="co">##  [27] labeling_0.4.2                GenomeInfoDbData_1.2.8       </span></span>
<span><span class="co">##  [29] RcppML_0.3.7                  polyclip_1.10-0              </span></span>
<span><span class="co">##  [31] pheatmap_1.0.12               farver_2.1.1                 </span></span>
<span><span class="co">##  [33] bit64_4.0.5                   rhdf5_2.40.0                 </span></span>
<span><span class="co">##  [35] rprojroot_2.0.3               parallelly_1.32.1            </span></span>
<span><span class="co">##  [37] vctrs_0.5.2                   generics_0.1.3               </span></span>
<span><span class="co">##  [39] xfun_0.32                     BiocFileCache_2.4.0          </span></span>
<span><span class="co">##  [41] R6_2.5.1                      DelayedArray_0.22.0          </span></span>
<span><span class="co">##  [43] bitops_1.0-7                  rhdf5filters_1.8.0           </span></span>
<span><span class="co">##  [45] spatstat.utils_3.0-1          cachem_1.0.6                 </span></span>
<span><span class="co">##  [47] assertthat_0.2.1              promises_1.2.0.1             </span></span>
<span><span class="co">##  [49] scales_1.2.1                  gtable_0.3.0                 </span></span>
<span><span class="co">##  [51] globals_0.16.0                goftest_1.2-3                </span></span>
<span><span class="co">##  [53] rlang_1.0.6                   zeallot_0.1.0                </span></span>
<span><span class="co">##  [55] systemfonts_1.0.4             splines_4.2.1                </span></span>
<span><span class="co">##  [57] lazyeval_0.2.2                spatstat.geom_3.0-3          </span></span>
<span><span class="co">##  [59] BiocManager_1.30.18           yaml_2.3.5                   </span></span>
<span><span class="co">##  [61] reshape2_1.4.4                abind_1.4-5                  </span></span>
<span><span class="co">##  [63] httpuv_1.6.5                  tools_4.2.1                  </span></span>
<span><span class="co">##  [65] ellipsis_0.3.2                jquerylib_0.1.4              </span></span>
<span><span class="co">##  [67] RColorBrewer_1.1-3            ggridges_0.5.3               </span></span>
<span><span class="co">##  [69] Rcpp_1.0.9                    plyr_1.8.7                   </span></span>
<span><span class="co">##  [71] zlibbioc_1.42.0               RCurl_1.98-1.8               </span></span>
<span><span class="co">##  [73] dbscan_1.1-10                 deldir_1.0-6                 </span></span>
<span><span class="co">##  [75] pbapply_1.5-0                 cowplot_1.1.1                </span></span>
<span><span class="co">##  [77] zoo_1.8-10                    ggrepel_0.9.3                </span></span>
<span><span class="co">##  [79] cluster_2.1.4                 fs_1.5.2                     </span></span>
<span><span class="co">##  [81] magrittr_2.0.3                data.table_1.14.2            </span></span>
<span><span class="co">##  [83] magick_2.7.3                  scattermore_0.8              </span></span>
<span><span class="co">##  [85] lmtest_0.9-40                 RANN_2.6.1                   </span></span>
<span><span class="co">##  [87] ggnewscale_0.4.8              fitdistrplus_1.1-8           </span></span>
<span><span class="co">##  [89] shinyjs_2.1.0                 mime_0.12                    </span></span>
<span><span class="co">##  [91] evaluate_0.16                 xtable_1.8-4                 </span></span>
<span><span class="co">##  [93] gridExtra_2.3                 compiler_4.2.1               </span></span>
<span><span class="co">##  [95] KernSmooth_2.23-20            crayon_1.5.1                 </span></span>
<span><span class="co">##  [97] htmltools_0.5.4               later_1.3.0                  </span></span>
<span><span class="co">##  [99] tidyr_1.3.0                   DBI_1.1.3                    </span></span>
<span><span class="co">## [101] ExperimentHub_2.4.0           dbplyr_2.2.1                 </span></span>
<span><span class="co">## [103] MASS_7.3-58.1                 rappdirs_0.3.3               </span></span>
<span><span class="co">## [105] Matrix_1.5-3                  cli_3.4.1                    </span></span>
<span><span class="co">## [107] gdata_2.18.0.1                parallel_4.2.1               </span></span>
<span><span class="co">## [109] igraph_1.3.4                  forcats_0.5.2                </span></span>
<span><span class="co">## [111] pkgconfig_2.0.3               pkgdown_2.0.6                </span></span>
<span><span class="co">## [113] sp_1.5-1                      plotly_4.10.0                </span></span>
<span><span class="co">## [115] spatstat.sparse_3.0-0         bslib_0.4.0                  </span></span>
<span><span class="co">## [117] XVector_0.36.0                stringr_1.5.0                </span></span>
<span><span class="co">## [119] digest_0.6.29                 sctransform_0.3.5            </span></span>
<span><span class="co">## [121] RcppAnnoy_0.0.19              spatstat.data_3.0-0          </span></span>
<span><span class="co">## [123] Biostrings_2.64.1             rmarkdown_2.15               </span></span>
<span><span class="co">## [125] leiden_0.4.2                  uwot_0.1.14                  </span></span>
<span><span class="co">## [127] curl_4.3.2                    gtools_3.9.3                 </span></span>
<span><span class="co">## [129] shiny_1.7.4                   lifecycle_1.0.3              </span></span>
<span><span class="co">## [131] nlme_3.1-159                  jsonlite_1.8.3               </span></span>
<span><span class="co">## [133] Rhdf5lib_1.18.2               desc_1.4.1                   </span></span>
<span><span class="co">## [135] viridisLite_0.4.1             fansi_1.0.3                  </span></span>
<span><span class="co">## [137] pillar_1.8.1                  lattice_0.20-45              </span></span>
<span><span class="co">## [139] KEGGREST_1.36.3               fastmap_1.1.0                </span></span>
<span><span class="co">## [141] httr_1.4.4                    survival_3.4-0               </span></span>
<span><span class="co">## [143] interactiveDisplayBase_1.34.0 glue_1.6.2                   </span></span>
<span><span class="co">## [145] png_0.1-7                     BiocVersion_3.15.2           </span></span>
<span><span class="co">## [147] bit_4.0.4                     stringi_1.7.8                </span></span>
<span><span class="co">## [149] sass_0.4.2                    HDF5Array_1.24.2             </span></span>
<span><span class="co">## [151] blob_1.2.3                    textshaping_0.3.6            </span></span>
<span><span class="co">## [153] AnnotationHub_3.4.0           memoise_2.0.1                </span></span>
<span><span class="co">## [155] irlba_2.3.5                   future.apply_1.9.0</span></span></code></pre>
</details><p><br></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
