<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Comparison of cell type mapping with NNLS • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Comparison of cell type mapping with NNLS">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/semla.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><h6 class="dropdown-header" data-toc-skip>Intro</h6></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a></li>
    <li><a class="dropdown-item" href="../articles/visiumHD.html">VisiumHD</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a></li>
    <li><a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a></li>
    <li><a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a></li>
    <li><a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial methods</h6></li>
    <li><a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a></li>
    <li><a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a></li>
    <li><a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a></li>
    <li><a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Image processing</h6></li>
    <li><a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a></li>
    <li><a class="dropdown-item" href="../articles/mask_images.html">Image masking</a></li>
    <li><a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a></li>
  </ul>
</li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/create_object.html">Create a Staffli object</a></li>
    <li><a class="dropdown-item" href="../articles/compare_cell_type_mapping_NNLS.html">Cell type mapping comparison</a></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_benchmark.html">Cell type mapping benchmark</a></li>
    <li><a class="dropdown-item" href="../articles/slide-seq.html">Slide-Seq data</a></li>
    <li><a class="dropdown-item" href="../articles/IF_data.html">IF data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Comparison of cell type mapping with NNLS</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 08 September
2025</h4>
      

      <div class="d-none name"><code>compare_cell_type_mapping_NNLS.Rmd</code></div>
    </div>

    
    
<p>As described in the tutorial <a href="https://ludvigla.github.io/semla/articles/cell_type_mapping_with_NNLS.html">Cell
type mapping with NNLS</a>, there is an implementation of a cell type
mapping approach in <code>semla</code> based on the Non-Negative Least
Squares (NNLS) method, as implemented in the <a href="https://github.com/zdebruine/RcppML" class="external-link">RcppML</a> R package. Cell
type mapping algorithms utilizes a well annotated scRNA-seq data set to
learn cell type expression profiles in order to infer their relative
abundance or proportions onto the spatial data. In short, the approach
implemented in <code>semla</code> works by initially estimating
enrichment scores for each available cell type in the scRNA-seq data and
assigning weights to genes based on their cell type specificity. This
information is in the following steps leveraged in the NNLS methods to
solve the following optimization problem in order to assign cell type
composition predictions to each spatial coordinate in the Visium
data,</p>
<math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mi mathvariant="normal">a</mi><mi mathvariant="normal">r</mi><mi mathvariant="normal">g</mi><mspace width="0.222em"></mspace><mi mathvariant="normal">m</mi><mi mathvariant="normal">i</mi><msub><mi mathvariant="normal">n</mi><mi mathvariant="normal">x</mi></msub></mrow><mrow><mo stretchy="true" form="prefix">|</mo><mo stretchy="false" form="prefix">|</mo><mi>A</mi><mi>x</mi><mo>−</mo><mi>y</mi><mo stretchy="true" form="postfix">|</mo></mrow><msubsup><mo stretchy="false" form="prefix">|</mo><mn>2</mn><mn>2</mn></msubsup><mrow><mtext mathvariant="normal">, subject to </mtext><mspace width="0.333em"></mspace></mrow><mi>x</mi><mo>≥</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\begin{equation}
\operatorname*{arg\:min_{x}} \left||Ax-y\right||_2^2 \text{, subject to } x \geq 0
\end{equation}</annotation></semantics></math><p>where <em>A</em> is a matrix with cell type enrichment scores,
<em>y</em> is a vector with mixed a mixed gene expression profile. The
solution for x represents estimates of the fractional abundances of cell
type estimates.</p>
<p><br></p>
<p>To demonstrate the utility of this approach, we have performed a
basic comparison with two other popular cell type mapping approaches,
applied to two tissue types.</p>
<p><br></p>
<p><strong>Cell type mapping methods</strong></p>
<ul>
<li><p><em>NNLS</em>: Debruine, Zach &amp; Melcher, Karsten &amp;
Triche, Timothy. (2021). Fast and robust non-negative matrix
factorization for single-cell experiments. <a href="https://doi.org/10.1101/2021.09.01.458620" class="external-link uri">https://doi.org/10.1101/2021.09.01.458620</a></p></li>
<li><p><em>Stereoscope</em>: Andersson, A., Bergenstråhle, J., Asp, M.
<em>et al.</em> Single-cell and spatial transcriptomics enables
probabilistic inference of cell type topography. <em>Commun Biol</em>
<strong>3</strong>, 565 (2020). <a href="https://doi.org/10.1038/s42003-020-01247-y" class="external-link uri">https://doi.org/10.1038/s42003-020-01247-y</a></p></li>
<li><p><em>cell2location</em>: Kleshchevnikov, V., Shmatko, A., Dann, E.
<em>et al.</em> Cell2location maps fine-grained cell types in spatial
transcriptomics. <em>Nat Biotechnol</em> <strong>40</strong>, 661–671
(2022). <a href="https://doi.org/10.1038/s41587-021-01139-4" class="external-link uri">https://doi.org/10.1038/s41587-021-01139-4</a></p></li>
</ul>
<p><br></p>
<p><strong>Data sets</strong></p>
<ul>
<li>
<p>Brain sagittal section (mouse).</p>
<ul>
<li><p>scRNA-seq: Allen Brain, mouse atlas reference single cell RNA-seq
data set</p></li>
<li><p>Visium: Mouse brain sagittal data (anterial + posterior) made
available by <a href="https://www.10xgenomics.com/cn/resources/datasets/mouse-brain-serial-section-1-sagittal-anterior-1-standard-1-0-0" class="external-link">10x
Genomics</a>, processed using the Space Ranger pipeline v1.0.0</p></li>
</ul>
</li>
<li>
<p>Kidney (mouse)</p>
<ul>
<li><p>scRNA-seq: Tabula Muris Senis droplet data from kidney, made
available by <a href="https://doi.org/10.1038%2Fs41586-020-2496-1" class="external-link">The
Tabula Muris Consortium</a></p></li>
<li><p>Visium: Mouse kidney coronal section data made available by <a href="https://www.10xgenomics.com/cn/resources/datasets/mouse-kidney-section-coronal-1-standard-1-1-0" class="external-link">10x
Genomics</a>, processed using the Space Ranger pipeline v1.1.0</p></li>
</ul>
</li>
</ul>
<p>The same Visium and reference scRNA-seq data sets were used as input
for all cell type deconvolution methods tested. The code displayed in
the <a href="https://ludvigla.github.io/semla/articles/cell_type_mapping_with_NNLS.html">Cell
type mapping with NNLS</a> vignette was used to produce the NNLS mapping
results, while <em>stereoscope</em> and <em>cell2location</em> were run
in python on high-performance computing servers using default parameter
settings.</p>
<p><br></p>
<div class="section level2">
<h2 id="mouse-brain">Mouse brain<a class="anchor" aria-label="anchor" href="#mouse-brain"></a>
</h2>
<p>The results of each cell type mapping approach is stored as an
<em>Assay</em> within the <code>semla</code> object, and the spot-wise
Pearson correlation values are computed using the <code><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor()</a></code>
function in R. The results are visualized as heatmaps using
<code>ggplot2</code> and <code><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile()</a></code>, where the fill color
corresponds to the strength of the correlation.</p>
<div class="float">
<img src="compare_cell_type_mapping_NNLS/mbrain_cor_heatmaps.jpg" alt="1. Correlation plots of cell type mapping results in mouse brain dataset."><div class="figcaption">1. Correlation plots of cell type mapping
results in mouse brain dataset.</div>
</div>
<p><br></p>
<p>In general, we can see that the correlation between mapped cell types
values is high between the NNLS method and the other two tested
approaches. As we do not have any ground truth values, it is
unfortunately impossible for us to tell which of these methods produce
results closest to the true cell type proportions in each spot.</p>
<p>We can also summarize the results by looking at the correlation
between matching cell types, to better view if the NNLS method seems to
share a higher concordance with a specific method for certain cell
types.</p>
<div class="float">
<img src="compare_cell_type_mapping_NNLS/mbrain_cor_bars.jpg" alt="2. Comparison of correlation values against NNLS for cell types in mouse brain dataset."><div class="figcaption">2. Comparison of correlation values against NNLS
for cell types in mouse brain dataset.</div>
</div>
<p><br></p>
<p>Here, we can for instance more easily spot a large difference in the
correlation values between the NNLS:stereoscope an NNLS:cell2location
for Pvalb cells, demonstrating an example of how different cell type
mapping algorithms can perform drastically different despite the same
input.</p>
<p>Lastly, let us have a look at the cell type mapping results from all
three approaches spatially.</p>
<div class="section level3 tabset">
<h3 class="tabset" id="spatial-cell-type-mapping">Spatial cell type mapping<a class="anchor" aria-label="anchor" href="#spatial-cell-type-mapping"></a>
</h3>






















<ul class="nav nav-tabs" id="spatial-cell-type-mapping" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#astro" id="astro-tab" type="button" role="tab" aria-controls="astro" aria-selected="true" class="active nav-link">Astro</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#endo" id="endo-tab" type="button" role="tab" aria-controls="endo" aria-selected="false" class="nav-link">Endo</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l23-it" id="l23-it-tab" type="button" role="tab" aria-controls="l23-it" aria-selected="false" class="nav-link">L2/3 IT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l4" id="l4-tab" type="button" role="tab" aria-controls="l4" aria-selected="false" class="nav-link">L4</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l5-it" id="l5-it-tab" type="button" role="tab" aria-controls="l5-it" aria-selected="false" class="nav-link">L5 IT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l5-pt" id="l5-pt-tab" type="button" role="tab" aria-controls="l5-pt" aria-selected="false" class="nav-link">L5 PT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l6-ct" id="l6-ct-tab" type="button" role="tab" aria-controls="l6-ct" aria-selected="false" class="nav-link">L6 CT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l6-it" id="l6-it-tab" type="button" role="tab" aria-controls="l6-it" aria-selected="false" class="nav-link">L6 IT</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#l6b" id="l6b-tab" type="button" role="tab" aria-controls="l6b" aria-selected="false" class="nav-link">L6b</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#lamp5" id="lamp5-tab" type="button" role="tab" aria-controls="lamp5" aria-selected="false" class="nav-link">Lamp5</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#macrophage" id="macrophage-tab" type="button" role="tab" aria-controls="macrophage" aria-selected="false" class="nav-link">Macrophage</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#meis2" id="meis2-tab" type="button" role="tab" aria-controls="meis2" aria-selected="false" class="nav-link">Meis2</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#np" id="np-tab" type="button" role="tab" aria-controls="np" aria-selected="false" class="nav-link">NP</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#oligo" id="oligo-tab" type="button" role="tab" aria-controls="oligo" aria-selected="false" class="nav-link">Oligo</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#peri" id="peri-tab" type="button" role="tab" aria-controls="peri" aria-selected="false" class="nav-link">Peri</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#pvalb" id="pvalb-tab" type="button" role="tab" aria-controls="pvalb" aria-selected="false" class="nav-link">Pvalb</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#serpinf1" id="serpinf1-tab" type="button" role="tab" aria-controls="serpinf1" aria-selected="false" class="nav-link">Serpinf1</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#smc" id="smc-tab" type="button" role="tab" aria-controls="smc" aria-selected="false" class="nav-link">SMC</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#sncg" id="sncg-tab" type="button" role="tab" aria-controls="sncg" aria-selected="false" class="nav-link">Sncg</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#sst" id="sst-tab" type="button" role="tab" aria-controls="sst" aria-selected="false" class="nav-link">Sst</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#vip" id="vip-tab" type="button" role="tab" aria-controls="vip" aria-selected="false" class="nav-link">Vip</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#vlmc" id="vlmc-tab" type="button" role="tab" aria-controls="vlmc" aria-selected="false" class="nav-link">VLMC</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="astro" role="tabpanel" aria-labelledby="astro-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-1.png" width="1344"></p>
</div>
<div class="tab-pane" id="endo" role="tabpanel" aria-labelledby="endo-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-2.png" width="1344"></p>
</div>
<div class="tab-pane" id="l23-it" role="tabpanel" aria-labelledby="l23-it-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-3.png" width="1344"></p>
</div>
<div class="tab-pane" id="l4" role="tabpanel" aria-labelledby="l4-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-4.png" width="1344"></p>
</div>
<div class="tab-pane" id="l5-it" role="tabpanel" aria-labelledby="l5-it-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-5.png" width="1344"></p>
</div>
<div class="tab-pane" id="l5-pt" role="tabpanel" aria-labelledby="l5-pt-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-6.png" width="1344"></p>
</div>
<div class="tab-pane" id="l6-ct" role="tabpanel" aria-labelledby="l6-ct-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-7.png" width="1344"></p>
</div>
<div class="tab-pane" id="l6-it" role="tabpanel" aria-labelledby="l6-it-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-8.png" width="1344"></p>
</div>
<div class="tab-pane" id="l6b" role="tabpanel" aria-labelledby="l6b-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-9.png" width="1344"></p>
</div>
<div class="tab-pane" id="lamp5" role="tabpanel" aria-labelledby="lamp5-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-10.png" width="1344"></p>
</div>
<div class="tab-pane" id="macrophage" role="tabpanel" aria-labelledby="macrophage-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-11.png" width="1344"></p>
</div>
<div class="tab-pane" id="meis2" role="tabpanel" aria-labelledby="meis2-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-12.png" width="1344"></p>
</div>
<div class="tab-pane" id="np" role="tabpanel" aria-labelledby="np-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-13.png" width="1344"></p>
</div>
<div class="tab-pane" id="oligo" role="tabpanel" aria-labelledby="oligo-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-14.png" width="1344"></p>
</div>
<div class="tab-pane" id="peri" role="tabpanel" aria-labelledby="peri-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-15.png" width="1344"></p>
</div>
<div class="tab-pane" id="pvalb" role="tabpanel" aria-labelledby="pvalb-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-16.png" width="1344"></p>
</div>
<div class="tab-pane" id="serpinf1" role="tabpanel" aria-labelledby="serpinf1-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-17.png" width="1344"></p>
</div>
<div class="tab-pane" id="smc" role="tabpanel" aria-labelledby="smc-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-18.png" width="1344"></p>
</div>
<div class="tab-pane" id="sncg" role="tabpanel" aria-labelledby="sncg-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-19.png" width="1344"></p>
</div>
<div class="tab-pane" id="sst" role="tabpanel" aria-labelledby="sst-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-20.png" width="1344"></p>
</div>
<div class="tab-pane" id="vip" role="tabpanel" aria-labelledby="vip-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-21.png" width="1344"></p>
</div>
<div class="tab-pane" id="vlmc" role="tabpanel" aria-labelledby="vlmc-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-21-22.png" width="1344"></p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="mouse-kidney">Mouse kidney<a class="anchor" aria-label="anchor" href="#mouse-kidney"></a>
</h2>
<p>The same comparison approach as for the mouse brain data is applied
to study the results for the mouse kidney sample, producing spot-wise
correlation values between the cell type mapping results using the three
different methods included for the comparison.</p>
<div class="float">
<img src="compare_cell_type_mapping_NNLS/mkidney_cor_heatmaps.jpg" alt="3. Correlation plots of cell type mapping results in mouse kidney dataset."><div class="figcaption">3. Correlation plots of cell type mapping
results in mouse kidney dataset.</div>
</div>
<div class="float">
<img src="compare_cell_type_mapping_NNLS/mkidney_cor_bars.jpg" alt="4. Comparison of correlation values against NNLS for cell types in mouse kidney dataset."><div class="figcaption">4. Comparison of correlation values against NNLS
for cell types in mouse kidney dataset.</div>
</div>
<p>Also for this data set, there is an overall high cell-cell
correlation between the NNLS results and those produced by stereoscope
and cell2location. However, we can for instance see that all three
methods seem to have difficulties correctly mapping B cells, as there
seem to be very little concordance across the method mapping
results.</p>
<div class="section level3 tabset">
<h3 class="tabset" id="spatial-cell-type-mapping-1">Spatial cell type mapping<a class="anchor" aria-label="anchor" href="#spatial-cell-type-mapping-1"></a>
</h3>















<ul class="nav nav-tabs" id="spatial-cell-type-mapping-1" role="tablist">
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#b-cell" id="b-cell-tab" type="button" role="tab" aria-controls="b-cell" aria-selected="true" class="active nav-link">B Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#brush-cell" id="brush-cell-tab" type="button" role="tab" aria-controls="brush-cell" aria-selected="false" class="nav-link">Brush Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#fenestrated-capillary-endothelial" id="fenestrated-capillary-endothelial-tab" type="button" role="tab" aria-controls="fenestrated-capillary-endothelial" aria-selected="false" class="nav-link">Fenestrated Capillary Endothelial</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#fibroblast" id="fibroblast-tab" type="button" role="tab" aria-controls="fibroblast" aria-selected="false" class="nav-link">Fibroblast</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#kidney-capillary-endothelial-cell" id="kidney-capillary-endothelial-cell-tab" type="button" role="tab" aria-controls="kidney-capillary-endothelial-cell" aria-selected="false" class="nav-link">Kidney Capillary Endothelial Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#kidney-collecting-duct-principal-cell" id="kidney-collecting-duct-principal-cell-tab" type="button" role="tab" aria-controls="kidney-collecting-duct-principal-cell" aria-selected="false" class="nav-link">Kidney Collecting Duct Principal Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#kidney-cortex-artery-cell" id="kidney-cortex-artery-cell-tab" type="button" role="tab" aria-controls="kidney-cortex-artery-cell" aria-selected="false" class="nav-link">Kidney Cortex Artery Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#kidney-distal-convoluted-tubule-epithelial-cell" id="kidney-distal-convoluted-tubule-epithelial-cell-tab" type="button" role="tab" aria-controls="kidney-distal-convoluted-tubule-epithelial-cell" aria-selected="false" class="nav-link">Kidney Distal Convoluted Tubule Epithelial Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#kidney-mesangial-cell" id="kidney-mesangial-cell-tab" type="button" role="tab" aria-controls="kidney-mesangial-cell" aria-selected="false" class="nav-link">Kidney Mesangial Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#kidney-proximal-convoluted-tubule-epithelial-cell" id="kidney-proximal-convoluted-tubule-epithelial-cell-tab" type="button" role="tab" aria-controls="kidney-proximal-convoluted-tubule-epithelial-cell" aria-selected="false" class="nav-link">Kidney Proximal Convoluted Tubule Epithelial Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#macrophage-1" id="macrophage-1-tab" type="button" role="tab" aria-controls="macrophage-1" aria-selected="false" class="nav-link">Macrophage</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#podocyte" id="podocyte-tab" type="button" role="tab" aria-controls="podocyte" aria-selected="false" class="nav-link">Podocyte</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#proximal-tube-epithelial-cell" id="proximal-tube-epithelial-cell-tab" type="button" role="tab" aria-controls="proximal-tube-epithelial-cell" aria-selected="false" class="nav-link">Proximal Tube Epithelial Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#t-cell" id="t-cell-tab" type="button" role="tab" aria-controls="t-cell" aria-selected="false" class="nav-link">T Cell</button></li>
<li role="presentation" class="nav-item"><button data-bs-toggle="tab" data-bs-target="#thick-ascending-tube-s-epithelial-cell" id="thick-ascending-tube-s-epithelial-cell-tab" type="button" role="tab" aria-controls="thick-ascending-tube-s-epithelial-cell" aria-selected="false" class="nav-link">Thick Ascending Tube S Epithelial Cell</button></li>
</ul>
<div class="tab-content">
<div class="active  tab-pane" id="b-cell" role="tabpanel" aria-labelledby="b-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-1.png" width="672"></p>
</div>
<div class="tab-pane" id="brush-cell" role="tabpanel" aria-labelledby="brush-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-2.png" width="672"></p>
</div>
<div class="tab-pane" id="fenestrated-capillary-endothelial" role="tabpanel" aria-labelledby="fenestrated-capillary-endothelial-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-3.png" width="672"></p>
</div>
<div class="tab-pane" id="fibroblast" role="tabpanel" aria-labelledby="fibroblast-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-4.png" width="672"></p>
</div>
<div class="tab-pane" id="kidney-capillary-endothelial-cell" role="tabpanel" aria-labelledby="kidney-capillary-endothelial-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-5.png" width="672"></p>
</div>
<div class="tab-pane" id="kidney-collecting-duct-principal-cell" role="tabpanel" aria-labelledby="kidney-collecting-duct-principal-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-6.png" width="672"></p>
</div>
<div class="tab-pane" id="kidney-cortex-artery-cell" role="tabpanel" aria-labelledby="kidney-cortex-artery-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-7.png" width="672"></p>
</div>
<div class="tab-pane" id="kidney-distal-convoluted-tubule-epithelial-cell" role="tabpanel" aria-labelledby="kidney-distal-convoluted-tubule-epithelial-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-8.png" width="672"></p>
</div>
<div class="tab-pane" id="kidney-mesangial-cell" role="tabpanel" aria-labelledby="kidney-mesangial-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-9.png" width="672"></p>
</div>
<div class="tab-pane" id="kidney-proximal-convoluted-tubule-epithelial-cell" role="tabpanel" aria-labelledby="kidney-proximal-convoluted-tubule-epithelial-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-10.png" width="672"></p>
</div>
<div class="tab-pane" id="macrophage-1" role="tabpanel" aria-labelledby="macrophage-1-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-11.png" width="672"></p>
</div>
<div class="tab-pane" id="podocyte" role="tabpanel" aria-labelledby="podocyte-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-12.png" width="672"></p>
</div>
<div class="tab-pane" id="proximal-tube-epithelial-cell" role="tabpanel" aria-labelledby="proximal-tube-epithelial-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-13.png" width="672"></p>
</div>
<div class="tab-pane" id="t-cell" role="tabpanel" aria-labelledby="t-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-14.png" width="672"></p>
</div>
<div class="tab-pane" id="thick-ascending-tube-s-epithelial-cell" role="tabpanel" aria-labelledby="thick-ascending-tube-s-epithelial-cell-tab">

<p><img src="compare_cell_type_mapping_NNLS_files/figure-html/unnamed-chunk-32-15.png" width="672"></p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>In summary, we here demonstrate the utility of the NNLS approach for
cell type decomposition of Visium data, using two separate data sets, by
comparing it with two well established methods, <em>stereoscope</em> and
<em>cell2location</em>. Overall, the concordance in the inferred cell
type spot proportion is high between all methods, showing that NNLS is
an acceptable alternative for performing a fast cell type mapping
analysis of your Visium data, where a good reference scRNA-seq data set
is available. We would still recommend users to try out additional
methods on their data in order to fully appreciate the variability in
the mapping results that may occur due to design differences of these
algorithms. Nonetheless, the NNLS approach provided in
<code>semla</code> can act as an initial test ground for optimizing cell
type annotation levels or filtering parameters in a quick and versatile
fashion, before moving on to a slightly more computationally heavy
approach.</p>
<p><br></p>
<hr>
<details open><summary><strong>Package versions</strong>
</summary><ul>
<li>
<code>semla</code>: 1.4.0</li>
</ul></details><details><summary><strong>Session info</strong>
</summary><pre><code><span><span class="co">## R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20.0.0</span></span>
<span><span class="co">## Running under: macOS Sequoia 15.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /Users/javierescudero/miniconda3/envs/r-semlaupd/lib/libopenblas.0.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Stockholm</span></span>
<span><span class="co">## tzcode source: system (macOS)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] patchwork_1.3.1    tidyr_1.3.1        tibble_3.2.1       semla_1.4.0       </span></span>
<span><span class="co">## [5] ggplot2_3.5.2      dplyr_1.1.4        Seurat_5.3.0       SeuratObject_5.1.0</span></span>
<span><span class="co">## [9] sp_2.2-0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3     rstudioapi_0.17.1      jsonlite_1.9.0        </span></span>
<span><span class="co">##   [4] magrittr_2.0.3         spatstat.utils_3.1-4   magick_2.8.7          </span></span>
<span><span class="co">##   [7] farver_2.1.2           rmarkdown_2.29         fs_1.6.5              </span></span>
<span><span class="co">##  [10] ragg_1.3.3             vctrs_0.6.5            ROCR_1.0-11           </span></span>
<span><span class="co">##  [13] spatstat.explore_3.4-3 htmltools_0.5.8.1      forcats_1.0.0         </span></span>
<span><span class="co">##  [16] sass_0.4.9             sctransform_0.4.2      parallelly_1.42.0     </span></span>
<span><span class="co">##  [19] KernSmooth_2.23-26     bslib_0.9.0            htmlwidgets_1.6.4     </span></span>
<span><span class="co">##  [22] desc_1.4.3             ica_1.0-3              plyr_1.8.9            </span></span>
<span><span class="co">##  [25] plotly_4.11.0          zoo_1.8-13             cachem_1.1.0          </span></span>
<span><span class="co">##  [28] igraph_2.1.4           mime_0.12              lifecycle_1.0.4       </span></span>
<span><span class="co">##  [31] pkgconfig_2.0.3        Matrix_1.7-2           R6_2.6.1              </span></span>
<span><span class="co">##  [34] fastmap_1.2.0          fitdistrplus_1.2-3     future_1.34.0         </span></span>
<span><span class="co">##  [37] shiny_1.10.0           digest_0.6.37          colorspace_2.1-1      </span></span>
<span><span class="co">##  [40] tensor_1.5.1           RSpectra_0.16-2        irlba_2.3.5.1         </span></span>
<span><span class="co">##  [43] textshaping_0.4.0      labeling_0.4.3         progressr_0.15.1      </span></span>
<span><span class="co">##  [46] spatstat.sparse_3.1-0  httr_1.4.7             polyclip_1.10-7       </span></span>
<span><span class="co">##  [49] abind_1.4-5            compiler_4.4.2         withr_3.0.2           </span></span>
<span><span class="co">##  [52] viridis_0.6.5          fastDummies_1.7.5      MASS_7.3-64           </span></span>
<span><span class="co">##  [55] tools_4.4.2            lmtest_0.9-40          httpuv_1.6.15         </span></span>
<span><span class="co">##  [58] future.apply_1.11.3    goftest_1.2-3          glue_1.8.0            </span></span>
<span><span class="co">##  [61] dbscan_1.2.2           nlme_3.1-167           promises_1.3.2        </span></span>
<span><span class="co">##  [64] grid_4.4.2             Rtsne_0.17             cluster_2.1.8         </span></span>
<span><span class="co">##  [67] reshape2_1.4.4         generics_0.1.3         gtable_0.3.6          </span></span>
<span><span class="co">##  [70] spatstat.data_3.1-6    data.table_1.17.0      spatstat.geom_3.4-1   </span></span>
<span><span class="co">##  [73] RcppAnnoy_0.0.22       ggrepel_0.9.6          RANN_2.6.2            </span></span>
<span><span class="co">##  [76] pillar_1.10.1          stringr_1.5.1          spam_2.11-1           </span></span>
<span><span class="co">##  [79] RcppHNSW_0.6.0         later_1.4.1            splines_4.4.2         </span></span>
<span><span class="co">##  [82] lattice_0.22-6         survival_3.8-3         deldir_2.0-4          </span></span>
<span><span class="co">##  [85] tidyselect_1.2.1       miniUI_0.1.1.1         pbapply_1.7-2         </span></span>
<span><span class="co">##  [88] knitr_1.50             gridExtra_2.3          scattermore_1.2       </span></span>
<span><span class="co">##  [91] xfun_0.53              matrixStats_1.5.0      stringi_1.8.4         </span></span>
<span><span class="co">##  [94] lazyeval_0.2.2         yaml_2.3.10            evaluate_1.0.5        </span></span>
<span><span class="co">##  [97] codetools_0.2-20       cli_3.6.4              uwot_0.2.3            </span></span>
<span><span class="co">## [100] xtable_1.8-4           reticulate_1.42.0      systemfonts_1.2.1     </span></span>
<span><span class="co">## [103] munsell_0.5.1          jquerylib_0.1.4        Rcpp_1.1.0            </span></span>
<span><span class="co">## [106] globals_0.16.3         spatstat.random_3.4-1  zeallot_0.2.0         </span></span>
<span><span class="co">## [109] png_0.1-8              spatstat.univar_3.1-3  parallel_4.4.2        </span></span>
<span><span class="co">## [112] pkgdown_2.1.1          dotCall64_1.2          listenv_0.9.1         </span></span>
<span><span class="co">## [115] viridisLite_0.4.2      scales_1.3.0           ggridges_0.5.6        </span></span>
<span><span class="co">## [118] purrr_1.0.4            rlang_1.1.5            cowplot_1.1.3         </span></span>
<span><span class="co">## [121] shinyjs_2.1.0</span></span></code></pre>
</details><p><br></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://github.com/jemorlanes" class="external-link">Javier Escudero Morlanes</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference%2Ffigures%2Fsr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
