<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="semla">
<title>Create dataset • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Create dataset">
<meta property="og:description" content="semla">
<meta property="og:image" content="https://ludvigla.github.io/semla/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/semla.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Intro</h6>
    <a class="dropdown-item" href="../articles/getting_started.html">Get started</a>
    <a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6>
    <a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a>
    <a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a>
    <a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a>
    <a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial methods</h6>
    <a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a>
    <a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a>
    <a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a>
    <a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Analysis</h6>
    <a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a>
    <a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a>
    <a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image processing</h6>
    <a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a>
    <a class="dropdown-item" href="../articles/mask_images.html">Image masking</a>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
  </div>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/create_object.html">Create a Staffli object</a>
    <a class="dropdown-item" href="../articles/compare_cell_type_mapping_NNLS.html">Cell type mapping comparison</a>
    <a class="dropdown-item" href="../articles/cell_type_mapping_benchmark.html">Cell type mapping benchmark</a>
    <a class="dropdown-item" href="../articles/slide-seq.html">Slide-Seq data</a>
    <a class="dropdown-item" href="../articles/IF_data.html">IF data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Create dataset</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 26 September
2023</h4>
      
      
      <div class="d-none name"><code>create_object.Rmd</code></div>
    </div>

    
    
<p>In this notebook, we’ll have a look at how you can create a
<code>Seurat</code> object compatible with <code>semla</code>.</p>
<div class="section level2">
<h2 id="load-libraries">Load libraries<a class="anchor" aria-label="anchor" href="#load-libraries"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ludvigla.github.io/semla/">semla</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/magick/" class="external-link">magick</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jeroen.r-universe.dev/jsonlite" class="external-link">jsonlite</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-coordinates">Load coordinates<a class="anchor" aria-label="anchor" href="#load-coordinates"></a>
</h2>
<p>The <code>Staffli</code> object requires image and coordinate data
that can be retrieved from the spaceranger output files.</p>
<p>Below we’ll use the mouse brain and mouse colon example data:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">he_imgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousebrain"</span>, </span>
<span>                         <span class="st">"spatial/tissue_lowres_image.jpg"</span>, </span>
<span>                          package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousecolon"</span>, </span>
<span>                         <span class="st">"spatial/tissue_lowres_image.jpg"</span>, </span>
<span>                         package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">spotfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousebrain"</span>, </span>
<span>                           <span class="st">"spatial/tissue_positions_list.csv"</span>, </span>
<span>                           package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span>,</span>
<span>               <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousecolon"</span>, </span>
<span>                           <span class="st">"spatial/tissue_positions_list.csv"</span>, </span>
<span>                           package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">jsonfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousebrain"</span>, </span>
<span>                           <span class="st">"spatial/scalefactors_json.json"</span>, </span>
<span>                           package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span>,</span>
<span>               <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousecolon"</span>, </span>
<span>                           <span class="st">"spatial/scalefactors_json.json"</span>, </span>
<span>                           package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>When loading coordinates from multiple samples, you need to change
the barcode IDs so that the suffix matches their respective sampleID.
For instance, a barcode in sample 1 might be called CATACAAAGCCGAACC-1
and in sample 2 the same barcode should be CATACAAAGCCGAACC-2. The
<code>coordinates</code> tibble below contains the barcode IDs, the
pixel coordinates and a column with the sampleIDs.</p>
<p><code>semla</code> provides <code>LoadSpatialCoordinates</code> to
make the task a little easier. Note that the suffix for the barcodes
have changed for sample 2.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read coordinates</span></span>
<span><span class="va">coordinates</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadSpatialCoordinates.html">LoadSpatialCoordinates</a></span><span class="op">(</span><span class="va">spotfiles</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"barcode"</span>, <span class="st">"pxl_col_in_fullres"</span>, <span class="st">"pxl_row_in_fullres"</span>, <span class="st">"sampleID"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Loading coordinates:</span></span></code></pre>
<pre><code><span><span class="co">## →   Finished loading coordinates for sample 1</span></span></code></pre>
<pre><code><span><span class="co">## →   Finished loading coordinates for sample 2</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Collected coordinates for <span style="color: #FF55FF;">5164</span> spots.</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">coordinates</span>, n <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 4</span></span></span>
<span><span class="co">##   barcode            pxl_col_in_fullres pxl_row_in_fullres sampleID</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> CATACAAAGCCGAACC-1               <span style="text-decoration: underline;">6</span>086               <span style="text-decoration: underline;">4</span>117        1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> CTGAGCAAGTAACAAG-1               <span style="text-decoration: underline;">5</span>062               <span style="text-decoration: underline;">4</span>472        1</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">coordinates</span>, n <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 4</span></span></span>
<span><span class="co">##   barcode            pxl_col_in_fullres pxl_row_in_fullres sampleID</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                           <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> GAAGGAGTCGAGTGCG-2               <span style="text-decoration: underline;">4</span>854               <span style="text-decoration: underline;">6</span>829        2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> TGACCAAATCTTAAAC-2               <span style="text-decoration: underline;">4</span>957               <span style="text-decoration: underline;">6</span>829        2</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check number of spots per sample</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">coordinates</span><span class="op">$</span><span class="va">sampleID</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    1    2 </span></span>
<span><span class="co">## 2560 2604</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="fetch-image-info">Fetch image info<a class="anchor" aria-label="anchor" href="#fetch-image-info"></a>
</h2>
<p>Next, we’ll fetch meta data for the sample H&amp;E images. Here, the
tibble must contain the width and height of <code>he_imgs</code> as well
as a sampleID column. Use <code>LoadImageInfo</code> to load the image
info:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create image_info</span></span>
<span><span class="va">image_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadImageInfo.html">LoadImageInfo</a></span><span class="op">(</span><span class="va">he_imgs</span><span class="op">)</span></span>
<span><span class="va">image_info</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 9</span></span></span>
<span><span class="co">##   format width height colorspace matte filesize density sampleID type         </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> JPEG     565    600 sRGB       FALSE   <span style="text-decoration: underline;">106</span>233 +72x+72 1        tissue_lowres</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> JPEG     600    541 sRGB       FALSE   <span style="text-decoration: underline;">141</span>710 +72x+72 2        tissue_lowres</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="define-image-dimensions">Define image dimensions<a class="anchor" aria-label="anchor" href="#define-image-dimensions"></a>
</h2>
<p>The <code>jsonfiles</code> contain scaling factors that allows us to
find the original H&amp;E image dimensions. We can load the scale
factors with <code>read_json</code> and store them in a tibble. The
tibble should also contain a ‘sampleID’ column. Use
<code>LoadScaleFactors</code> to load the scaling factors:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read scalefactors</span></span>
<span><span class="va">scalefactors</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadScaleFactors.html">LoadScaleFactors</a></span><span class="op">(</span><span class="va">jsonfiles</span><span class="op">)</span></span>
<span><span class="va">scalefactors</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 5</span></span></span>
<span><span class="co">##   spot_diameter_fullres tissue_hires_scalef fiducial_diameter_fullres</span></span>
<span><span class="co">##                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>                 143.                0.104                      215.</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>                  67.0               0.202                      108.</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: tissue_lowres_scalef &lt;dbl&gt;, sampleID &lt;chr&gt;</span></span></span></code></pre>
<p>Once we have the scale factors, we can add additional columns to
<code>image_info</code> which specify the dimensions of the original
H&amp;E images. Use <code>UpdateImageInfo</code> to update
<code>image_info</code> with the scaling factors:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add additional columns to image_info using scale factors</span></span>
<span><span class="va">image_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/UpdateImageInfo.html">UpdateImageInfo</a></span><span class="op">(</span><span class="va">image_info</span>, <span class="va">scalefactors</span><span class="op">)</span></span>
<span><span class="va">image_info</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 10</span></span></span>
<span><span class="co">##   format width height full_width full_height colorspace filesize density</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> JPEG     565    600     <span style="text-decoration: underline;">18</span>120.      <span style="text-decoration: underline;">19</span>242. sRGB         <span style="text-decoration: underline;">106</span>233 +72x+72</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> JPEG     600    541      <span style="text-decoration: underline;">9</span>901.       <span style="text-decoration: underline;">8</span>927. sRGB         <span style="text-decoration: underline;">141</span>710 +72x+72</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: sampleID &lt;chr&gt;, type &lt;chr&gt;</span></span></span></code></pre>
<p>Finally, we are ready to create the <code>Staffli</code> object. Here
we’ll provide the <code>he_imgs</code> so that we can load the images
later with <code>LoadImages</code>. The <code>coordinates</code> are
stored in the <code>meta_data</code> slot. <code>image_info</code> will
be used by the plot functions provided in <code>semla</code> to define
the dimensions of the plot area.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create Staffli object</span></span>
<span><span class="va">staffli_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CreateStaffliObject.html">CreateStaffliObject</a></span><span class="op">(</span>imgs <span class="op">=</span> <span class="va">he_imgs</span>, </span>
<span>                                      meta_data <span class="op">=</span> <span class="va">coordinates</span>, </span>
<span>                                      image_info <span class="op">=</span> <span class="va">image_info</span>, </span>
<span>                                      scalefactors <span class="op">=</span> <span class="va">scalefactors</span><span class="op">)</span></span>
<span><span class="va">staffli_object</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Staffli </span></span>
<span><span class="co">## 5164 spots across 2 samples.</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="use-a-staffli-object-in-seurat">Use a <code>Staffli</code> object in <code>Seurat</code><a class="anchor" aria-label="anchor" href="#use-a-staffli-object-in-seurat"></a>
</h2>
<p>Now let’s see how we can incorporate our <code>Staffli</code> object
in <code>Seurat</code>. First, we’ll load the gene expression matrices
for our samples and merge them.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get paths for expression matrices</span></span>
<span><span class="va">expr_matrix_files</span> <span class="op">&lt;-</span> <span class="va">he_imgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousebrain"</span>, </span>
<span>                         <span class="st">"filtered_feature_bc_matrix.h5"</span>, </span>
<span>                          package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span>,</span>
<span>             <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousecolon"</span>, </span>
<span>                         <span class="st">"filtered_feature_bc_matrix.h5"</span>, </span>
<span>                         package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Before merging the matrices, it’s important to rename the barcodes so
that they match the barcodes in <code>Staffli_object</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exprMatList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">expr_matrix_files</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">exprMat</span> <span class="op">&lt;-</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/Read10X_h5.html" class="external-link">Read10X_h5</a></span><span class="op">(</span><span class="va">expr_matrix_files</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">exprMat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"-\\d*"</span>, <span class="co"># Replace barcode suffix with sampleID</span></span>
<span>                          replacement <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="va">i</span><span class="op">)</span>, </span>
<span>                          x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">exprMat</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">exprMat</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Merge expression matrices</span></span>
<span><span class="va">mergedExprMat</span> <span class="op">&lt;-</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/RowMergeSparseMatrices.html" class="external-link">RowMergeSparseMatrices</a></span><span class="op">(</span><span class="va">exprMatList</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">exprMatList</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, use the <code>LoadAndMergeMatrices</code> function
from <code>semla</code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mergedExprMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/LoadAndMergeMatrices.html">LoadAndMergeMatrices</a></span><span class="op">(</span>samplefiles <span class="op">=</span> <span class="va">expr_matrix_files</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create Seurat object</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">mergedExprMat</span><span class="op">)</span></span></code></pre></div>
<p>Rearrange <code>Staffli</code> object to match <code>Seurat</code>
object. If the barcode are mismatched, make sure to subset the two
object to contain intersecting barcodes.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">staffli_object</span><span class="op">@</span><span class="va">meta_data</span> <span class="op">&lt;-</span> <span class="va">staffli_object</span><span class="op">@</span><span class="va">meta_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>, <span class="va">staffli_object</span><span class="op">@</span><span class="va">meta_data</span><span class="op">$</span><span class="va">barcode</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<p>Check that the <code>Staffli</code> object matches
<code>Seurat</code> object:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">staffli_object</span><span class="op">@</span><span class="va">meta_data</span><span class="op">$</span><span class="va">barcode</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>Place <code>Staffli</code> object in the <code>Seurat</code>
object:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">@</span><span class="va">tools</span><span class="op">$</span><span class="va">Staffli</span> <span class="op">&lt;-</span> <span class="va">staffli_object</span></span></code></pre></div>
<p>Now the <code>semla</code> functions should be compatible with the
<code>Seurat</code> object!</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, features <span class="op">=</span> <span class="st">"nFeature_RNA"</span><span class="op">)</span></span></code></pre></div>
<p><img src="create_object_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="va">se</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, features <span class="op">=</span> <span class="st">"nFeature_RNA"</span>, image_use <span class="op">=</span> <span class="st">"raw"</span>, override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="create_object_files/figure-html/unnamed-chunk-14-2.png" width="700"></p>
<p><br></p>
<hr>
<details open><summary><strong>Package version</strong>
</summary><ul>
<li>
<code>semla</code>: 1.1.6</li>
</ul></details><details><summary><strong>Session info</strong>
</summary><div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices datasets  utils     methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] tibble_3.2.1       jsonlite_1.8.7     magick_2.7.4       semla_1.1.6       </span></span>
<span><span class="co">## [5] ggplot2_3.4.3      dplyr_1.1.3        SeuratObject_4.1.3 Seurat_4.3.0.1    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] systemfonts_1.0.4      plyr_1.8.8             igraph_1.5.1          </span></span>
<span><span class="co">##   [4] lazyeval_0.2.2         sp_2.0-0               splines_4.2.3         </span></span>
<span><span class="co">##   [7] listenv_0.9.0          scattermore_1.2        digest_0.6.33         </span></span>
<span><span class="co">##  [10] htmltools_0.5.6        fansi_1.0.4            magrittr_2.0.3        </span></span>
<span><span class="co">##  [13] memoise_2.0.1          tensor_1.5             cluster_2.1.4         </span></span>
<span><span class="co">##  [16] ROCR_1.0-11            globals_0.16.2         matrixStats_1.0.0     </span></span>
<span><span class="co">##  [19] pkgdown_2.0.7          spatstat.sparse_3.0-2  colorspace_2.1-0      </span></span>
<span><span class="co">##  [22] ggrepel_0.9.3          textshaping_0.3.6      xfun_0.40             </span></span>
<span><span class="co">##  [25] progressr_0.14.0       spatstat.data_3.0-1    zeallot_0.1.0         </span></span>
<span><span class="co">##  [28] survival_3.5-7         zoo_1.8-12             glue_1.6.2            </span></span>
<span><span class="co">##  [31] polyclip_1.10-4        gtable_0.3.4           leiden_0.4.3          </span></span>
<span><span class="co">##  [34] future.apply_1.11.0    abind_1.4-5            scales_1.2.1          </span></span>
<span><span class="co">##  [37] spatstat.random_3.1-5  miniUI_0.1.1.1         Rcpp_1.0.11           </span></span>
<span><span class="co">##  [40] viridisLite_0.4.2      xtable_1.8-4           reticulate_1.31       </span></span>
<span><span class="co">##  [43] bit_4.0.5              htmlwidgets_1.6.2      httr_1.4.7            </span></span>
<span><span class="co">##  [46] RColorBrewer_1.1-3     ellipsis_0.3.2         ica_1.0-3             </span></span>
<span><span class="co">##  [49] pkgconfig_2.0.3        farver_2.1.1           sass_0.4.7            </span></span>
<span><span class="co">##  [52] uwot_0.1.16            deldir_1.0-9           utf8_1.2.3            </span></span>
<span><span class="co">##  [55] tidyselect_1.2.0       labeling_0.4.3         rlang_1.1.1           </span></span>
<span><span class="co">##  [58] reshape2_1.4.4         later_1.3.1            munsell_0.5.0         </span></span>
<span><span class="co">##  [61] tools_4.2.3            cachem_1.0.8           cli_3.6.1             </span></span>
<span><span class="co">##  [64] dbscan_1.1-11          generics_0.1.3         ggridges_0.5.4        </span></span>
<span><span class="co">##  [67] evaluate_0.21          stringr_1.5.0          fastmap_1.1.1         </span></span>
<span><span class="co">##  [70] yaml_2.3.7             ragg_1.2.5             goftest_1.2-3         </span></span>
<span><span class="co">##  [73] knitr_1.43             bit64_4.0.5            fs_1.6.3              </span></span>
<span><span class="co">##  [76] fitdistrplus_1.1-11    purrr_1.0.2            RANN_2.6.1            </span></span>
<span><span class="co">##  [79] pbapply_1.7-2          future_1.33.0          nlme_3.1-163          </span></span>
<span><span class="co">##  [82] mime_0.12              hdf5r_1.3.8            compiler_4.2.3        </span></span>
<span><span class="co">##  [85] plotly_4.10.2          png_0.1-8              spatstat.utils_3.0-3  </span></span>
<span><span class="co">##  [88] bslib_0.5.1            stringi_1.7.12         highr_0.10            </span></span>
<span><span class="co">##  [91] desc_1.4.2             forcats_1.0.0          lattice_0.21-8        </span></span>
<span><span class="co">##  [94] Matrix_1.5-1           shinyjs_2.1.0          vctrs_0.6.3           </span></span>
<span><span class="co">##  [97] pillar_1.9.0           lifecycle_1.0.3        BiocManager_1.30.22   </span></span>
<span><span class="co">## [100] spatstat.geom_3.2-4    lmtest_0.9-40          jquerylib_0.1.4       </span></span>
<span><span class="co">## [103] RcppAnnoy_0.0.21       data.table_1.14.8      cowplot_1.1.1         </span></span>
<span><span class="co">## [106] irlba_2.3.5.1          httpuv_1.6.11          patchwork_1.1.3       </span></span>
<span><span class="co">## [109] R6_2.5.1               promises_1.2.1         renv_1.0.2            </span></span>
<span><span class="co">## [112] KernSmooth_2.23-22     gridExtra_2.3          parallelly_1.36.0     </span></span>
<span><span class="co">## [115] codetools_0.2-19       MASS_7.3-60            rprojroot_2.0.3       </span></span>
<span><span class="co">## [118] withr_2.5.0            sctransform_0.3.5      parallel_4.2.3        </span></span>
<span><span class="co">## [121] grid_4.2.3             tidyr_1.3.0            rmarkdown_2.24        </span></span>
<span><span class="co">## [124] Rtsne_0.16             spatstat.explore_3.2-1 shiny_1.7.5</span></span></code></pre>
</details><p><br></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
