<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="semla">
<title>Digital unrolling • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Digital unrolling">
<meta property="og:description" content="semla">
<meta property="og:image" content="https://ludvigla.github.io/semla/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.3.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/semla.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Intro</h6>
    <a class="dropdown-item" href="../articles/getting_started.html">Get started</a>
    <a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a>
    <a class="dropdown-item" href="../articles/visiumHD.html">VisiumHD</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6>
    <a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a>
    <a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a>
    <a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a>
    <a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial methods</h6>
    <a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a>
    <a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a>
    <a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a>
    <a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Analysis</h6>
    <a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a>
    <a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a>
    <a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image processing</h6>
    <a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a>
    <a class="dropdown-item" href="../articles/mask_images.html">Image masking</a>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/create_object.html">Create a Staffli object</a>
    <a class="dropdown-item" href="../articles/compare_cell_type_mapping_NNLS.html">Cell type mapping comparison</a>
    <a class="dropdown-item" href="../articles/cell_type_mapping_benchmark.html">Cell type mapping benchmark</a>
    <a class="dropdown-item" href="../articles/slide-seq.html">Slide-Seq data</a>
    <a class="dropdown-item" href="../articles/IF_data.html">IF data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Digital unrolling</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 19 March
2025</h4>
      
      
      <div class="d-none name"><code>digital_unrolling.rmd</code></div>
    </div>

    
    
<p>In this tutorial, we will look at how to perform digital unrolling of
folded “swiss roll” tissue sections.</p>
<p>The data set comes from a “swiss roll” of a mouse colon which was
published by M. Parigi et al. (including yours truly) <a href="https://www.nature.com/articles/s41467-022-28497-0" class="external-link"><em>The
spatial transcriptomic landscape of the healing mouse intestine
following damage</em></a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ludvigla.github.io/semla/">semla</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="mouse-colon-data">Mouse colon data<a class="anchor" aria-label="anchor" href="#mouse-colon-data"></a>
</h2>
<p>Let’s load and have a quick look at the data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mcolon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousecolon"</span>, </span>
<span>                                 <span class="st">"se_mcolon"</span>, </span>
<span>                                 package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_mcolon</span> <span class="op">&lt;-</span> <span class="va">se_mcolon</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## ── <span style="font-weight: bold;">Loading H&amp;E images</span> ──</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Loading image from /private/var/folders/z8/nrcst881607gn95xh3_4qvb00000gp/T/Rtmpf80zo9/temp_libpath1c4b5544360a/semla/extdata/mousecolon/spatial/tissue_lowres_image.jpg</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Scaled image from 541x600 to 400x444 pixels</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Saving loaded H&amp;E images as 'rasters' in Seurat object</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ImagePlot.html">ImagePlot</a></span><span class="op">(</span><span class="va">se_mcolon</span><span class="op">)</span></span></code></pre></div>
<p><img src="digital_unrolling_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>In the H&amp;E image, we can clearly see the organization of large
part of the lower intestine. The outermost part of the roll is the
ascending colon and the center of the roll is the rectum.</p>
<p>We can identify genes with a regionalized pattern pretty easily but
it would we great if we could somehow unroll the tissue so that we can
investigate gene expression patterns along the proximal-distal axis.</p>
<p>Below is a spatial map of the expression of Fxyd4 which is
up-regulated closer to the distal part of the colon.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se_mcolon</span>, features <span class="op">=</span> <span class="st">"Fxyd4"</span>, </span>
<span>            image_use <span class="op">=</span> <span class="st">"raw"</span>, override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkblue"</span>, <span class="st">"cyan"</span>, <span class="st">"yellow"</span>, <span class="st">"red"</span>, <span class="st">"darkred"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="digital_unrolling_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="pre-processing">Pre-processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h2>
<p>The goal of this tutorial is to show how <code>semla</code> can be
used to find an “unrolled” coordinate system. To achieve this, we can
start a shiny application with <code><a href="../reference/CutSpatialNetwork.html">CutSpatialNetwork()</a></code> which
will allow us to do some interesting stuff. More about that later.</p>
<p>Before we get started, we need to do a little bit of pre-processing.
First, we will tile the H&amp;E image using <code><a href="../reference/TileImage.html">TileImage()</a></code>.
Tiling is necessary to make it possible to visualize the H&amp;E images
in a more interactive way. The path to the raw H&amp;E image can be
found with <code>GetStaffli(se_mcolon)@imgs</code>.</p>
<p>NB: If the image file has been moved, <code>TileImage</code> the path
might no longer be valid. Make sure that the path is valid before
proceeding. To get decent resolution, we you should use the
“tissue_hires_image.png” output by spaceranger.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">im</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetStaffli.html">GetStaffli</a></span><span class="op">(</span><span class="va">se_mcolon</span><span class="op">)</span><span class="op">@</span><span class="va">imgs</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"File exists: %s"</span>, <span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">im</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "File exists: TRUE"</span></span></code></pre>
<p>Now we can load the image with <code>magick</code>, tile the image
and export it to a local directory. The function returns the path to the
directory the tiles were exported to. We will save this to a variable
called <code>tilepath</code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/magick/" class="external-link">magick</a></span><span class="op">)</span></span>
<span><span class="va">im</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://docs.ropensci.org/magick/reference/editing.html" class="external-link">image_read</a></span><span class="op">(</span><span class="va">im</span><span class="op">)</span></span>
<span><span class="va">tilepath</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TileImage.html">TileImage</a></span><span class="op">(</span>im <span class="op">=</span> <span class="va">im</span>, outpath <span class="op">=</span> <span class="st">"~/Downloads/"</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/CutSpatialNetwork.html">CutSpatialNetwork()</a></code> function requires another file
to work properly. This file holds information about the “spatial
network” (see <code><a href="../reference/get-network.html">GetSpatialNetwork()</a></code> for details) which will
be used in the shiny application to define how spots in the Visium data
are connected to each other.</p>
<p>To demonstrate what this means, we can generate a spatial network and
visualize it.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spatnet</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get-network.html">GetSpatialNetwork</a></span><span class="op">(</span><span class="va">se_mcolon</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">spatnet</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x_start</span>, xend <span class="op">=</span> <span class="va">x_end</span>, y <span class="op">=</span> <span class="va">y_start</span>, yend <span class="op">=</span> <span class="va">y_end</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="digital_unrolling_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>In the plot above, you can see the edges that connects neighboring
spots to each other.</p>
<p>Now let’s export this spatial network to make it available for our
shiny application.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/export_graph.html">export_graph</a></span><span class="op">(</span><span class="va">se_mcolon</span>, sampleID <span class="op">=</span> <span class="fl">1</span>, outdir <span class="op">=</span> <span class="va">tilepath</span><span class="op">$</span><span class="va">datapath</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prepare-network">Prepare network<a class="anchor" aria-label="anchor" href="#prepare-network"></a>
</h2>
<p>Now we are ready to run the app! Once the app loads, you should see
the H&amp;E image with spots and edges. The goal is to disconnect the
layer of the roll by cutting the edges between spots in different
layers. Cutting can be done by holding the <code>SHIFT</code> key while
move the cursor across edges which will make them red. You can also
“repair” edges by holding the <code>CTRL</code> key.</p>
<p>The image below illustrates how edges between two layers have been
sliced.</p>
<div class="float">
<img src="cut.png" alt="Sliced edges are colored in red"><div class="figcaption"><em>Sliced edges are colored in red</em></div>
</div>
<p>Once all edges has been cut, you have to press the
<code>Quit &amp; Save</code> button. If you just close the app without
pressing “Quit &amp; Save”, the results will not be returned back to
R.</p>
<p>NB: We need to make sure to save the output to a variable. Here, we
will save the output to <code>tidy_network</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tidy_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CutSpatialNetwork.html">CutSpatialNetwork</a></span><span class="op">(</span><span class="va">se_mcolon</span>, datadir <span class="op">=</span> <span class="va">tilepath</span><span class="op">$</span><span class="va">datapath</span><span class="op">)</span></span></code></pre></div>
<p>For downstream steps, it is <em>crucial</em> that the layers have
been separated appropriately. Even a single edge between two layer will
make it impossible to run down stream steps reliably. The image below
demonstrates a situation where the separation will fail.</p>
<div class="float">
<img src="wrong.png" alt="The arrow points at two edges (colored in black) which still connects two separate layers"><div class="figcaption"><em>The arrow points at two edges (colored in
black) which still connects two separate layers</em></div>
</div>
<p>Luckily, any changes you made in the app are by default written to
the spatial network file, so if you miss something, you can simply run
the app again and fix the issue.</p>
<p>Let’s plot our full network and our filtered network:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidygraph.data-imaginist.com" class="external-link">tidygraph</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">edges_full</span> <span class="op">&lt;-</span> <span class="va">tidy_network</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span><span class="va">edges</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">edges_filtered</span> <span class="op">&lt;-</span> <span class="va">tidy_network</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span><span class="va">edges</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">edges_full</span>, </span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, xend <span class="op">=</span> <span class="va">x_end</span>, <span class="va">y</span>, yend <span class="op">=</span> <span class="va">y_end</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Full network"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">edges_filtered</span>, </span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, xend <span class="op">=</span> <span class="va">x_end</span>, <span class="va">y</span>, yend <span class="op">=</span> <span class="va">y_end</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Filtered network"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="digital_unrolling_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>And just like that, we have separated the the layers!</p>
</div>
<div class="section level2">
<h2 id="unrolling-the-network">Unrolling the network<a class="anchor" aria-label="anchor" href="#unrolling-the-network"></a>
</h2>
<p>Now comes the part where we do the actual “unrolling” using
<code><a href="../reference/AdjustTissueCoordinates.html">AdjustTissueCoordinates()</a></code>. There are a few important things
to note here. <code>AdjustTissueCoordinates</code> applies an algorithm
to sort nodes in the spatial graph between the end points. Distances
between nodes are used as a proxy for the actual distances in the
tissue. These distances between nodes are the shortest paths, i.e. the
minimum number of edges needed to visit (geodesic). For obvious reasons,
distances cannot be calculated between nodes that are not connected, and
for this reason, the function only handles connected graphs.</p>
<p>Unfortunately, our roll is not fully connected, so the function will
only deal with the largest sub graph.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_networks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/morphers.html" class="external-link">to_components</a></span><span class="op">(</span><span class="va">tidy_network</span> <span class="op">|&gt;</span> </span>
<span>                                <span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span><span class="va">edges</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                                <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">edges_labeled</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">bind_rows</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">all_networks</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">all_networks</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span><span class="va">edges</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sub graph"</span>, <span class="va">i</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">edges_labeled</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, xend <span class="op">=</span> <span class="va">x_end</span>, <span class="va">y</span>, yend <span class="op">=</span> <span class="va">y_end</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html" class="external-link">geom_segment</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span><span class="op">~</span><span class="va">graph</span><span class="op">)</span></span></code></pre></div>
<p><img src="digital_unrolling_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>The second, smaller sub graph will be discarded.</p>
<p>Another thing to note is that this function relies on a few
assumptions about the shape of the folded tissue. If you were to try it
on somethings completely different, let’s say a tissue folded in a
zig-zag shape, it might not work as well.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tidy_network_adjusted</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AdjustTissueCoordinates.html">AdjustTissueCoordinates</a></span><span class="op">(</span>full_graph <span class="op">=</span> <span class="va">tidy_network</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Removing 620 edges</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Checking for disconnected graphs</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> <span style="color: #FF55FF;">More than 1 subgraph identified. Keeping the largest subgraph</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Calculating pairwise geodesics between nodes in graph</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Identifying end points</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Finding shortest path beetween end points</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Checking location of nodes relative to shortest path nodes</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Finished!</span></span></code></pre>
<p>Our <code>tidy_network_adjusted</code> variable holds a
<code>tibble</code> with the node coordinates and the new
<code>x_dist</code> and <code>y_dist</code> coordinates. As a sanity
check, we can color our original spatial map by the new distance values
to see if they make sense.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tidy_network_adjusted</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, color <span class="op">=</span> <span class="va">x_dist</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colours <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">9</span>, name <span class="op">=</span> <span class="st">"Spectral"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                          <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">tidy_network_adjusted</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, color <span class="op">=</span> <span class="va">y_dist</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colours <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">9</span>, name <span class="op">=</span> <span class="st">"Spectral"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>                          <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_fixed.html" class="external-link">coord_fixed</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_reverse</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="digital_unrolling_files/figure-html/unnamed-chunk-13-1.png" width="864"></p>
<p>The <code>x_dist</code> values, which is arguably the most relevant
for regionalization patterns, looks quite good. <code>y_dist</code>
corresponds to the geodesic distances from the nodes along the shortest
path, where negative values indicate that the spots are located outside
the shortest path and positive values indicate that the spots are
located inside the shortest path.</p>
<p>Now that we have our new coordinate system, we can start looking at
regionalization of gene expression along the proximal-distal axis.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">selected_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Fxyd4"</span>, <span class="st">"Slc37a2"</span>, <span class="st">"Hmgcs2"</span>, <span class="st">"Cyp4b1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_mcolon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">se_mcolon</span><span class="op">)</span></span>
<span><span class="va">gg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/FetchData.html" class="external-link">FetchData</a></span><span class="op">(</span><span class="va">se_mcolon</span>, vars <span class="op">=</span> <span class="va">selected_genes</span>, slot <span class="op">=</span> <span class="st">"scale.data"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">tidy_network_adjusted</span>, by <span class="op">=</span> <span class="st">"name"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="va">selected_genes</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"variable"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">gg</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">x_dist</span>, <span class="va">value</span>, color <span class="op">=</span> <span class="va">variable</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="digital_unrolling_files/figure-html/unnamed-chunk-14-1.png" width="768"></p>
<p>With these coordinates in our hands, it is quite straightforward to
extract genes whose expression depend on distance along the
proximal-distal axis. This will not be part of this tutorial, but below
is a rather simple way of sorting genes by expression in a heatmap.
Here, it becomes quite clear what genes are regionalized. There is one
part of the heatmap that stands out, just on the right side of the
center, which are the genes that are expressed in a lymphoid structure
located in that part of the colon.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sorted_spots</span> <span class="op">&lt;-</span> <span class="va">tidy_network_adjusted</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">x_dist</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">name</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get scale data</span></span>
<span><span class="va">scale_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">se_mcolon</span>, </span>
<span>                           slot <span class="op">=</span> <span class="st">"scale.data"</span>, </span>
<span>                           assay <span class="op">=</span> <span class="st">"Spatial"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">se_mcolon</span><span class="op">)</span>, <span class="va">sorted_spots</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Bin data</span></span>
<span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cut.html" class="external-link">cut</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">scale_data</span><span class="op">)</span>, breaks <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="va">binMat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">cbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">bins</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">lvl</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="va">scale_data</span><span class="op">[</span>, <span class="va">bins</span> <span class="op">==</span> <span class="va">lvl</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="va">binMat</span>, cluster_cols <span class="op">=</span> <span class="cn">FALSE</span>, border_color <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                   breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span>, length.out <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>, </span>
<span>                   color <span class="op">=</span> <span class="fu">scico</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scico/man/scico.html" class="external-link">scico</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"bam"</span>, n <span class="op">=</span> <span class="fl">51</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="digital_unrolling_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="other-applications">Other applications<a class="anchor" aria-label="anchor" href="#other-applications"></a>
</h2>
<p>The tools described here could also be used for other applications.
Maybe we are interested in a specific part of a tissue where we want to
define an axis. Let’s make a manual selection of the outermost layer of
the mouse brain and subset the <code>Seurat</code> object to contain
spots from this layer.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousebrain"</span>, <span class="st">"se_mbrain"</span>, package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span>
<span><span class="va">se_mbrain</span><span class="op">$</span><span class="va">region</span> <span class="op">&lt;-</span> <span class="st">"NA"</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FeatureViewer.html">FeatureViewer</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se_mbrain_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subset.html">SubsetSTData</a></span><span class="op">(</span><span class="va">se_mbrain</span>, expression <span class="op">=</span> <span class="va">region</span> <span class="op">==</span> <span class="st">"outer_layer"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se_mbrain_subset</span>, column_name <span class="op">=</span> <span class="st">"region"</span>, image_use <span class="op">=</span> <span class="st">"raw"</span><span class="op">)</span></span></code></pre></div>
<p><img src="digital_unrolling/mBrain.jpg"></p>
<p>Then we can create a spatial network from this layer and modify it a
bit to make it compatible with <code>AdjustTissueCoordinates</code>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spatial_network</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get-network.html">GetSpatialNetwork</a></span><span class="op">(</span><span class="va">se_mbrain_subset</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">nodes</span> <span class="op">&lt;-</span> <span class="va">spatial_network</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">from</span>, <span class="va">x_start</span>, <span class="va">y_start</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">from</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">from</span>, x <span class="op">=</span> <span class="va">x_start</span>, y <span class="op">=</span> <span class="va">y_start</span><span class="op">)</span></span>
<span><span class="va">edges</span> <span class="op">&lt;-</span> <span class="va">spatial_network</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">from</span>, <span class="va">to</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>keep <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">spatial_network</span> <span class="op">&lt;-</span> <span class="fu">tidygraph</span><span class="fu">::</span><span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/tbl_graph.html" class="external-link">tbl_graph</a></span><span class="op">(</span>nodes <span class="op">=</span> <span class="va">nodes</span>, edges <span class="op">=</span> <span class="va">edges</span>, directed <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Now that we have the spatial network, we can run
<code>AdjustTissueCoordinates</code> to compute distances between the
end points of the selected region:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run adjust tissue coordinates</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/AdjustTissueCoordinates.html">AdjustTissueCoordinates</a></span><span class="op">(</span>full_graph <span class="op">=</span> <span class="va">spatial_network</span><span class="op">)</span></span>
<span><span class="va">se_mbrain_subset</span><span class="op">$</span><span class="va">x_dist</span> <span class="op">&lt;-</span> <span class="va">test</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">test</span><span class="op">$</span><span class="va">name</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">se_mbrain_subset</span><span class="op">)</span><span class="op">)</span>, <span class="st">"x_dist"</span>, drop <span class="op">=</span> <span class="cn">TRUE</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Plot x_dist</span></span>
<span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se_mbrain_subset</span>, features <span class="op">=</span> <span class="st">"x_dist"</span>, image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>            colors <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">9</span>, name <span class="op">=</span> <span class="st">"Spectral"</span><span class="op">)</span>, </span>
<span>            override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, pt_size <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="digital_unrolling/mBrain_x_dist.jpg"></p>
<p><br></p>
<hr>
<details open><summary><strong>Package version</strong>
</summary><ul>
<li>
<code>semla</code>: 1.3.1</li>
</ul></details><details><summary><strong>Session info</strong>
</summary><div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.3 (2024-02-29)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20.0.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS 15.3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /Users/javierescudero/miniconda3/envs/r-semla/lib/libopenblas.0.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Stockholm</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] tidyr_1.3.1        tibble_3.2.1       patchwork_1.2.0    tidygraph_1.3.1   </span></span>
<span><span class="co">## [5] semla_1.3.1        ggplot2_3.5.0      dplyr_1.1.4        SeuratObject_5.0.1</span></span>
<span><span class="co">## [9] Seurat_4.3.0.1    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3     rstudioapi_0.15.0      jsonlite_1.8.8        </span></span>
<span><span class="co">##   [4] magrittr_2.0.3         spatstat.utils_3.0-5   magick_2.8.3          </span></span>
<span><span class="co">##   [7] farver_2.1.1           rmarkdown_2.26         fs_1.6.3              </span></span>
<span><span class="co">##  [10] ragg_1.3.3             vctrs_0.6.5            ROCR_1.0-11           </span></span>
<span><span class="co">##  [13] memoise_2.0.1          spatstat.explore_3.2-6 htmltools_0.5.7       </span></span>
<span><span class="co">##  [16] forcats_1.0.0          sass_0.4.8             sctransform_0.4.1     </span></span>
<span><span class="co">##  [19] parallelly_1.38.0      KernSmooth_2.23-22     bslib_0.6.1           </span></span>
<span><span class="co">##  [22] htmlwidgets_1.6.4      desc_1.4.3             ica_1.0-3             </span></span>
<span><span class="co">##  [25] plyr_1.8.9             plotly_4.10.4          zoo_1.8-12            </span></span>
<span><span class="co">##  [28] cachem_1.0.8           igraph_2.0.2           mime_0.12             </span></span>
<span><span class="co">##  [31] lifecycle_1.0.4        pkgconfig_2.0.3        Matrix_1.6-3          </span></span>
<span><span class="co">##  [34] R6_2.5.1               fastmap_1.1.1          fitdistrplus_1.1-11   </span></span>
<span><span class="co">##  [37] future_1.34.0          shiny_1.8.0            digest_0.6.34         </span></span>
<span><span class="co">##  [40] colorspace_2.1-0       tensor_1.5             irlba_2.3.5.1         </span></span>
<span><span class="co">##  [43] textshaping_0.3.7      labeling_0.4.3         progressr_0.14.0      </span></span>
<span><span class="co">##  [46] fansi_1.0.6            spatstat.sparse_3.0-3  mgcv_1.9-1            </span></span>
<span><span class="co">##  [49] httr_1.4.7             polyclip_1.10-6        abind_1.4-5           </span></span>
<span><span class="co">##  [52] compiler_4.3.3         withr_3.0.0            highr_0.10            </span></span>
<span><span class="co">##  [55] MASS_7.3-60            tools_4.3.3            lmtest_0.9-40         </span></span>
<span><span class="co">##  [58] httpuv_1.6.14          future.apply_1.11.1    goftest_1.2-3         </span></span>
<span><span class="co">##  [61] glue_1.7.0             dbscan_1.1-12          nlme_3.1-164          </span></span>
<span><span class="co">##  [64] promises_1.2.1         grid_4.3.3             Rtsne_0.17            </span></span>
<span><span class="co">##  [67] cluster_2.1.6          reshape2_1.4.4         generics_0.1.3        </span></span>
<span><span class="co">##  [70] gtable_0.3.4           spatstat.data_3.0-4    data.table_1.15.2     </span></span>
<span><span class="co">##  [73] sp_2.1-3               utf8_1.2.4             spatstat.geom_3.2-9   </span></span>
<span><span class="co">##  [76] RcppAnnoy_0.0.22       ggrepel_0.9.5          RANN_2.6.1            </span></span>
<span><span class="co">##  [79] pillar_1.9.0           stringr_1.5.1          spam_2.10-0           </span></span>
<span><span class="co">##  [82] later_1.3.2            splines_4.3.3          lattice_0.22-5        </span></span>
<span><span class="co">##  [85] survival_3.5-8         deldir_2.0-4           tidyselect_1.2.0      </span></span>
<span><span class="co">##  [88] miniUI_0.1.1.1         pbapply_1.7-2          knitr_1.45            </span></span>
<span><span class="co">##  [91] gridExtra_2.3          scattermore_1.2        xfun_0.42             </span></span>
<span><span class="co">##  [94] matrixStats_1.2.0      pheatmap_1.0.12        stringi_1.8.3         </span></span>
<span><span class="co">##  [97] lazyeval_0.2.2         yaml_2.3.8             evaluate_0.23         </span></span>
<span><span class="co">## [100] codetools_0.2-19       cli_3.6.2              uwot_0.1.16           </span></span>
<span><span class="co">## [103] xtable_1.8-4           reticulate_1.35.0      systemfonts_1.0.5     </span></span>
<span><span class="co">## [106] munsell_0.5.0          jquerylib_0.1.4        Rcpp_1.0.12           </span></span>
<span><span class="co">## [109] globals_0.16.3         spatstat.random_3.2-3  zeallot_0.1.0         </span></span>
<span><span class="co">## [112] png_0.1-8              parallel_4.3.3         ellipsis_0.3.2        </span></span>
<span><span class="co">## [115] pkgdown_2.0.7          dotCall64_1.1-1        listenv_0.9.1         </span></span>
<span><span class="co">## [118] viridisLite_0.4.2      scales_1.3.0           ggridges_0.5.6        </span></span>
<span><span class="co">## [121] leiden_0.4.3.1         purrr_1.0.2            scico_1.5.0           </span></span>
<span><span class="co">## [124] rlang_1.1.3            cowplot_1.1.3          shinyjs_2.1.0</span></span></code></pre>
</details><p><br></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://github.com/jemorlanes" class="external-link">Javier Escudero Morlanes</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
