<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="semla">
<title>Non-negative matrix factorization • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Non-negative matrix factorization">
<meta property="og:description" content="semla">
<meta property="og:image" content="https://ludvigla.github.io/semla/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/semla.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Intro</h6>
    <a class="dropdown-item" href="../articles/getting_started.html">Get started</a>
    <a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6>
    <a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a>
    <a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a>
    <a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a>
    <a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial methods</h6>
    <a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a>
    <a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a>
    <a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a>
    <a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Analysis</h6>
    <a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a>
    <a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a>
    <a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image processing</h6>
    <a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a>
    <a class="dropdown-item" href="../articles/mask_images.html">Image masking</a>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
  </div>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/create_object.html">Create a Staffli object</a>
    <a class="dropdown-item" href="../articles/compare_cell_type_mapping_NNLS.html">Cell type mapping comparison</a>
    <a class="dropdown-item" href="../articles/cell_type_mapping_benchmark.html">Cell type mapping benchmark</a>
    <a class="dropdown-item" href="../articles/slide-seq.html">Slide-Seq data</a>
    <a class="dropdown-item" href="../articles/IF_data.html">IF data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Non-negative matrix factorization</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 26 September
2023</h4>
      
      
      <div class="d-none name"><code>NNMF.Rmd</code></div>
    </div>

    
    
<p>Non-negative matrix factorization (NNMF or NMF) can be a useful
technique to deconvolve 10x Visium data, in particular when a reference
single-cell RNA-seq data is not available to conduct cell type
deconvolution. In this vignette we will demonstrate how you can apply
this method with <code>semla</code> on spatial data.</p>
<p>We recommend using the <a href="https://zdebruine.github.io/singlet/" class="external-link">singlet</a> R package for
NNMF which handles <code>Seurat</code> objects. <code>singlet</code>
uses the ultra fast NNMF implementation from <a href="https://zdebruine.github.io/RcppML/" class="external-link">RcppML</a>. Both of these
packages are developed by Zach DeBruines lab and you can find more
information about their work and tools at <a href="https://github.com/zdebruine" class="external-link uri">https://github.com/zdebruine</a> and <a href="https://www.zachdebruine.com/" class="external-link uri">https://www.zachdebruine.com/</a>.</p>
<p><code>RcppML</code> is available on <a href="https://cran.r-project.org/web/packages/RcppML/index.html" class="external-link">CRAN</a>
and <a href="https://github.com/zdebruine/RcppML" class="external-link">GitHub</a>, and
singlet can be installed from GitHub with
<code>devtools::install_github("zdebruine/singlet")</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ludvigla.github.io/semla/">semla</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zdebruine/singlet" class="external-link">singlet</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<div class="section level2">
<h2 id="run-nmf">Run NMF<a class="anchor" aria-label="anchor" href="#run-nmf"></a>
</h2>
<p>We’ll use a mouse brain Visium data set provided in
<code>SeuratData</code> for the NNMF. The method automatically runs
cross-validation to find the best rank and learns a model at that
rank.</p>
<p>The factorization can be computed on all genes, but below we have
filtered the data to include only the top variable features to speed up
the computation.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Here we'll use an example data set provided in SeuratData</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"stxBrain"</span>, type <span class="op">=</span> <span class="st">"anterior1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Update se_mbrain for compatibility with semla</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/UpdateSeuratForSemla.html">UpdateSeuratForSemla</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Normalize data and find top variable features</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="va">se_mbrain</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># OPTIONAL: subset data to improve computational speed</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="va">se_mbrain</span><span class="op">[</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/singlet/man/RunNMF.html" class="external-link">RunNMF</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">nmf</span><span class="op">@</span><span class="va">feature.loadings</span><span class="op">)</span></span></code></pre></div>
<p>We can plot the cross-validation results and find that the optimal
rank decided by the method is 22, which determines the number of factors
we obtain from the NMF run.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/singlet/man/RankPlot.html" class="external-link">RankPlot</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span></code></pre></div>
<p><img src="NNMF/rank_plot.jpg"></p>
<p><br></p>
<div class="section level3">
<h3 id="spatial-visualization">Spatial visualization<a class="anchor" aria-label="anchor" href="#spatial-visualization"></a>
</h3>
<p>The results are stored as a <code>DimReduc</code> object in our
<code>Seurat</code> object and we can map the factors spatially with
<code><a href="../reference/visualize-features.html">MapFeatures()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"NMF_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, </span>
<span>            override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            colors <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">magma</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">11</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="NNMF/spatial_plot.jpg"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="explore-gene-contributions">Explore gene contributions<a class="anchor" aria-label="anchor" href="#explore-gene-contributions"></a>
</h3>
<p>We can also investigate the gene loadings for each factor with
<code><a href="../reference/PlotFeatureLoadings.html">PlotFeatureLoadings()</a></code> which will give us an idea about what
the top contributing genes for each factor are. Below we can see the top
30 contributing genes for NMF_1 and NMF_2.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotFeatureLoadings.html">PlotFeatureLoadings</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>                    dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, </span>
<span>                    reduction <span class="op">=</span> <span class="st">"nmf"</span>, </span>
<span>                    nfeatures <span class="op">=</span> <span class="fl">30</span>,</span>
<span>                    mode <span class="op">=</span> <span class="st">"dotplot"</span>, </span>
<span>                    fill <span class="op">=</span> <span class="st">"darkmagenta"</span>,</span>
<span>                    pt_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="NNMF/gene_loadings.jpg"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="make-composite-plot">Make composite plot<a class="anchor" aria-label="anchor" href="#make-composite-plot"></a>
</h3>
<p>We can also map multiple factors spatially in a single plot with
<code><a href="../reference/visualize-multiple-features.html">MapMultipleFeatures()</a></code>. Note that the visualization will
only display the most dominant factor for each spot. See
<code>?MapMultipleFeatures()</code> for details.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span>
<span></span>
<span><span class="va">factor_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'#e6194b'</span>, <span class="st">'#3cb44b'</span>, <span class="st">'#ffe119'</span>, <span class="st">'#4363d8'</span>, <span class="st">'#f58231'</span>, </span>
<span>                   <span class="st">'#911eb4'</span>, <span class="st">'#46f0f0'</span>, <span class="st">'#f032e6'</span>, <span class="st">'#bcf60c'</span>, <span class="st">'#fabebe'</span>, </span>
<span>                   <span class="st">'#008080'</span>, <span class="st">'#e6beff'</span>, <span class="st">'#9a6324'</span>, <span class="st">'#fffac8'</span>, <span class="st">'#800000'</span>, <span class="st">'#aaffc3'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select non-overlapping factors</span></span>
<span><span class="va">selected_factors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">6</span>, <span class="fl">7</span>, <span class="fl">9</span>, <span class="fl">10</span>, <span class="fl">12</span>, <span class="fl">13</span>, <span class="fl">15</span>, <span class="fl">16</span>, <span class="fl">17</span>, <span class="fl">19</span>, <span class="fl">20</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/visualize-multiple-features.html">MapMultipleFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>                    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"NMF_"</span>, <span class="va">selected_factors</span><span class="op">)</span>, </span>
<span>                    colors <span class="op">=</span> <span class="va">factor_colors</span>, </span>
<span>                    image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                    override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                    pt_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="NNMF/multiple_factors.jpg"></p>
<p><br></p>
</div>
<div class="section level3">
<h3 id="explore-gene-contributions-heatmap">Explore gene contributions (heatmap)<a class="anchor" aria-label="anchor" href="#explore-gene-contributions-heatmap"></a>
</h3>
<p>Similarly, we can also summarize the top feature loadings for each
factor with a heatmap:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotFeatureLoadings.html">PlotFeatureLoadings</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>                    dims <span class="op">=</span> <span class="va">selected_factors</span>, </span>
<span>                    reduction <span class="op">=</span> <span class="st">"nmf"</span>, </span>
<span>                    nfeatures <span class="op">=</span> <span class="fl">30</span>, </span>
<span>                    mode <span class="op">=</span> <span class="st">"heatmap"</span>, </span>
<span>                    gradient_colors <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">magma</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">11</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="NNMF/multiple_factors_heatmap.jpg"></p>
</div>
</div>
<div class="section level2">
<h2 id="functional-enrichment-analysis-fea">Functional Enrichment Analysis (FEA)<a class="anchor" aria-label="anchor" href="#functional-enrichment-analysis-fea"></a>
</h2>
<p>If you want to get the feature loading values for each factor, you
can extract them from the <code>DimReduc</code> object. This can for
example be useful to run Functional Enrichment Analysis (FEA).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># fetch feature.loadings from DimReduc object</span></span>
<span><span class="va">nmf_loadings</span> <span class="op">&lt;-</span> <span class="va">se_mbrain</span><span class="op">[[</span><span class="st">"nmf"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">feature.loadings</span></span>
<span></span>
<span><span class="co"># Convert to long format and group data by factor</span></span>
<span><span class="va">gene_loadings_sorted</span> <span class="op">&lt;-</span> <span class="va">nmf_loadings</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span>var <span class="op">=</span> <span class="st">"gene"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">nmf_loadings</span><span class="op">)</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"fctr"</span>, values_to <span class="op">=</span> <span class="st">"loading"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>fctr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">fctr</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">nmf_loadings</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">fctr</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">fctr</span>, <span class="op">-</span><span class="va">loading</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Extract top 10 genes per factor</span></span>
<span><span class="va">gene_loadings_sorted</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 220 × 3</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Groups:   fctr [22]</span></span></span>
<span><span class="co">##    gene    fctr  loading</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> Penk    NMF_1 0.007<span style="text-decoration: underline;">57</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> Pcp4    NMF_1 0.007<span style="text-decoration: underline;">53</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> Ppp1r1b NMF_1 0.007<span style="text-decoration: underline;">21</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> Gpr88   NMF_1 0.006<span style="text-decoration: underline;">50</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> Hpca    NMF_1 0.006<span style="text-decoration: underline;">39</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> Ppp3ca  NMF_1 0.006<span style="text-decoration: underline;">08</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> Pde1b   NMF_1 0.006<span style="text-decoration: underline;">08</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> Arpp21  NMF_1 0.005<span style="text-decoration: underline;">98</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> Nrgn    NMF_1 0.005<span style="text-decoration: underline;">85</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> Pde10a  NMF_1 0.005<span style="text-decoration: underline;">82</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 210 more rows</span></span></span></code></pre>
<p>Run functional enrichment analysis on top 10 genes from factor 1:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gprofiler2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get gene sets</span></span>
<span><span class="va">gene_set_nmf_1</span> <span class="op">&lt;-</span> <span class="va">gene_loadings_sorted</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">fctr</span> <span class="op">==</span> <span class="st">"NMF_1"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Run FEA</span></span>
<span><span class="va">fea_results_nmf_1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gprofiler2/man/gost.html" class="external-link">gost</a></span><span class="op">(</span>query <span class="op">=</span> <span class="va">gene_set_nmf_1</span><span class="op">$</span><span class="va">gene</span>, ordered_query <span class="op">=</span> <span class="cn">TRUE</span>, organism <span class="op">=</span> <span class="st">"mmusculus"</span>, sources <span class="op">=</span> <span class="st">"GO:BP"</span><span class="op">)</span><span class="op">$</span><span class="va">result</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Look at results</span></span>
<span><span class="va">fea_results_nmf_1</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">p_value</span>, <span class="va">term_size</span>, <span class="va">query_size</span>, <span class="va">intersection_size</span>, <span class="va">term_name</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 23 × 5</span></span></span>
<span><span class="co">##      p_value term_size query_size intersection_size term_name                   </span></span>
<span><span class="co">##        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> 0.000<span style="text-decoration: underline;">082</span>5        36          7                 3 response to amphetamine     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> 0.000<span style="text-decoration: underline;">265</span>        250          7                 4 locomotory behavior         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> 0.000<span style="text-decoration: underline;">301</span>        188          9                 4 learning                    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> 0.000<span style="text-decoration: underline;">302</span>         55          7                 3 response to amine           </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> 0.000<span style="text-decoration: underline;">384</span>       <span style="text-decoration: underline;">1</span>329          8                 6 response to abiotic stimulus</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> 0.000<span style="text-decoration: underline;">885</span>        417          6                 4 response to salt            </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> 0.002<span style="text-decoration: underline;">14</span>         760          9                 5 behavior                    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> 0.002<span style="text-decoration: underline;">64</span>         324          9                 4 learning or memory          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> 0.003<span style="text-decoration: underline;">42</span>         475          7                 4 response to radiation       </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> 0.004<span style="text-decoration: underline;">01</span>         360          9                 4 cognition                   </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 13 more rows</span></span></span></code></pre>
<p><br></p>
<hr>
<details open><summary><strong>Package versions</strong>
</summary><ul>
<li><p><code>semla</code>: 1.1.6</p></li>
<li><p><code>RcppML</code>: 0.5.6</p></li>
<li><p><code>singlet</code>: 0.99.27</p></li>
</ul></details><details><summary><strong>Session info</strong>
</summary><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.3 (2023-03-15)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices datasets  utils     methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] stxBrain.SeuratData_0.1.1 SeuratData_0.2.2         </span></span>
<span><span class="co">## [3] singlet_0.99.27           RcppML_0.5.6             </span></span>
<span><span class="co">## [5] semla_1.1.6               ggplot2_3.4.3            </span></span>
<span><span class="co">## [7] dplyr_1.1.3               SeuratObject_4.1.3       </span></span>
<span><span class="co">## [9] Seurat_4.3.0.1           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fastmatch_1.1-4        systemfonts_1.0.4      plyr_1.8.8            </span></span>
<span><span class="co">##   [4] igraph_1.5.1           lazyeval_0.2.2         sp_2.0-0              </span></span>
<span><span class="co">##   [7] splines_4.2.3          BiocParallel_1.32.6    listenv_0.9.0         </span></span>
<span><span class="co">##  [10] scattermore_1.2        digest_0.6.33          htmltools_0.5.6       </span></span>
<span><span class="co">##  [13] magick_2.7.4           fansi_1.0.4            magrittr_2.0.3        </span></span>
<span><span class="co">##  [16] memoise_2.0.1          tensor_1.5             cluster_2.1.4         </span></span>
<span><span class="co">##  [19] ROCR_1.0-11            limma_3.54.2           globals_0.16.2        </span></span>
<span><span class="co">##  [22] matrixStats_1.0.0      pkgdown_2.0.7          spatstat.sparse_3.0-2 </span></span>
<span><span class="co">##  [25] colorspace_2.1-0       rappdirs_0.3.3         ggrepel_0.9.3         </span></span>
<span><span class="co">##  [28] textshaping_0.3.6      xfun_0.40              crayon_1.5.2          </span></span>
<span><span class="co">##  [31] jsonlite_1.8.7         progressr_0.14.0       spatstat.data_3.0-1   </span></span>
<span><span class="co">##  [34] zeallot_0.1.0          survival_3.5-7         zoo_1.8-12            </span></span>
<span><span class="co">##  [37] glue_1.6.2             polyclip_1.10-4        gtable_0.3.4          </span></span>
<span><span class="co">##  [40] leiden_0.4.3           future.apply_1.11.0    msigdbr_7.5.1         </span></span>
<span><span class="co">##  [43] abind_1.4-5            scales_1.2.1           spatstat.random_3.1-5 </span></span>
<span><span class="co">##  [46] miniUI_0.1.1.1         Rcpp_1.0.11            viridisLite_0.4.2     </span></span>
<span><span class="co">##  [49] xtable_1.8-4           reticulate_1.31        htmlwidgets_1.6.2     </span></span>
<span><span class="co">##  [52] httr_1.4.7             fgsea_1.24.0           RColorBrewer_1.1-3    </span></span>
<span><span class="co">##  [55] ellipsis_0.3.2         ica_1.0-3              pkgconfig_2.0.3       </span></span>
<span><span class="co">##  [58] sass_0.4.7             uwot_0.1.16            deldir_1.0-9          </span></span>
<span><span class="co">##  [61] utf8_1.2.3             tidyselect_1.2.0       rlang_1.1.1           </span></span>
<span><span class="co">##  [64] reshape2_1.4.4         later_1.3.1            munsell_0.5.0         </span></span>
<span><span class="co">##  [67] tools_4.2.3            cachem_1.0.8           cli_3.6.1             </span></span>
<span><span class="co">##  [70] dbscan_1.1-11          generics_0.1.3         ggridges_0.5.4        </span></span>
<span><span class="co">##  [73] evaluate_0.21          stringr_1.5.0          fastmap_1.1.1         </span></span>
<span><span class="co">##  [76] yaml_2.3.7             ragg_1.2.5             goftest_1.2-3         </span></span>
<span><span class="co">##  [79] babelgene_22.9         knitr_1.43             fs_1.6.3              </span></span>
<span><span class="co">##  [82] fitdistrplus_1.1-11    purrr_1.0.2            RANN_2.6.1            </span></span>
<span><span class="co">##  [85] pbapply_1.7-2          future_1.33.0          nlme_3.1-163          </span></span>
<span><span class="co">##  [88] mime_0.12              compiler_4.2.3         plotly_4.10.2         </span></span>
<span><span class="co">##  [91] png_0.1-8              spatstat.utils_3.0-3   tibble_3.2.1          </span></span>
<span><span class="co">##  [94] bslib_0.5.1            stringi_1.7.12         desc_1.4.2            </span></span>
<span><span class="co">##  [97] forcats_1.0.0          lattice_0.21-8         Matrix_1.5-1          </span></span>
<span><span class="co">## [100] shinyjs_2.1.0          vctrs_0.6.3            pillar_1.9.0          </span></span>
<span><span class="co">## [103] lifecycle_1.0.3        BiocManager_1.30.22    spatstat.geom_3.2-4   </span></span>
<span><span class="co">## [106] lmtest_0.9-40          jquerylib_0.1.4        RcppAnnoy_0.0.21      </span></span>
<span><span class="co">## [109] data.table_1.14.8      cowplot_1.1.1          irlba_2.3.5.1         </span></span>
<span><span class="co">## [112] httpuv_1.6.11          patchwork_1.1.3        R6_2.5.1              </span></span>
<span><span class="co">## [115] promises_1.2.1         renv_1.0.2             KernSmooth_2.23-22    </span></span>
<span><span class="co">## [118] gridExtra_2.3          parallelly_1.36.0      codetools_0.2-19      </span></span>
<span><span class="co">## [121] MASS_7.3-60            rprojroot_2.0.3        withr_2.5.0           </span></span>
<span><span class="co">## [124] sctransform_0.3.5      parallel_4.2.3         grid_4.2.3            </span></span>
<span><span class="co">## [127] tidyr_1.3.0            rmarkdown_2.24         Rtsne_0.16            </span></span>
<span><span class="co">## [130] spatstat.explore_3.2-1 shiny_1.7.5</span></span></code></pre>
</details><p><br></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
