<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="semla">
<title>Non-negative matrix factorization â€¢ semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Non-negative matrix factorization">
<meta property="og:description" content="semla">
<meta property="og:image" content="https://ludvigla.github.io/semla/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-inverse navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/semla.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Intro</h6>
    <a class="dropdown-item" href="../articles/getting_started.html">Get started</a>
    <a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive</h6>
    <a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6>
    <a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a>
    <a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a>
    <a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial methods</h6>
    <a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a>
    <a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a>
    <a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a>
    <a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Analysis</h6>
    <a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a>
    <a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a>
    <a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image processing</h6>
    <a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Non-negative matrix factorization</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 13 February
2023</h4>
      
      
      <div class="d-none name"><code>NNMF.Rmd</code></div>
    </div>

    
    
<p>Non-negative matrix factorization (NNMF or NMF) can be a useful
technique to deconvolve 10x Visium data, in particular when a reference
single-cell RNA-seq data is not available to conduct cell type
deconvolution. In this vignette we will demonstrate how you can apply
this method with <code>semla</code> on spatial data.</p>
<p>We recommend using the <a href="https://zdebruine.github.io/singlet/" class="external-link">singlet</a> R package for
NNMF which handles <code>Seurat</code> objects. <code>singlet</code>
uses the ultra fast NNMF implementation from <a href="https://zdebruine.github.io/RcppML/" class="external-link">RcppML</a>. Both of these
packages are developed by Zach DeBruines lab and you can find more
information about their work and tools at <a href="https://github.com/zdebruine" class="external-link uri">https://github.com/zdebruine</a> and at <a href="https://www.zachdebruine.com/" class="external-link uri">https://www.zachdebruine.com/</a>.</p>
<p><code>RcppML</code> is available on <a href="https://cran.r-project.org/web/packages/RcppML/index.html" class="external-link">CRAN</a>
and <a href="https://github.com/zdebruine/RcppML" class="external-link">GitHub</a>, and
singlet can be installed from GitHub with
<code>devtools::install_github("zdebruine/singlet")</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ludvigla.github.io/semla/">semla</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zdebruine/singlet" class="external-link">singlet</a></span><span class="op">)</span></span></code></pre></div>
<p><br></p>
<p>Weâ€™ll use our mouse brain test data for the NNMF. The method
automatically runs cross-validation to find the best rank and to learn a
model at that rank.</p>
<p>NB: The mouse brain test data only includes a subset of the entire
data of 188 unique genes. With the full dataset, we would be able to
find a higher optimal rank and therefore obtain more factors.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousebrain"</span>, </span>
<span>                                 <span class="st">"se_mbrain"</span>, </span>
<span>                                 package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/singlet/man/RunNMF.html" class="external-link">RunNMF</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## running with sparse optimization</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## REPLICATE  1 / 3 </span></span>
<span><span class="co">## k = 2 , rep = 1 </span></span>
<span><span class="co">##    test_error = 1.4626e+00 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 2 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 3 , rep = 1 </span></span>
<span><span class="co">##    test_error = 1.1130e+00 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 3 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 6 , rep = 1 </span></span>
<span><span class="co">##    test_error = 1.0116e+00 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 11 , rep = 1 </span></span>
<span><span class="co">##    test_error = 9.6053e-01 </span></span>
<span><span class="co">##    overfit (overfit_score = 2.4498e-04 )</span></span>
<span><span class="co">##    best rank in replicate = 6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 8 , rep = 1 </span></span>
<span><span class="co">##    test_error = 9.7487e-01 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 8 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 9 , rep = 1 </span></span>
<span><span class="co">##    test_error = 9.6914e-01 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 9 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 10 , rep = 1 </span></span>
<span><span class="co">##    test_error = 9.6743e-01 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 10 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## REPLICATE  2 / 3 </span></span>
<span><span class="co">## k = 2 , rep = 2 </span></span>
<span><span class="co">##    test_error = 1.4324e+00 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 2 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 3 , rep = 2 </span></span>
<span><span class="co">##    test_error = 1.1140e+00 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 3 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 6 , rep = 2 </span></span>
<span><span class="co">##    test_error = 9.9703e-01 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 11 , rep = 2 </span></span>
<span><span class="co">##    test_error = 9.5834e-01 </span></span>
<span><span class="co">##    overfit (overfit_score = 1.5347e-03 )</span></span>
<span><span class="co">##    best rank in replicate = 6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 8 , rep = 2 </span></span>
<span><span class="co">##    test_error = 9.6048e-01 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 8 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 9 , rep = 2 </span></span>
<span><span class="co">##    test_error = 9.6746e-01 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 8 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 7 , rep = 2 </span></span>
<span><span class="co">##    test_error = 9.7535e-01 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 8 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## REPLICATE  3 / 3 </span></span>
<span><span class="co">## k = 2 , rep = 3 </span></span>
<span><span class="co">##    test_error = 1.4391e+00 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 2 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 3 , rep = 3 </span></span>
<span><span class="co">##    test_error = 1.1060e+00 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 3 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 6 , rep = 3 </span></span>
<span><span class="co">##    test_error = 1.0093e+00 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 11 , rep = 3 </span></span>
<span><span class="co">##    test_error = 9.7064e-01 </span></span>
<span><span class="co">##    overfit (overfit_score = 2.9483e-03 )</span></span>
<span><span class="co">##    best rank in replicate = 6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 8 , rep = 3 </span></span>
<span><span class="co">##    test_error = 9.9007e-01 </span></span>
<span><span class="co">##    overfit (overfit_score = 5.3016e-04 )</span></span>
<span><span class="co">##    best rank in replicate = 6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 5 , rep = 3 </span></span>
<span><span class="co">##    test_error = 1.0257e+00 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 6 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## k = 7 , rep = 3 </span></span>
<span><span class="co">##    test_error = 9.8727e-01 </span></span>
<span><span class="co">##    not overfit</span></span>
<span><span class="co">##    best rank in replicate = 7 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Unmasking test set</span></span>
<span><span class="co">## Fitting final model at k = 11</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">nmf</span><span class="op">@</span><span class="va">feature.loadings</span><span class="op">)</span></span></code></pre></div>
<p>We can plot the cross-validation results and find that the optimal
rank decided by the method is 11 which determines the number of factors
we obtain from the NMF run.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/singlet/man/RankPlot.html" class="external-link">RankPlot</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span></code></pre></div>
<p><img src="NNMF_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p><br></p>
<p>The results are stored as a <code>DimReduc</code> object in our
<code>Seurat</code> object and we can map the factors spatially with
<code><a href="../reference/visualize-features.html">MapFeatures()</a></code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"NMF_"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">k</span><span class="op">)</span>, </span>
<span>            override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            colors <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">magma</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">11</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="NNMF_files/figure-html/unnamed-chunk-6-1.png" width="864"></p>
<p><br></p>
<p>We can also investigate the gene loadings for each factor with
<code><a href="../reference/PlotFeatureLoadings.html">PlotFeatureLoadings()</a></code> which will give us an idea about what
the top contributing genes for each factor are. Below we can see the top
30 contributing genes for NMF_1 and NMF_2.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotFeatureLoadings.html">PlotFeatureLoadings</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>                    dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, </span>
<span>                    reduction <span class="op">=</span> <span class="st">"nmf"</span>, </span>
<span>                    nfeatures <span class="op">=</span> <span class="fl">30</span>,</span>
<span>                    mode <span class="op">=</span> <span class="st">"dotplot"</span>, </span>
<span>                    fill <span class="op">=</span> <span class="st">"darkmagenta"</span>,</span>
<span>                    pt_size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="NNMF_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p><br></p>
<p>We can also map multiple factors spatially in a single plot with
<code><a href="../reference/visualize-multiple-features.html">MapMultipleFeatures()</a></code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## â”€â”€ <span style="font-weight: bold;">Loading H&amp;E images</span> â”€â”€</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">â„¹</span> Loading image from /Users/ludviglarsson/miniconda3/envs/R4.2/lib/R/library/semla/extdata/mousebrain/spatial/tissue_hires_image.jpg</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">â„¹</span> Scaled image from 2000x1882 to 400x376 pixels</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">â„¹</span> Saving loaded H&amp;E images as 'rasters' in Seurat object</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">factor_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#332288"</span>, <span class="st">"#88CCEE"</span>, <span class="st">"#44AA99"</span>, <span class="st">"#117733"</span>, </span>
<span>                   <span class="st">"#999933"</span>, <span class="st">"#DDCC77"</span>, <span class="st">"#CC6677"</span>, <span class="st">"#882255"</span>, <span class="st">"#AA4499"</span><span class="op">)</span></span>
<span><span class="va">n_k_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">k</span><span class="op">&lt;</span><span class="fl">9</span>, <span class="va">k</span>, <span class="fl">9</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/visualize-multiple-features.html">MapMultipleFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>                    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"NMF_"</span>, <span class="fl">1</span><span class="op">:</span><span class="va">n_k_plot</span><span class="op">)</span>, </span>
<span>                    colors <span class="op">=</span> <span class="va">factor_colors</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">n_k_plot</span><span class="op">]</span>, </span>
<span>                    image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                    override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                    pt_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required namespace: ggnewscale</span></span></code></pre>
<p><img src="NNMF_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p><br></p>
<p>Similarly, we can also summarize the top feature loadings for each
factor with a heatmap:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotFeatureLoadings.html">PlotFeatureLoadings</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>                    dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">k</span>, </span>
<span>                    reduction <span class="op">=</span> <span class="st">"nmf"</span>, </span>
<span>                    nfeatures <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                    mode <span class="op">=</span> <span class="st">"heatmap"</span>, </span>
<span>                    gradient_colors <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">magma</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">11</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="NNMF_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>NB: Although we are able to pick up distinct patterns in our data
with the factor analysis, our mouse brain dataset only contain a
fraction of the full dataset which partly explains why the spatial maps
and gene loadings for some factors are noisy.</p>
<p><br></p>
<hr>
<details open><summary><strong>Package versions</strong>
</summary><ul>
<li><p><code>semla</code>: 0.1.0</p></li>
<li><p><code>RcppML</code>: 0.3.7</p></li>
<li><p><code>singlet</code>: 0.99.1023</p></li>
</ul></details><details><summary><strong>Session info</strong>
</summary><div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin13.4.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /Users/ludviglarsson/miniconda3/envs/R4.2/lib/libopenblasp-r0.3.21.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] singlet_0.99.1023  semla_0.1.0        ggplot2_3.3.6      dplyr_1.0.9       </span></span>
<span><span class="co">## [5] SeuratObject_4.1.3 Seurat_4.3.0      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fastmatch_1.1-3        systemfonts_1.0.4      plyr_1.8.7            </span></span>
<span><span class="co">##   [4] igraph_1.3.4           lazyeval_0.2.2         sp_1.5-1              </span></span>
<span><span class="co">##   [7] splines_4.2.1          BiocParallel_1.30.3    listenv_0.8.0         </span></span>
<span><span class="co">##  [10] scattermore_0.8        digest_0.6.29          htmltools_0.5.4       </span></span>
<span><span class="co">##  [13] viridis_0.6.2          magick_2.7.3           fansi_1.0.3           </span></span>
<span><span class="co">##  [16] magrittr_2.0.3         memoise_2.0.1          tensor_1.5            </span></span>
<span><span class="co">##  [19] cluster_2.1.4          ROCR_1.0-11            limma_3.52.3          </span></span>
<span><span class="co">##  [22] globals_0.16.0         matrixStats_0.62.0     RcppML_0.3.7          </span></span>
<span><span class="co">##  [25] pkgdown_2.0.6          spatstat.sparse_3.0-0  colorspace_2.0-3      </span></span>
<span><span class="co">##  [28] ggrepel_0.9.1          textshaping_0.3.6      xfun_0.32             </span></span>
<span><span class="co">##  [31] jsonlite_1.8.3         progressr_0.10.1       spatstat.data_3.0-0   </span></span>
<span><span class="co">##  [34] zeallot_0.1.0          survival_3.4-0         zoo_1.8-10            </span></span>
<span><span class="co">##  [37] glue_1.6.2             polyclip_1.10-0        gtable_0.3.0          </span></span>
<span><span class="co">##  [40] leiden_0.4.2           future.apply_1.9.0     msigdbr_7.5.1         </span></span>
<span><span class="co">##  [43] abind_1.4-5            scales_1.2.1           DBI_1.1.3             </span></span>
<span><span class="co">##  [46] spatstat.random_3.0-1  miniUI_0.1.1.1         Rcpp_1.0.9            </span></span>
<span><span class="co">##  [49] viridisLite_0.4.1      xtable_1.8-4           reticulate_1.26       </span></span>
<span><span class="co">##  [52] htmlwidgets_1.5.4      httr_1.4.4             fgsea_1.22.0          </span></span>
<span><span class="co">##  [55] RColorBrewer_1.1-3     ellipsis_0.3.2         ica_1.0-3             </span></span>
<span><span class="co">##  [58] farver_2.1.1           pkgconfig_2.0.3        sass_0.4.2            </span></span>
<span><span class="co">##  [61] uwot_0.1.14            deldir_1.0-6           utf8_1.2.2            </span></span>
<span><span class="co">##  [64] labeling_0.4.2         tidyselect_1.1.2       rlang_1.0.6           </span></span>
<span><span class="co">##  [67] reshape2_1.4.4         later_1.3.0            munsell_0.5.0         </span></span>
<span><span class="co">##  [70] tools_4.2.1            cachem_1.0.6           cli_3.4.1             </span></span>
<span><span class="co">##  [73] dbscan_1.1-10          generics_0.1.3         ggridges_0.5.3        </span></span>
<span><span class="co">##  [76] evaluate_0.16          stringr_1.4.1          fastmap_1.1.0         </span></span>
<span><span class="co">##  [79] yaml_2.3.5             ragg_1.2.2             goftest_1.2-3         </span></span>
<span><span class="co">##  [82] babelgene_22.9         knitr_1.39             fs_1.5.2              </span></span>
<span><span class="co">##  [85] fitdistrplus_1.1-8     purrr_0.3.4            RANN_2.6.1            </span></span>
<span><span class="co">##  [88] pbapply_1.5-0          future_1.27.0          nlme_3.1-159          </span></span>
<span><span class="co">##  [91] mime_0.12              compiler_4.2.1         rstudioapi_0.14       </span></span>
<span><span class="co">##  [94] plotly_4.10.0          png_0.1-7              spatstat.utils_3.0-1  </span></span>
<span><span class="co">##  [97] tibble_3.1.8           bslib_0.4.0            stringi_1.7.8         </span></span>
<span><span class="co">## [100] highr_0.9              desc_1.4.1             forcats_0.5.2         </span></span>
<span><span class="co">## [103] lattice_0.20-45        Matrix_1.5-3           shinyjs_2.1.0         </span></span>
<span><span class="co">## [106] vctrs_0.5.1            pillar_1.8.1           lifecycle_1.0.3       </span></span>
<span><span class="co">## [109] spatstat.geom_3.0-3    lmtest_0.9-40          jquerylib_0.1.4       </span></span>
<span><span class="co">## [112] RcppAnnoy_0.0.19       data.table_1.14.2      cowplot_1.1.1         </span></span>
<span><span class="co">## [115] irlba_2.3.5            httpuv_1.6.5           patchwork_1.1.2       </span></span>
<span><span class="co">## [118] R6_2.5.1               promises_1.2.0.1       KernSmooth_2.23-20    </span></span>
<span><span class="co">## [121] gridExtra_2.3          parallelly_1.32.1      codetools_0.2-18      </span></span>
<span><span class="co">## [124] MASS_7.3-58.1          assertthat_0.2.1       rprojroot_2.0.3       </span></span>
<span><span class="co">## [127] withr_2.5.0            sctransform_0.3.5      parallel_4.2.1        </span></span>
<span><span class="co">## [130] grid_4.2.1             tidyr_1.2.0            rmarkdown_2.15        </span></span>
<span><span class="co">## [133] Rtsne_0.16             ggnewscale_0.4.8       spatstat.explore_3.0-5</span></span>
<span><span class="co">## [136] shiny_1.7.4</span></span></code></pre>
</details><p><br></p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
