<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Visualization of numeric features • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Visualization of numeric features">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/semla.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><h6 class="dropdown-header" data-toc-skip>Intro</h6></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a></li>
    <li><a class="dropdown-item" href="../articles/visiumHD.html">VisiumHD</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a></li>
    <li><a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a></li>
    <li><a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a></li>
    <li><a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial methods</h6></li>
    <li><a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a></li>
    <li><a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a></li>
    <li><a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a></li>
    <li><a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Image processing</h6></li>
    <li><a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a></li>
    <li><a class="dropdown-item" href="../articles/mask_images.html">Image masking</a></li>
    <li><a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/create_object.html">Create a Staffli object</a></li>
    <li><a class="dropdown-item" href="../articles/compare_cell_type_mapping_NNLS.html">Cell type mapping comparison</a></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_benchmark.html">Cell type mapping benchmark</a></li>
    <li><a class="dropdown-item" href="../articles/slide-seq.html">Slide-Seq data</a></li>
    <li><a class="dropdown-item" href="../articles/IF_data.html">IF data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Visualization of numeric features</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 08 September
2025</h4>
      

      <div class="d-none name"><code>numeric_features.Rmd</code></div>
    </div>

    
    
<p>In the visualization tutorials, we’ll have a look at different ways
of creating spatial plots with <code>semla</code>.</p>
<p>Functions such as <code><a href="../reference/visualize-features.html">MapFeatures()</a></code> and
<code><a href="../reference/visualize-labels.html">MapLabels()</a></code> produce patchworks (see R package <a href="https://patchwork.data-imaginist.com/" class="external-link">patchwork</a>) which are
easy to manipulate after they have been created. The
<code>patchwork</code> R package is extremely versatile and makes it
easy to customize your figures!</p>
<p>For those who are familiar with <code>Seurat</code>, these functions
are similar to <code><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialFeaturePlot()</a></code> and
<code><a href="https://satijalab.org/seurat/reference/SpatialPlot.html" class="external-link">SpatialDimPlot()</a></code> in the sense that the first can be used to
visualize numeric data and the latter can be used to color data points
based on categorical data.</p>
<p>If you are interested in more advanced features - including details
about how to use the patchwork system and other visualization methods -
you can skip directly to the <a href="https://ludvigla.github.io/semla/articles/advanced_visualization.html">‘Advanced
visualization’ tutorial</a>.</p>
<p>Here we’ll have a look at basic usage of the
<code><a href="../reference/visualize-features.html">MapFeatures()</a></code> function.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ludvigla.github.io/semla/">semla</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>First we need to load some 10x Visium data. Here we’ll use a mouse
brain tissue dataset and a mouse colon dataset that are shipped with
<code>semla</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load data</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, </span>
<span>                                        <span class="st">"mousebrain/se_mbrain"</span>, </span>
<span>                                        package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_mbrain</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="st">"mousebrain"</span></span>
<span><span class="va">se_mcolon</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, </span>
<span>                                        <span class="st">"mousecolon/se_mcolon"</span>, </span>
<span>                                        package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_mcolon</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">&lt;-</span> <span class="st">"mousecolon"</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge.html">MergeSTData</a></span><span class="op">(</span><span class="va">se_mbrain</span>, <span class="va">se_mcolon</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="map-numeric-features">Map numeric features<a class="anchor" aria-label="anchor" href="#map-numeric-features"></a>
</h2>
<p>Let’s get started with <code><a href="../reference/visualize-features.html">MapFeatures()</a></code>. The most basic
usage is to map gene expression spatially:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">rocket</a></span><span class="op">(</span><span class="fl">11</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>                 features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Th"</span>, <span class="st">"Clu"</span><span class="op">)</span>, </span>
<span>                 colors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="numeric_features_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>We can also plot other numeric features present within the
<code>meta.data</code> slot, such as the quality metrics
‘nCount_Spatial’ and ‘nFeature_Spatial’. To view the distribution of the
selected feature per sample, the function MapFeaturesSummary() can be
used to add a subplot next to the spatial plot (choose from histogram,
box, violin, or density plot).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MapFeaturesSummary.html">MapFeaturesSummary</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>                        features <span class="op">=</span> <span class="st">"nFeature_Spatial"</span>, </span>
<span>                        subplot_type <span class="op">=</span> <span class="st">"violin"</span>,</span>
<span>                        colors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="numeric_features_files/figure-html/unnamed-chunk-4-1.png" width="864"></p>
<p>If the distribution of your gene or feature is very skewed, it can
also ease the visualization to modify the <code>max_cutoff</code> and/or
<code>min_cutoff</code> arguments, where you provide a number between
0-1, corresponding the k-th percentile of the feature’s data range. By
doing this, it can help you create more distinct visualizations, where
spots with the more extreme values - see example below, where the plot
to the right has been capped by minimum and maximum percentiles:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>                  section_number <span class="op">=</span> <span class="fl">1</span>, <span class="co"># Show only the first sample</span></span>
<span>                  features <span class="op">=</span> <span class="st">"Ptgds"</span>, </span>
<span>                  max_cutoff <span class="op">=</span> <span class="fl">1</span>, <span class="co"># Full data</span></span>
<span>                  colors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>                  section_number <span class="op">=</span> <span class="fl">1</span>, <span class="co"># Show only the first sample</span></span>
<span>                  features <span class="op">=</span> <span class="st">"Ptgds"</span>, </span>
<span>                  max_cutoff <span class="op">=</span> <span class="fl">0.95</span>, <span class="co"># Max value set to the 95th percentile</span></span>
<span>                  min_cutoff <span class="op">=</span> <span class="fl">0.05</span>, <span class="co"># Min value set to the 5th percentile</span></span>
<span>                  colors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">|</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="numeric_features_files/figure-html/unnamed-chunk-5-1.png" width="864"></p>
</div>
<div class="section level2">
<h2 id="map-dimensionality-reduction-vectors">Map dimensionality reduction vectors<a class="anchor" aria-label="anchor" href="#map-dimensionality-reduction-vectors"></a>
</h2>
<p><code>MapFeatures</code> handles any type of numeric features which
can be fetched using the <code>Seurat</code> function
<code><a href="https://satijalab.github.io/seurat-object/reference/FetchData.html" class="external-link">FetchData()</a></code>. This includes latent vectors from
dimensionality reduction methods.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">## Warning: The `slot` argument of `SetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">## Centering and scaling data matrix</span></span></code></pre>
<pre><code><span><span class="co">## PC_ 1 </span></span>
<span><span class="co">## Positive:  Ptgds, Snap25, Rtn1, Cck, Nrgn, Eef1a2, Snca, Olfm1, Uchl1, Mobp </span></span>
<span><span class="co">##     Mbp, Plp1, Stmn2, Fabp5, Hpca, Pvalb, Nnat, Nptxr, Cpne6, Npy </span></span>
<span><span class="co">##     Clu, Mag, Pcp4, Apod, Scg2, Crym, Hbb-bs, Mal, Arc, Mog </span></span>
<span><span class="co">## Negative:  Car1, Cyp2c55, Cnn1, Aqp8, Dsp, Emp1, 1810065E05Rik, Tgm3, Acta2, Cd24a </span></span>
<span><span class="co">##     Hmgcs2, Tagln, Fabp2, Col1a2, Myh11, H2-Aa, Actg2, S100g, Car4, Pyy </span></span>
<span><span class="co">##     Ces2e, Apol10a, Mgat4c, Prdx6, Slc37a2, Prkcd, Ang4, Cd74, Atp12a, Myl9 </span></span>
<span><span class="co">## PC_ 2 </span></span>
<span><span class="co">## Positive:  Slc6a3, Th, Mog, Chrna6, Apod, Mag, Drd2, Opalin, En1, Mal </span></span>
<span><span class="co">##     Lhx1os, Aldh1a1, Slc18a2, Slc10a4, Mobp, Cnp, Dlk1, Plp1, Ret, Sncg </span></span>
<span><span class="co">##     Spp1, Mbp, Ddc, Ptgds, Tcf7l2, Calb2, Pitx2, Tnnt1, Pvalb, Slc13a4 </span></span>
<span><span class="co">## Negative:  Nov, Lamp5, Egr1, Nr4a1, Crym, Neurod6, Arc, Nptxr, Myh7, Coro1a </span></span>
<span><span class="co">##     Sst, Cort, Npy, Tmsb4x, Spink8, Cpne6, Grp, Mfge8, Nrgn, Rprm </span></span>
<span><span class="co">##     Fibcd1, Myl4, Hpca, Chgb, Vip, Snca, Cck, Tac2, Olfm1, Trh </span></span>
<span><span class="co">## PC_ 3 </span></span>
<span><span class="co">## Positive:  Slc18a2, En1, Th, Slc10a4, Slc6a3, Chrna6, Dlk1, Sncg, Drd2, Ret </span></span>
<span><span class="co">##     Ddc, Calb2, Cpne7, Hpcal1, Dcn, Scg2, Nnat, Vip, Coro1a, Prph </span></span>
<span><span class="co">##     Stmn2, Nos1, Mfge8, Myh7, Ecel1, Tmsb4x, Htr3a, Trh, Snca, Tac1 </span></span>
<span><span class="co">## Negative:  Myoc, Hbb-bt, Hba-a1, Mog, Opalin, Gfap, Slc13a4, Mag, Mal, Mobp </span></span>
<span><span class="co">##     Apod, Pvalb, Ptgds, C1ql2, Tnnt1, Reg3b, Plp1, Lhx1os, Tcf7l2, Hba-a2 </span></span>
<span><span class="co">##     Mbp, Iapp, Fabp2, Hbb-bs, Fmod, Defb37, Myl4, Car1, Cyp2c55, Spp1 </span></span>
<span><span class="co">## PC_ 4 </span></span>
<span><span class="co">## Positive:  H2-DMb2, Ly6d, Ighd, Il22ra2, Cd52, Cd79b, Cd79a, Ighm, Clu, C3 </span></span>
<span><span class="co">##     Spib, Mfge8, Lyz2, Ccl20, Ubd, Cnp, Iglc2, Hba-a2, Ly6g, Ighg2b </span></span>
<span><span class="co">##     Myoc, H2-Eb1, Hbb-bs, Hbb-bt, Hba-a1, Apod, Slc13a4, Cd74, Vtn, Fabp5 </span></span>
<span><span class="co">## Negative:  En1, Cyp4b1, Ces1g, Chrna6, Slc10a2, Defb37, Th, Sct, Slc18a2, Slc10a4 </span></span>
<span><span class="co">##     Slc6a3, Slc51a, Nov, Iapp, Drd2, Tmigd1, Fabp2, Dlk1, Reg3b, Myh7 </span></span>
<span><span class="co">##     Cpne7, 1810065E05Rik, Cyp2d26, Apol10a, Cyp2c55, Emp1, Spink8, Nts, Mgat4c, S100g </span></span>
<span><span class="co">## PC_ 5 </span></span>
<span><span class="co">## Positive:  Trbc2, Myl4, Arc, Egr1, Nr4a1, Ighm, Coro1a, Lamp5, Ly6d, Cplx3 </span></span>
<span><span class="co">##     Ighd, Th, En1, H2-DMb2, Neurod6, Opalin, Cd79b, Chrna6, Cd79a, Il22ra2 </span></span>
<span><span class="co">##     Mobp, Mog, Spib, Drd2, Pvalb, Cnp, Mal, Mag, Cd52, Slc6a3 </span></span>
<span><span class="co">## Negative:  Myoc, Slc13a4, Fmod, Dcn, Gfap, Hba-a2, Hbb-bt, Hba-a1, Vtn, Trh </span></span>
<span><span class="co">##     Ecel1, Hbb-bs, Slc6a20a, Nnat, Mgp, Nos1, Spp1, Fibcd1, Htr3a, Vgll3 </span></span>
<span><span class="co">##     Cpne7, Calb2, Hpcal1, C1ql2, Spink8, Crym, Myl9, Lypd1, Rcn1, Cartpt</span></span></code></pre>
<p>When plotting numeric features that are centered at 0, it is more
appropriate to also center the color scale and select a ‘divergent’
color palette.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>            features <span class="op">=</span> <span class="st">"PC_2"</span>, </span>
<span>            center_zero <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            section_number <span class="op">=</span> <span class="fl">1</span>, </span>
<span>            pt_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>            colors <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">11</span>, name <span class="op">=</span> <span class="st">"RdBu"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="numeric_features_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="overlay-maps-on-images">Overlay maps on images<a class="anchor" aria-label="anchor" href="#overlay-maps-on-images"></a>
</h2>
<p>If we want to create a map with the H&amp;E images we can do this by
setting <code>image_use = raw</code>. But before we can do this, we need
to load the images into our <em>Seurat</em> object:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="va">se</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span><span class="fl">11</span>, <span class="st">"Spectral"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>                 features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Th"</span>, <span class="st">"Clu"</span><span class="op">)</span>, </span>
<span>                 image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                 colors <span class="op">=</span> <span class="va">cols</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="numeric_features_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Right now it’s quite difficult to see the tissue underneath the
spots. We can add some opacity to the colors which is scaled by the
feature values to make spots with low expression transparent:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>                 features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Th"</span>, <span class="st">"Clu"</span><span class="op">)</span>, </span>
<span>                 image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                 colors <span class="op">=</span> <span class="va">cols</span>, </span>
<span>                 scale_alpha <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="numeric_features_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="crop-image">Crop image<a class="anchor" aria-label="anchor" href="#crop-image"></a>
</h2>
<p>We can crop the images manually by defining a <code>crop_area</code>.
The <code>crop_area</code> should be a vector of length four defining
the opposite corners of a rectangle, where the x- and y-axes are defined
from 0-1.</p>
<p>In order to more easily see how this rectangle could be defined we
can get some help by adding a grid to the plot.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>                 section_number <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                 features <span class="op">=</span> <span class="st">"Clu"</span>, </span>
<span>                 image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                 color <span class="op">=</span> <span class="va">cols</span>, </span>
<span>                 pt_alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x<span class="op">=</span><span class="st">"x-axis"</span>, y<span class="op">=</span><span class="st">"y-axis"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>panel.grid.major <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_line</a></span><span class="op">(</span>linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span>, </span>
<span>        axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>        axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="numeric_features_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>The Clu gene expression reveals where we have an area of lymphoid
tissue in the colon sample. Now if we want to crop out the GALT tissue
in the mouse colon sample we can cut the image at:</p>
<ul>
<li><p>x-left = 0.45</p></li>
<li><p>y-left = 0.55</p></li>
<li><p>x-right = 0.65</p></li>
<li><p>y-right = 0.7</p></li>
</ul>
<p>Provide these coordinates to the <code>crop_area</code> argument in
that order (x-left, y-left, x-right, y-right).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cd79a"</span>, <span class="st">"Clu"</span><span class="op">)</span>, image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                 pt_size <span class="op">=</span> <span class="fl">3</span>, section_number <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                 color <span class="op">=</span> <span class="va">cols</span>, crop_area <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.45</span>, <span class="fl">0.55</span>, <span class="fl">0.65</span>, <span class="fl">0.7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="numeric_features_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>And we can patch together a nice figure showing the expression both
at a global level and inside the GALT:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p_global</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>                        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cd79a"</span>, <span class="st">"Clu"</span><span class="op">)</span>, </span>
<span>                        image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                        scale_alpha <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                        pt_size <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                        section_number <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                        color <span class="op">=</span> <span class="va">cols</span>, </span>
<span>                        override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_GALT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, </span>
<span>                      features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Cd79a"</span>, <span class="st">"Clu"</span><span class="op">)</span>, </span>
<span>                      image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                      scale_alpha <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                      pt_size <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                      section_number <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                      color <span class="op">=</span> <span class="va">cols</span>, </span>
<span>                      crop_area <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.45</span>, <span class="fl">0.55</span>, <span class="fl">0.65</span>, <span class="fl">0.7</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>        plot.subtitle <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>, </span>
<span>        legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">p_global</span> <span class="op">/</span> <span class="va">p_GALT</span><span class="op">)</span></span></code></pre></div>
<p><img src="numeric_features_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p><br></p>
<hr>
<details open><summary><strong>Package version</strong>
</summary><ul>
<li>
<code>semla</code>: 1.4.0</li>
</ul></details><p><br></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://github.com/jemorlanes" class="external-link">Javier Escudero Morlanes</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference%2Ffigures%2Fsr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
