<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="semla">
<title>Radial distances • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Radial distances">
<meta property="og:description" content="semla">
<meta property="og:image" content="https://ludvigla.github.io/semla/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/semla.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Intro</h6>
    <a class="dropdown-item" href="../articles/getting_started.html">Get started</a>
    <a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive</h6>
    <a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6>
    <a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a>
    <a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a>
    <a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial methods</h6>
    <a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a>
    <a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a>
    <a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a>
    <a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Analysis</h6>
    <a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a>
    <a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a>
    <a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image processing</h6>
    <a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a>
    <a class="dropdown-item" href="../articles/mask_images.html">Image masking</a>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Staffli</h6>
    <a class="dropdown-item" href="../articles/create_object.html">Create Staffli object</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Radial distances</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 25 March
2023</h4>
      
      
      <div class="d-none name"><code>radial_distances.Rmd</code></div>
    </div>

    
    
<p>In this notebook, we’ll have a look at so called ‘spatial methods’
provided in <code>semla</code>. ‘spatial methods’ refers to analysis
methods which somehow incorporate the position of spots in their
calculations.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ludvigla.github.io/semla/">semla</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thomasp85/scico" class="external-link">scico</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>First we need to download some 10x Visium data. Here we’ll use a
breast cancer dataset provided by 10x:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create an empty folder in current directory</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"./bc_visium"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Breast_Cancer_Block_A_Section_1/V1_Breast_Cancer_Block_A_Section_1_filtered_feature_bc_matrix.h5"</span>, destfile <span class="op">=</span> <span class="st">"./bc_visium/filtered_feature_bc_matrix.h5"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Breast_Cancer_Block_A_Section_1/V1_Breast_Cancer_Block_A_Section_1_spatial.tar.gz"</span>, destfile <span class="op">=</span> <span class="st">"./bc_visium/spatial.tar.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span>tarfile <span class="op">=</span> <span class="st">"./bc_visium/spatial.tar.gz"</span>, </span>
<span>      exdir <span class="op">=</span>  <span class="st">"./bc_visium/"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">"./bc_visium/spatial.tar.gz"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="st">"bc_visium/filtered_feature_bc_matrix.h5"</span></span>
<span><span class="va">imgs</span> <span class="op">&lt;-</span> <span class="st">"bc_visium/spatial/tissue_hires_image.png"</span></span>
<span><span class="va">spotfiles</span> <span class="op">&lt;-</span> <span class="st">"bc_visium/spatial/tissue_positions_list.csv"</span></span>
<span><span class="va">json</span> <span class="op">&lt;-</span> <span class="st">"bc_visium/spatial/scalefactors_json.json"</span></span>
<span></span>
<span><span class="co"># Create a tibble/data.frame with file paths</span></span>
<span><span class="va">infoTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">imgs</span>, <span class="va">spotfiles</span>, <span class="va">json</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create Seurat object</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReadVisiumData.html">ReadVisiumData</a></span><span class="op">(</span>infoTable <span class="op">=</span> <span class="va">infoTable</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>Next, we’ll use <code>Seurat</code> to run a simple data
processing/analysis workflow:</p>
<ul>
<li><p>normalize data</p></li>
<li><p>scale data</p></li>
<li><p>find variable features</p></li>
<li><p>run dimensionality reduction by PCA</p></li>
<li><p>Create SNN graph</p></li>
<li><p>Cluster data</p></li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of nodes: 3813</span></span>
<span><span class="co">## Number of edges: 141050</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Running Louvain algorithm...</span></span>
<span><span class="co">## Maximum modularity in 10 random starts: 0.8713</span></span>
<span><span class="co">## Number of communities: 13</span></span>
<span><span class="co">## Elapsed time: 0 seconds</span></span></code></pre>
<p>Now that we have identified clusters in our data based on the spot
expression profiles, we can plot these clusters spatially with
<code>MapLabels</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## ── <span style="font-weight: bold;">Loading H&amp;E images</span> ──</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Loading image from /Users/ludviglarsson/10x_data/BC_block_S1/spatial/tissue_hires_image.png</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Scaled image from 2000x2000 to 400x400 pixels</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Saving loaded H&amp;E images as 'rasters' in Seurat object</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"seurat_clusters"</span>, </span>
<span>          image_use <span class="op">=</span> <span class="st">"raw"</span>, pt_alpha <span class="op">=</span> <span class="fl">0.6</span>, pt_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>For the purpose of this analysis, we just need to define a region of
interest in our tissue sections. We could just as well use the tissue
morphology as a basis to define this region. From the plot above, we can
see that some clusters are confined to distinct regions, for example
cluster 9:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">$</span><span class="va">cluster_9</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"9"</span>, <span class="st">"9"</span>, <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"cluster_9"</span>, override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>          image_use <span class="op">=</span> <span class="st">"raw"</span>, drop_na <span class="op">=</span> <span class="cn">TRUE</span>, pt_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="disconnect-regions">Disconnect regions<a class="anchor" aria-label="anchor" href="#disconnect-regions"></a>
</h2>
<p>Cluster 9 is concentrated to a region rich with cancer cells but
there are also a few spots in other parts of the tissue. At this point
we might only be interested in one of these regions in which case we can
split cluster_9 into spatially disconnected compartments. For this
purpose, we can use <code>DisconnectRegions</code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/disconnect-regions.html">DisconnectRegions</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"seurat_clusters"</span>, selected_groups <span class="op">=</span> <span class="st">"9"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Extracting disconnected components for group '9'</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Detecting disconnected regions for 178 spots</span></span></code></pre>
<pre><code><span><span class="co">## Loading required namespace: tidygraph</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Found 5 disconnected graph(s) in data</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Sorting disconnected regions by decreasing size</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Found 2 singletons in data</span></span></code></pre>
<pre><code><span><span class="co">## →   These will be labeled as 'singletons'</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"9_split"</span>, override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>          image_use <span class="op">=</span> <span class="st">"raw"</span>, drop_na <span class="op">=</span> <span class="cn">TRUE</span>, pt_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Each spatially disconnected region in cluster 9 now has its own
label, order from largest to smallest. Singletons are spots that are
completely isolated from other spots with the same label. One could of
course achieve the same result by doing a manual selection of spots, for
example using the <code>FeatureViewer</code> function, but
<code>DisconnectRegions</code> makes it easier to automate. At this
point, it might not be entirely clear why one would need this feature,
but hopefully this will make sense when we look at ‘radial distances’ in
the next section.</p>
</div>
<div class="section level2">
<h2 id="radial-distance">Radial distance<a class="anchor" aria-label="anchor" href="#radial-distance"></a>
</h2>
<p>Imagine that we are interested in looking at the expression of
certain genes as a function of distance to a region of interest. We can
compute distances from a region of interest (ROI) using
<code>RadialDistances</code>. Here we’ll use cluster 3 as our ROI for
the computation.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">$</span><span class="va">cluster_3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"3"</span>, <span class="st">"3"</span>, <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"cluster_3"</span>, override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>          image_use <span class="op">=</span> <span class="st">"raw"</span>, drop_na <span class="op">=</span> <span class="cn">TRUE</span>, pt_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/radial-distance.html">RadialDistance</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"seurat_clusters"</span>, selected_groups <span class="op">=</span> <span class="st">"3"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Running calculations for sample 1</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Calculating radial distances for group '3'</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Removing 3 spots with 0 neighbors.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Extracting border spots from a region with 413 spots</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 75 spots on borders</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 413 spots inside borders</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 3400 spots outside borders</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Returning radial distances</span></span></code></pre>
<p>We can illustrate the results by coloring the spots based on the
radial distances from cluster “3”:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, features <span class="op">=</span> <span class="st">"r_dist_3"</span>, center_zero <span class="op">=</span> <span class="cn">TRUE</span>, pt_size <span class="op">=</span> <span class="fl">2</span>, </span>
<span>            colors <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">11</span>, name <span class="op">=</span> <span class="st">"RdBu"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>The distances are calculated from the border of the ROI, where
positive values represent the radial distances out from the ROI and
negative values represent the radial distances towards the center of the
ROI.</p>
<div class="section level3">
<h3 id="pixel-coordinates">Pixel coordinates<a class="anchor" aria-label="anchor" href="#pixel-coordinates"></a>
</h3>
<p>The distances are given as pixels relative to the image pixels
coordinates that were loaded when creating the <code>Seurat</code>
object. If we know the conversion factor between pixels in our original
H&amp;E image and microns, we can easily convert the radial distances to
microns. In this example data, the center-to-center spot distance is 273
pixels in the image and we know that the actual center-to-center spot
distance is 100 microns.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">$</span><span class="va">r_dist_3</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fl">100</span><span class="op">/</span><span class="fl">273</span><span class="op">)</span><span class="op">*</span><span class="va">se</span><span class="op">$</span><span class="va">r_dist_3</span></span></code></pre></div>
<p>Alternatively, you can set <code>convert_to_microns = TRUE</code> to
make <code>RadialDistance</code> do the conversion for you. Note that
the conversion assumes that you are working with Visium data and that at
least some spots are adjacent in the dataset.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/radial-distance.html">RadialDistance</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"seurat_clusters"</span>, </span>
<span>                     selected_groups <span class="op">=</span> <span class="st">"3"</span>, convert_to_microns <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Running calculations for sample 1</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Calculating radial distances for group '3'</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Removing 3 spots with 0 neighbors.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Extracting border spots from a region with 413 spots</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 75 spots on borders</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 413 spots inside borders</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 3400 spots outside borders</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Returning radial distances</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="convert-radial-distances">Convert radial distances<a class="anchor" aria-label="anchor" href="#convert-radial-distances"></a>
</h3>
<p>The scale of the distances is somewhat inconvenient for our color
scale, so we can apply some transformation to make them easier to
visualize, for example the square root of the distance:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">$</span><span class="va">r_dist_3_sqrt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">r_dist_3</span><span class="op">)</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">se</span><span class="op">$</span><span class="va">r_dist_3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, features <span class="op">=</span> <span class="st">"r_dist_3_sqrt"</span>, center_zero <span class="op">=</span> <span class="cn">TRUE</span>, pt_size <span class="op">=</span> <span class="fl">2</span>, </span>
<span>            colors <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">11</span>, name <span class="op">=</span> <span class="st">"RdBu"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>With these distances, we can now explore the expression of certain
genes as a function of distance from our ROI. Here we are only
interested in the microenvironment outside of our cluster 3, so we will
filter the data to have a maximum radial distance of 1000 microns
outside of cluster 3. The red dashed line represents the tumor
border.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sel_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CRISP3"</span>, <span class="st">"IGLC2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html" class="external-link">FetchData</a></span><span class="op">(</span><span class="va">se</span>, vars <span class="op">=</span> <span class="va">sel_genes</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">r_dist_3</span> <span class="op">&lt;</span> <span class="fl">1e3</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="va">sel_genes</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"variable"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">r_dist_3</span>, <span class="va">value</span>, color <span class="op">=</span> <span class="va">variable</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"gam"</span>, formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu">s</span><span class="op">(</span><span class="va">x</span>, bs <span class="op">=</span> <span class="st">"cs"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"border"</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Here we can see that the expression of CRISP3 is high inside cluster
“3” and declines rapidly at the border whereas IGLC2 show the opposite
trend. We can also visualize these trends spatially:</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="fu"><a href="../reference/subset.html">SubsetSTData</a></span><span class="op">(</span><span class="va">se</span>, expression <span class="op">=</span> <span class="va">r_dist_3</span> <span class="op">&lt;</span> <span class="fl">1e3</span><span class="op">)</span>, </span>
<span>            features <span class="op">=</span> <span class="va">sel_genes</span>, override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, scale_alpha <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            image_use <span class="op">=</span> <span class="st">"raw"</span>, pt_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<p>So why bother using radial distances? Having access to this
information makes it possible to identify genes that change with
distance to a ROI. In the example above, we could for example identify
genes whose expression decrease or increase with distance from the tumor
which gives us a useful tool to characterize the tumor
microenvironment.</p>
<p>A perhaps even more useful application is to explore the relative
abundance of cell types around or inside the tumor border, which we can
do if we compute cell type proportions from our data fist.</p>
</div>
<div class="section level3">
<h3 id="focused-search">Focused search<a class="anchor" aria-label="anchor" href="#focused-search"></a>
</h3>
<p>It is rarely the case that the microenvironment is homogeneous in all
directions from the region of interest. In our breast cancer tissue
section, we can clearly see that the tissue composition is different
depending on which direction you look from our region of interest.
<code>RadialDistance</code> therefore provides an option to narrow down
the search space to a predefined interval of angles.</p>
<p>The plot below (generated with <code>AnglePlot</code>) should give
you an idea about how the angles are calculated for a certain region.
The ROI center is defined by taking the median of the x, y coordinates
for the spots of the ROI. The angles are defined clockwise where 0° =
right, 90° = down, …</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/AnglePlot.html">AnglePlot</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"cluster_3"</span>, selected_group <span class="op">=</span> <span class="st">"3"</span>, pt_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>          crop_area <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, image_use <span class="op">=</span> <span class="st">"raw"</span>, radius <span class="op">=</span> <span class="fl">0.3</span>, drop_na <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="section level4">
<h4 id="disconnected-regions">Disconnected regions<a class="anchor" aria-label="anchor" href="#disconnected-regions"></a>
</h4>
<p>Before you proceed, note that it is important that the ROI is
spatially connected. If not, <code>RadialDistance</code> will complain
that it finds disconnected components and will abort the function. This
is intended to prevent users from defining a center outside of the ROI
which would likely happen if multiple ROIs are present. It doesn’t make
much sense to compute radial distances from one center when there are
multiple centers present. Instead, you can split the disconnected
components with <code>DisconnectRegions</code> and focus on one ROI.</p>
<p>Now that we have an idea about how the angles are defined, we can
provide an interval of angles to narrow down the search area and
calculate radial distances. The color scale in the plot below represents
the radial distances from the tumor (cluster 3) center.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/radial-distance.html">RadialDistance</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"seurat_clusters"</span>, selected_groups <span class="op">=</span> <span class="st">"3"</span>, </span>
<span>                     angles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">200</span>, <span class="fl">240</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Running calculations for sample 1</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Calculating radial distances for group '3'</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Removing 3 spots with 0 neighbors.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Extracting border spots from a region with 413 spots</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 75 spots on borders</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 413 spots inside borders</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 3400 spots outside borders</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Returning radial distances</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Setting search area between 200 and 240 degrees from region center</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, features <span class="op">=</span> <span class="st">"r_dist_3"</span>, pt_size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<p>If we want to go for a more exploratory approach, we can split the
radial distances into slices by setting <code>angles_nbreaks</code>. In
the example below, we narrow the search area to angles between 120-320
degrees and split it into 6 even slices.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/radial-distance.html">RadialDistance</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"seurat_clusters"</span>, selected_groups <span class="op">=</span> <span class="st">"3"</span>, </span>
<span>                    angles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">120</span>, <span class="fl">320</span><span class="op">)</span>, angles_nbreaks <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Running calculations for sample 1</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Calculating radial distances for group '3'</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Removing 3 spots with 0 neighbors.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Extracting border spots from a region with 413 spots</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 75 spots on borders</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 413 spots inside borders</span></span></code></pre>
<pre><code><span><span class="co">## →   Detected 3400 spots outside borders</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> Returning radial distances</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Setting search area between 120 and 320 degrees from region center</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Splitting search area into 6 interval(s)</span></span></code></pre>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se</span>, column_name <span class="op">=</span> <span class="st">"intervals_3"</span>, pt_size <span class="op">=</span> <span class="fl">2</span>, drop_na <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          colors <span class="op">=</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html" class="external-link">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">6</span>, name <span class="op">=</span> <span class="st">"Spectral"</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<p>Inside our <code>Seurat</code> object meta data slot, we now have
three new columns: ‘angle_3’, ‘r_dist_3’ and ‘intervals_3’. The
‘intervals_3’ columns tells us what slice each spot belongs to. With
this data, we can modify our previous plot and split it by slice. If we
look at the same genes as before (CRISP3 and IGLC2), we can see that the
trends is quite similar in all 6 directions:</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sel_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CRISP3"</span>, <span class="st">"IGLC2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html" class="external-link">FetchData</a></span><span class="op">(</span><span class="va">se</span>, vars <span class="op">=</span> <span class="va">sel_genes</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">r_dist_3</span> <span class="op">&lt;</span> <span class="fl">1e3</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="va">sel_genes</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"variable"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">r_dist_3</span>, <span class="va">value</span>, color <span class="op">=</span> <span class="va">variable</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"gam"</span>, formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu">s</span><span class="op">(</span><span class="va">x</span>, bs <span class="op">=</span> <span class="st">"cs"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"border"</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">intervals_3</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<p>However, the trend varies substantially depending on direction for
other genes. Below we can see the expression along the radial axis for
three genes. Depending on the direction, these genes have very distinct
expression profiles.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sel_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PGM5-AS1"</span>, <span class="st">"CXCL14"</span>, <span class="st">"IGHG3"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/FetchData.html" class="external-link">FetchData</a></span><span class="op">(</span><span class="va">se</span>, vars <span class="op">=</span> <span class="va">sel_genes</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">r_dist_3</span> <span class="op">&lt;</span> <span class="fl">1e3</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html" class="external-link">all_of</a></span><span class="op">(</span><span class="va">sel_genes</span><span class="op">)</span>, names_to <span class="op">=</span> <span class="st">"variable"</span>, values_to <span class="op">=</span> <span class="st">"value"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">variable</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>value_centered <span class="op">=</span> <span class="va">value</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="co"># Note that the values have been centered</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">r_dist_3</span>, <span class="va">value_centered</span>, color <span class="op">=</span> <span class="va">variable</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">geom_smooth</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"gam"</span>, formula <span class="op">=</span> <span class="va">y</span> <span class="op">~</span> <span class="fu">s</span><span class="op">(</span><span class="va">x</span>, bs <span class="op">=</span> <span class="st">"cs"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, color <span class="op">=</span> <span class="st">"border"</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span> <span class="va">intervals_3</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>If we map the expression of these genes spatially, it is quite clear
that these genes are expressed in different niches of the
microenvironment. In fact, two of these “niches” are mostly composed of
tumor cells.</p>
<p>Hopefully, this demonstration will serve as a motivation to carefully
think through how radial distances should be used for specific
biological questions. The tools provided in <code>semla</code> are easy
to adapt to many different scenarios.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se</span>, crop_area <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.45</span>, <span class="fl">0.6</span>, <span class="fl">0.8</span>, <span class="fl">0.93</span><span class="op">)</span>,</span>
<span>            features <span class="op">=</span> <span class="va">sel_genes</span>, scale_alpha <span class="op">=</span> <span class="cn">TRUE</span>, ncol <span class="op">=</span> <span class="fl">3</span>,</span>
<span>            image_use <span class="op">=</span> <span class="st">"raw"</span>, pt_size <span class="op">=</span> <span class="fl">1.5</span>, colors <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">9</span><span class="op">)</span>,</span>
<span>            max_cutoff <span class="op">=</span> <span class="fl">0.99</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"right"</span>, legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="radial_distances_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p><br></p>
<hr>
<details open><summary><strong>Package version</strong>
</summary><ul>
<li>
<code>semla</code>: 1.0.0</li>
</ul></details><details><summary><strong>Session info</strong>
</summary><div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin13.4.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Big Sur ... 10.16</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /Users/ludviglarsson/miniconda3/envs/R4.2/lib/libopenblasp-r0.3.21.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] tidyr_1.3.0        scico_1.3.1        patchwork_1.1.2    tibble_3.1.8      </span></span>
<span><span class="co">## [5] semla_1.0.0        ggplot2_3.4.1      dplyr_1.1.0        SeuratObject_4.1.3</span></span>
<span><span class="co">## [9] Seurat_4.3.0      </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] systemfonts_1.0.4      plyr_1.8.7             igraph_1.3.4          </span></span>
<span><span class="co">##   [4] lazyeval_0.2.2         sp_1.5-1               splines_4.2.1         </span></span>
<span><span class="co">##   [7] listenv_0.8.0          scattermore_0.8        digest_0.6.29         </span></span>
<span><span class="co">##  [10] htmltools_0.5.4        viridis_0.6.2          magick_2.7.3          </span></span>
<span><span class="co">##  [13] fansi_1.0.3            magrittr_2.0.3         memoise_2.0.1         </span></span>
<span><span class="co">##  [16] tensor_1.5             cluster_2.1.4          ROCR_1.0-11           </span></span>
<span><span class="co">##  [19] globals_0.16.0         matrixStats_0.62.0     pkgdown_2.0.6         </span></span>
<span><span class="co">##  [22] spatstat.sparse_3.0-0  colorspace_2.0-3       ggrepel_0.9.3         </span></span>
<span><span class="co">##  [25] textshaping_0.3.6      xfun_0.32              crayon_1.5.1          </span></span>
<span><span class="co">##  [28] jsonlite_1.8.3         progressr_0.10.1       spatstat.data_3.0-0   </span></span>
<span><span class="co">##  [31] zeallot_0.1.0          survival_3.4-0         zoo_1.8-10            </span></span>
<span><span class="co">##  [34] glue_1.6.2             polyclip_1.10-0        gtable_0.3.0          </span></span>
<span><span class="co">##  [37] leiden_0.4.2           future.apply_1.9.0     abind_1.4-5           </span></span>
<span><span class="co">##  [40] scales_1.2.1           spatstat.random_3.0-1  miniUI_0.1.1.1        </span></span>
<span><span class="co">##  [43] Rcpp_1.0.9             viridisLite_0.4.1      xtable_1.8-4          </span></span>
<span><span class="co">##  [46] reticulate_1.26        bit_4.0.4              htmlwidgets_1.5.4     </span></span>
<span><span class="co">##  [49] httr_1.4.4             RColorBrewer_1.1-3     ellipsis_0.3.2        </span></span>
<span><span class="co">##  [52] ica_1.0-3              pkgconfig_2.0.3        farver_2.1.1          </span></span>
<span><span class="co">##  [55] sass_0.4.2             uwot_0.1.14            deldir_1.0-6          </span></span>
<span><span class="co">##  [58] utf8_1.2.2             labeling_0.4.2         tidyselect_1.2.0      </span></span>
<span><span class="co">##  [61] rlang_1.0.6            reshape2_1.4.4         later_1.3.0           </span></span>
<span><span class="co">##  [64] munsell_0.5.0          tools_4.2.1            cachem_1.0.6          </span></span>
<span><span class="co">##  [67] cli_3.4.1              dbscan_1.1-10          generics_0.1.3        </span></span>
<span><span class="co">##  [70] ggridges_0.5.3         evaluate_0.16          stringr_1.5.0         </span></span>
<span><span class="co">##  [73] fastmap_1.1.0          yaml_2.3.5             ragg_1.2.2            </span></span>
<span><span class="co">##  [76] goftest_1.2-3          knitr_1.39             bit64_4.0.5           </span></span>
<span><span class="co">##  [79] fs_1.5.2               fitdistrplus_1.1-8     tidygraph_1.2.2       </span></span>
<span><span class="co">##  [82] purrr_1.0.1            RANN_2.6.1             pbapply_1.5-0         </span></span>
<span><span class="co">##  [85] future_1.27.0          nlme_3.1-159           mime_0.12             </span></span>
<span><span class="co">##  [88] hdf5r_1.3.5            compiler_4.2.1         rstudioapi_0.14       </span></span>
<span><span class="co">##  [91] plotly_4.10.0          png_0.1-7              spatstat.utils_3.0-1  </span></span>
<span><span class="co">##  [94] bslib_0.4.0            stringi_1.7.8          highr_0.9             </span></span>
<span><span class="co">##  [97] desc_1.4.1             forcats_0.5.2          lattice_0.20-45       </span></span>
<span><span class="co">## [100] Matrix_1.5-3           shinyjs_2.1.0          vctrs_0.5.2           </span></span>
<span><span class="co">## [103] pillar_1.8.1           lifecycle_1.0.3        spatstat.geom_3.0-3   </span></span>
<span><span class="co">## [106] lmtest_0.9-40          jquerylib_0.1.4        RcppAnnoy_0.0.19      </span></span>
<span><span class="co">## [109] data.table_1.14.2      cowplot_1.1.1          irlba_2.3.5           </span></span>
<span><span class="co">## [112] httpuv_1.6.5           R6_2.5.1               promises_1.2.0.1      </span></span>
<span><span class="co">## [115] KernSmooth_2.23-20     gridExtra_2.3          parallelly_1.32.1     </span></span>
<span><span class="co">## [118] codetools_0.2-18       MASS_7.3-58.1          rprojroot_2.0.3       </span></span>
<span><span class="co">## [121] withr_2.5.0            sctransform_0.3.5      mgcv_1.8-40           </span></span>
<span><span class="co">## [124] parallel_4.2.1         grid_4.2.1             rmarkdown_2.15        </span></span>
<span><span class="co">## [127] Rtsne_0.16             spatstat.explore_3.0-5 shiny_1.7.4</span></span></code></pre>
</details><p><br></p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
