<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Spatial autocorrelation • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Spatial autocorrelation">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/semla.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><h6 class="dropdown-header" data-toc-skip>Intro</h6></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a></li>
    <li><a class="dropdown-item" href="../articles/visiumHD.html">VisiumHD</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a></li>
    <li><a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a></li>
    <li><a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a></li>
    <li><a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial methods</h6></li>
    <li><a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a></li>
    <li><a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a></li>
    <li><a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a></li>
    <li><a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Image processing</h6></li>
    <li><a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a></li>
    <li><a class="dropdown-item" href="../articles/mask_images.html">Image masking</a></li>
    <li><a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/create_object.html">Create a Staffli object</a></li>
    <li><a class="dropdown-item" href="../articles/compare_cell_type_mapping_NNLS.html">Cell type mapping comparison</a></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_benchmark.html">Cell type mapping benchmark</a></li>
    <li><a class="dropdown-item" href="../articles/slide-seq.html">Slide-Seq data</a></li>
    <li><a class="dropdown-item" href="../articles/IF_data.html">IF data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Spatial autocorrelation</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 03 March
2025</h4>
      

      <div class="d-none name"><code>spatial_autocorrelation.Rmd</code></div>
    </div>

    
    
<p><code>semla</code> offers a fast method to compute spatial
autocorrelation scores for numeric features.</p>
<p>Spatial autocorrelation is a term used to describe the presence of
systematic spatial variation. Positive spatial autocorrelation of a
feature is the tendency for regions that are close together in space to
have similar values for that feature.</p>
<p>A simple example is when you have an anatomical structure or a tissue
type that spans across multiple neighboring spots in an SRT experiment,
for example a gland, an immune infiltrate or a region of the brain.
Inside such structures, you might find that the expression levels of
certain genes (or other features) are highly similar and hence these
genes have a positive spatial autocorrelation.</p>
<p>Perhaps the most widely used method for computing spatial
autocorrelation scores is <a href="https://en.wikipedia.org/wiki/Moran%27s_I" class="external-link">Moran’s I</a>. The
computation of Moran’s I doesn’t scale well and often require some sort
of binning strategy to mitigate this problem.</p>
<p>The method provided in <code>semla</code> uses a fast alternative
which relies on computing the “spatial lag” vector of each feature. Here
the spatial lag of a numeric feature is defined as the averaged value of
that feature in adjacent spots. Once this score has been calculated, we
can estimate spatial autocorrelation by computing the Pearson
correlation between the original vector and its spatial lag vector.</p>
<p>We will demonstrate how this method, using the function
<code><a href="../reference/cor-features.html">CorSpatialFeatures()</a></code>, work with our mouse brain data
set.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ludvigla.github.io/semla/">semla</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mousebrain"</span>,</span>
<span>                                 <span class="st">"se_mbrain"</span>,</span>
<span>                                 package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Compute spatial autocorrelation for variable features</span></span>
<span><span class="va">spatgenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cor-features.html">CorSpatialFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## ── <span style="font-weight: bold;">Computing spatial autocorrelation</span> ──</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Sample 1:</span></span></code></pre>
<pre><code><span><span class="co">## →   Cleaned out spots with 0 adjacent neighbors</span></span></code></pre>
<pre><code><span><span class="co">## →   Computed feature lag expression</span></span></code></pre>
<pre><code><span><span class="co">## →   Computed feature spatial autocorrelation scores</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span>   Returning results</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spatgenes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 170 × 2</span></span></span>
<span><span class="co">##    gene     cor</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> Mbp    0.870</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> Nrgn   0.857</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> Cck    0.813</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> Olfm1  0.798</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> Slc6a3 0.798</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> Trh    0.770</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> Mobp   0.763</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> Th     0.757</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> Crym   0.756</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> Tmsb4x 0.753</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 160 more rows</span></span></span></code></pre>
<p>By default, the method computes scores for the variable features of
the <code>Seurat</code> objects returns the results in a list with one
element per tissue section. If we plot the top ranked genes, we will see
that these genes have a distinct spatial profile:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>            features <span class="op">=</span> <span class="va">spatgenes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span>, </span>
<span>            override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colors <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">magma</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">11</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span>, </span>
<span>            min_cutoff <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial_autocorrelation_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p><br></p>
<p>In contrast, genes with a more random spatial distribution are
assigned a low correlation score:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">spatgenes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 2</span></span></span>
<span><span class="co">##   gene               cor</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> 2010109I03Rik -<span style="color: #BB0000;">0.002</span><span style="color: #BB0000; text-decoration: underline;">85</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> Slc51a        -<span style="color: #BB0000;">0.004</span><span style="color: #BB0000; text-decoration: underline;">72</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> H2-DMb2       -<span style="color: #BB0000;">0.005</span><span style="color: #BB0000; text-decoration: underline;">51</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> Ighd          -<span style="color: #BB0000;">0.005</span><span style="color: #BB0000; text-decoration: underline;">69</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> Cd79b         -<span style="color: #BB0000;">0.010</span><span style="color: #BB0000; text-decoration: underline;">8</span> </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> Emp1          -<span style="color: #BB0000;">0.025</span><span style="color: #BB0000; text-decoration: underline;">3</span></span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span>, </span>
<span>            features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">spatgenes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">$</span><span class="va">gene</span>, </span>
<span>            override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            colors <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">magma</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">11</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial_autocorrelation_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p><br></p>
<p>With these spatial autocorrelation scores in our hands, we can for
example prioritize genes with defined spatial structures for down stream
analysis. As an example, hemoglobin transcripts are often highly
variable due to the presence of blood vessels that are dispersed across
tissue sections. These blood vessels might be of little interest if the
goal is to identify larger “spatial domains” in the tissue. In fact, the
hemoglobin transcripts “Hbb-bs”, “Hba-a1”, “Hba-a2” and “Hbb-bt” are the
top four most variable genes in our dataset and the expression of these
transcripts might influence the downstream steps such as data-driven
clustering.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hemoglobin_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Hbb-bs"</span>, <span class="st">"Hba-a1"</span>, <span class="st">"Hba-a2"</span>, <span class="st">"Hbb-bt"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">hemoglobin_genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1 2 3 4</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span>, features <span class="op">=</span> <span class="va">hemoglobin_genes</span>, override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial_autocorrelation_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p><br></p>
<p>Let’s demonstrate with an example how we can use spatial
autocorrelation scores to assign higher importance to transcripts with
more distinct spatial expression. In the histogram below, we can see the
correlation scores for the genes in our data set and we select genes
with a score higher than 0.5 which leaves us with 70 genes to be used
for downstream analysis.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spatgenes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>selected <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">cor</span> <span class="op">&gt;</span> <span class="fl">0.5</span>, <span class="st">"selected"</span>, <span class="st">"discarded"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">cor</span>, fill <span class="op">=</span> <span class="va">selected</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre>
<p><img src="spatial_autocorrelation_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p><br></p>
<p>Now we can compare the results of data-driven clustering when using
our top 70 “spatial genes” with the results when using the to 70 most
variable genes.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_var_features</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">se_mbrain</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">70</span><span class="op">]</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="va">se_mbrain</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>features <span class="op">=</span> <span class="va">top_var_features</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span>, resolution <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span></span>
<span><span class="va">se_mbrain</span><span class="op">$</span><span class="va">clusters_var_genes</span> <span class="op">&lt;-</span> <span class="va">se_mbrain</span><span class="op">$</span><span class="va">seurat_clusters</span></span>
<span></span>
<span><span class="va">top_spatial_features</span> <span class="op">&lt;-</span> <span class="va">spatgenes</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">gene</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">70</span><span class="op">]</span></span>
<span><span class="va">se_mbrain</span> <span class="op">&lt;-</span> <span class="va">se_mbrain</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>features <span class="op">=</span> <span class="va">top_spatial_features</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>verbose <span class="op">=</span> <span class="cn">FALSE</span>, resolution <span class="op">=</span> <span class="fl">1.0</span><span class="op">)</span></span>
<span><span class="va">se_mbrain</span><span class="op">$</span><span class="va">clusters_spatial_genes</span> <span class="op">&lt;-</span> <span class="va">se_mbrain</span><span class="op">$</span><span class="va">seurat_clusters</span></span></code></pre></div>
<p><br></p>
<p>At a first glance, the results might not look too different from each
other. However, if we look closely, we can see that when we focus on the
top variable genes, some clusters are more dispersed across the tissue
section.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se_mbrain</span>, column_name <span class="op">=</span> <span class="st">"clusters_var_genes"</span>, split_labels <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>          override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, section_number <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial_autocorrelation_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se_mbrain</span>, column_name <span class="op">=</span> <span class="st">"clusters_spatial_genes"</span>, split_labels <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>          override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>, section_number <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="spatial_autocorrelation_files/figure-html/unnamed-chunk-8-2.png" width="700"></p>
<p><br></p>
<p>The selection of genes boils down to what you are trying to achieve
with your analysis. Note that if you base your selection on spatial
autocorrelation, you might end up hiding sources of biological
variability such as blood vessels or infiltrating immune cells. If the
goal is to identify more general structure (“spatial domains”), spatial
autocorrelation can be a useful metric to base gene selection on.</p>
<p><br></p>
<p>It is also possible to compute p-values for each correlation score by
passing the argument <code>calculate_pvalue = TRUE</code> to the
<code><a href="../reference/cor-features.html">CorSpatialFeatures()</a></code> function (<em>semla</em> v. ≥1.1.7).
P-values and adjusted p-values will appear in the results table as
separate columns (<code>pval</code>, <code>adj.pval</code>), where the
adjusted p-value is obtained using the Benjamini &amp; Hochberg (BH)
method, also know as FDR.</p>
<p><br></p>
<hr>
<details open><summary><strong>Package version</strong>
</summary><ul>
<li>
<code>semla</code>: 1.3.0</li>
</ul></details><p><br></p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://github.com/jemorlanes" class="external-link">Javier Escudero Morlanes</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
