<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>VisiumHD: loading and basic plotting • semla</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="VisiumHD: loading and basic plotting">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/semla.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials">
<li><h6 class="dropdown-header" data-toc-skip>Intro</h6></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a></li>
    <li><a class="dropdown-item" href="../articles/visiumHD.html">VisiumHD</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a></li>
    <li><a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a></li>
    <li><a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a></li>
    <li><a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial methods</h6></li>
    <li><a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a></li>
    <li><a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a></li>
    <li><a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a></li>
    <li><a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Image processing</h6></li>
    <li><a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a></li>
    <li><a class="dropdown-item" href="../articles/mask_images.html">Image masking</a></li>
    <li><a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a></li>
  </ul>
</li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/create_object.html">Create a Staffli object</a></li>
    <li><a class="dropdown-item" href="../articles/compare_cell_type_mapping_NNLS.html">Cell type mapping comparison</a></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_benchmark.html">Cell type mapping benchmark</a></li>
    <li><a class="dropdown-item" href="../articles/slide-seq.html">Slide-Seq data</a></li>
    <li><a class="dropdown-item" href="../articles/IF_data.html">IF data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>VisiumHD: loading and basic plotting</h1>
            
            <h4 data-toc-skip class="date">Last compiled: 30 September
2024</h4>
      

      <div class="d-none name"><code>visiumHD.Rmd</code></div>
    </div>

    
    
<p>In this article, we will show how to load and work with VisiumHD data
using <code>semla</code>.</p>
<p>We are going to be using the FFPE mouse brain test dataset that is
available at 10x Genomics <a href="https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-of-mouse-brain-he" class="external-link">website</a>.
Specifically, we are going to use the <em>binned_outputs</em> folder.
Mouse brain, FFPE, 6.5mm.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ludvigla.github.io/semla/">semla</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="introduction-to-visiumhd">Introduction to VisiumHD<a class="anchor" aria-label="anchor" href="#introduction-to-visiumhd"></a>
</h3>
<p>The high-resolution VisiumHD commercialized by 10x Genomics in spring
2024 follows the same philosophy as the traditional Visium platform we
are used to: spatially barcoded spots capture the transcriptome (or most
of the transcriptome since it is probe based but now there is FF
protocol), and the location of the transcripts can be then mapped to the
tissue by the spatial sequence that is specific to each spot. While
similar in philosophy, both technologies have some fundamental
differences:</p>
<ol style="list-style-type: decimal">
<li>The capture area in Visium are rounded spots distributed in a
<em>honeycomb</em> or <em>hexagonal</em> pattern, while in VisiumHD they
are squares distributed in a <em>grid</em> pattern.</li>
<li>In the traditional Visium, there is a distance of 100um between the
spot’s center, whereas in VisiumHD there is no gaps in between. All the
spots are immediately next to each-other, leaving no space.</li>
<li>In Visium, the spots are 55um in diameter, meaning that the spatial
resolution for the technology is 55um. In VisiumHD, the spots/squares
are 2um in side, meaning that the spatial resolution is of 2um. In
theory, this should allow for a sub-cellular resolution of the
transcriptome. To put this into context, there is around
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>55</mn><mi>/</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup><annotation encoding="application/x-tex">(55/2)^2</annotation></semantics></math>
VisiumHD spots per Visium spot.</li>
</ol>
<p>So, 3 main things regarding VisiumHD: the spots are now squares
instead of circles, the spots have no empty space in between them, and
the size of the spots is 2um. This is a bit ugly, but maybe it can help
put these numbers into context:</p>
<div class="float">
<img src="visiumHD/visiumhd_in_context.png" alt="How many VisiumHD spots there is in one Visium spot!"><div class="figcaption">How many VisiumHD spots there is in one Visium
spot!</div>
</div>
<p>As a side note, maybe it is more accurate to refer to VisiumHD’s
capture areas as <em>squares</em>, but due to old habit we will still
refer to them as <em>spots</em>.</p>
</div>
<div class="section level3">
<h3 id="data-sparsity">Data sparsity<a class="anchor" aria-label="anchor" href="#data-sparsity"></a>
</h3>
<p>An inherited problem of high-resolution spatial transcriptomics, be
that sequencing-based (like BGI’s <a href="https://en.stomics.tech/products/stereo-seq-transcriptomics-solution/list.html" class="external-link">Stereo-seq</a>,
Curio’s <a href="https://curiobioscience.com/press/curio-bioscience-launches-spatial-transcriptomics-solution/" class="external-link">Slide-seq</a>,
and other alternatives) or imaging-based (like 10x Genomics’s <a href="https://www.10xgenomics.com/platforms/xenium" class="external-link">ISS platform</a>
Xenium, VIZGEN’s <a href="https://vizgen.com/merscope-ultra/" class="external-link">MERSCOPE</a>, and others), is
that the data tends to be quite sparse: few data points per spatial
region. A common approach to circumvent this issue is to
<strong>bin/group</strong> the spatial regions to increase the density
of the data. Nevertheless, the cost of this binning is that you lose
spatial resolution. There are different ways to bin the spatial data,
but the most common approaches are via cell segmentation and grid
binning. So, with high-resolution data, you have the following problem:
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>H</mi><mi>i</mi><mi>g</mi><mi>h</mi><mi>S</mi><mi>p</mi><mi>a</mi><mi>t</mi><mi>i</mi><mi>a</mi><mi>l</mi><mi>R</mi><mi>e</mi><mi>s</mi><mi>o</mi><mi>l</mi><mi>u</mi><mi>t</mi><mi>i</mi><mi>o</mi><mi>n</mi><mo>=</mo><mi>S</mi><mi>p</mi><mi>a</mi><mi>r</mi><mi>s</mi><mi>e</mi><mi>D</mi><mi>a</mi><mi>t</mi><mi>a</mi><mo>=</mo><mi>B</mi><mi>i</mi><mi>n</mi><mi>n</mi><mi>i</mi><mi>n</mi><mi>g</mi><mo>=</mo><mi>N</mi><mi>o</mi><mi>t</mi><mi>H</mi><mi>i</mi><mi>g</mi><mi>h</mi><mi>S</mi><mi>p</mi><mi>a</mi><mi>t</mi><mi>i</mi><mi>a</mi><mi>l</mi><mi>R</mi><mi>e</mi><mi>s</mi><mi>o</mi><mi>l</mi><mi>u</mi><mi>t</mi><mi>i</mi><mi>o</mi><mi>n</mi><mi>A</mi><mi>n</mi><mi>y</mi><mi>m</mi><mi>o</mi><mi>r</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">HighSpatialResolution = SparseData = Binning = NotHighSpatialResolutionAnymore</annotation></semantics></math>.</p>
<p>This is a problem the field is facing, and it will be interesting to
see what creative approaches the community takes in order to maximize
the information while minimizing the decrease in spatial resolution. As
of the time this markdown was last knitted, the approach taken by 10x
Genomics by default is <strong>grid binning</strong>. What this means is
that all spots that are within a certain distance of each other will be
grouped under the same spot, and their expression data summed up. The
spots in VisiumHD are 2um in side, so if we talk about an <em>8um
bin</em> we mean that the new spots are going to be squares of 8um in
side: we will be adding up the expression of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>4</mn><mo>*</mo><mn>4</mn></mrow><annotation encoding="application/x-tex">4*4</annotation></semantics></math>
adjacent 2um spots to generate one 8um spot.</p>
</div>
<div class="section level3">
<h3 id="grid-binning">Grid binning<a class="anchor" aria-label="anchor" href="#grid-binning"></a>
</h3>
<p>In VisiumHD, the “original” count data is the count data at the 2um
spatial resolution, but 10x Genomics provides several
<em>binned_outputs</em> by default when running <a href="https://www.10xgenomics.com/support/software/space-ranger/latest/analysis/outputs/output-overview" class="external-link">spaceranger</a>:
2, 8 and 16, with the option of specifying a custom resolution when
calling spaceranger. At the 8um bin, each spot is a sum of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>4</mn><mo>*</mo><mn>4</mn></mrow><annotation encoding="application/x-tex">4*4</annotation></semantics></math>
2um spots; at 16um bin, each spot is a sum of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>8</mn><mo>*</mo><mn>8</mn></mrow><annotation encoding="application/x-tex">8*8</annotation></semantics></math>
2um spots; and so on.</p>
<p>For each one of these resolutions, spaceranger returns an entire
folder with the structure we are used to from the normal Visium. This
folder will contain the spatial position of each spot, the count matrix,
the tissue HE image and the json file required to map the spots to the
HE image.</p>
</div>
<div class="section level3">
<h3 id="folder-structure">Folder structure<a class="anchor" aria-label="anchor" href="#folder-structure"></a>
</h3>
<p>We can think of each subfolder in the <em>binned_outputs</em>
directory as it’s own sample. So for each tissue section, we will have
at least “3 samples”: one for the 2um resolution, another for the 8um
and another for the 16um. But do not forget that all the bin sizes above
2um are made up by adding up the “original” count data, which is the 2um
data.</p>
</div>
</div>
<div class="section level2">
<h2 id="load-data">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<p>First we need to load some 10x VisiumHD data. You can run the code
chunk below or follow this <a href="https://www.10xgenomics.com/datasets/visium-hd-cytassist-gene-expression-libraries-of-mouse-brain-he" class="external-link">link</a>
to download it manually.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>tmpdir <span class="ot">&lt;-</span> <span class="st">"."</span> <span class="co"># Set current wd or change to tmpdir()</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(tmpdir, <span class="st">"/mousebrain"</span>))</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>targetdir <span class="ot">&lt;-</span> <span class="fu">paste0</span>(tmpdir, <span class="st">"/mousebrain/visium-hd"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">dir.create</span>(targetdir)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>destfile <span class="ot">&lt;-</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/mouse_brain_HD.tar.gz"</span>)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/3.0.0/Visium_HD_Mouse_Brain/Visium_HD_Mouse_Brain_binned_outputs.tar.gz"</span>, <span class="at">destfile =</span> destfile)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="co"># now we decompress it</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="fu">untar</span>(<span class="fu">paste0</span>(targetdir, <span class="st">"mouse_brain_HD.tar.gz"</span>)</span></code></pre></div>
<p>As mentioned above, we can consider each resolution as its separate
section, and load them as such into R using <code>semla</code>. Let’s
construct the <code>infoTable</code> with information regarding all the
resolutions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># directory where the data is stored</span></span>
<span><span class="va">st.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"mousebrain/visium-hd/binned_outputs"</span><span class="op">)</span></span>
<span><span class="co"># specify resolutions to load into memory</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">8</span>, <span class="fl">16</span><span class="op">)</span>, <span class="st">"um"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span><span class="va">all.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">res</span>, collapse <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># retrieve the directories for each resolution</span></span>
<span><span class="va">res.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="va">st.dir</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="va">all.res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># build the infoTable for all resolutions</span></span>
<span><span class="va">infoTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">res.dir</span>,</span>
<span>                       full.names <span class="op">=</span> <span class="cn">TRUE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"^filtered_feature.+.h5$"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  spotfiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">res.dir</span>,</span>
<span>                         full.names <span class="op">=</span> <span class="cn">TRUE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         pattern <span class="op">=</span> <span class="st">"parquet$|positions.csv$"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  imgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">res.dir</span>,</span>
<span>                    recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"hires"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="va">all.res</span><span class="op">)</span>,</span>
<span>  json <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">st.dir</span>,</span>
<span>                    recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"^scalefactors"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="va">all.res</span><span class="op">)</span>,</span>
<span>  resolution <span class="op">=</span> <span class="va">res</span>,</span>
<span>  sample_ID <span class="op">=</span> <span class="st">"mousebrain"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now that we have built the infoTable, we can load the VisiumHD data
into memory. Be aware that the 2um data is very big, so due to RAM
issues you might not be able to work with this data locally. For
example, if you want to build the <code>infoTable</code> to store only
the 8 and 16um resolutions, you could do the following:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">st.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"mousebrain/visium-hd/binned_outputs"</span><span class="op">)</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">16</span><span class="op">)</span>, <span class="st">"um"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span><span class="va">all.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">res</span>, collapse <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># retrieve the directories for each resolution</span></span>
<span><span class="va">res.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="va">st.dir</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="va">all.res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># build the infoTable for all resolutions</span></span>
<span><span class="va">infoTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">res.dir</span>,</span>
<span>                       full.names <span class="op">=</span> <span class="cn">TRUE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"^filtered_feature.+.h5$"</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  spotfiles <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">res.dir</span>,</span>
<span>                         full.names <span class="op">=</span> <span class="cn">TRUE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                         pattern <span class="op">=</span> <span class="st">"parquet$|positions.csv$"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  imgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">res.dir</span>,</span>
<span>                    recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"hires"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="va">all.res</span><span class="op">)</span>,</span>
<span>  json <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">st.dir</span>,</span>
<span>                    recursive <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                    full.names <span class="op">=</span> <span class="cn">TRUE</span>, pattern <span class="op">=</span> <span class="st">"^scalefactors"</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_subset.html" class="external-link">str_subset</a></span><span class="op">(</span><span class="va">all.res</span><span class="op">)</span>,</span>
<span>  resolution <span class="op">=</span> <span class="va">res</span>,</span>
<span>  sample_ID <span class="op">=</span> <span class="st">"mousebrain"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Regardless of which resolution you want to load, with the built
<code>infoTable</code> now we can load the data using
<code><a href="../reference/ReadVisiumData.html">ReadVisiumData()</a></code> like you would in your standard Visium
sections. If you have the 2um resolution in the infoTable, this can take
a bit of time.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se.hd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ReadVisiumData.html">ReadVisiumData</a></span><span class="op">(</span><span class="va">infoTable</span><span class="op">)</span></span>
<span><span class="va">se.hd</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 19059 features across 492460 samples within 1 assay </span></span>
<span><span class="co">## Active assay: Spatial (19059 features, 0 variable features)</span></span></code></pre>
<p>As mentioned before, now each resolution of the VisiumHD experiment
is stored in the Seurat object as an individual experiment. Let’s take a
look at the HE images. You can note how, even though we are labeling
each image by expressing from which resolution they come from, they are
exactly the same image.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se.hd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load-images.html">LoadImages</a></span><span class="op">(</span><span class="va">se.hd</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/ImagePlot.html">ImagePlot</a></span><span class="op">(</span><span class="va">se.hd</span>, label_by <span class="op">=</span> <span class="st">"resolution"</span><span class="op">)</span></span></code></pre></div>
<p><img src="visiumHD_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>This object is fully compatible with all of <code>semla</code>’s
functions, so you can proceed normally with your analysis!</p>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>Two of the main issues when working with high-resolution spatial
transcriptomics is <strong>data sparsity</strong> and the <strong>size
of the datasets.</strong> For example, do not be surprised if your
standard analysis does not run at all at the higher resolutions in
VisiumHD data. For standard Seurat clustering (v4 and prior), we
recommend working on a high performance computing cluster.</p>
<p>Thus, we are going to perform the standard analysis pipeline on the 8
and 16um bins.</p>
<div class="section level3">
<h3 id="seurat-clustering">Seurat clustering<a class="anchor" aria-label="anchor" href="#seurat-clustering"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 16um</span></span>
<span><span class="va">se.16</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subset.html">SubsetSTData</a></span><span class="op">(</span><span class="va">se.hd</span>, expression <span class="op">=</span> <span class="va">resolution</span> <span class="op">==</span> <span class="st">"016um"</span><span class="op">)</span></span>
<span><span class="va">se.16</span> <span class="op">&lt;-</span>  <span class="va">se.16</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">NormalizeData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">ScaleData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">RunPCA</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">FindNeighbors</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">FindClusters</span><span class="op">(</span>resolution <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 8um</span></span>
<span><span class="va">se.8</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subset.html">SubsetSTData</a></span><span class="op">(</span><span class="va">se.hd</span>, expression <span class="op">=</span> <span class="va">resolution</span> <span class="op">==</span> <span class="st">"08um"</span><span class="op">)</span></span>
<span><span class="va">se.8</span> <span class="op">&lt;-</span>  <span class="va">se.8</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">NormalizeData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">ScaleData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">FindVariableFeatures</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">RunPCA</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">FindNeighbors</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"pca"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">FindClusters</span><span class="op">(</span>resolution <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="non-negative-matrix-factorization">Non-negative matrix factorization<a class="anchor" aria-label="anchor" href="#non-negative-matrix-factorization"></a>
</h3>
<p>In the same way we can perform the standard Seurat analysis pipeline,
we can also follow our <a href="https://ludvigla.github.io/semla/articles/NNMF.html">tutorial</a>
for performing non-Negative matrix factorization (NMF).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/zdebruine/singlet" class="external-link">singlet</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 16um</span></span>
<span><span class="va">se.16</span> <span class="op">&lt;-</span> <span class="va">se.16</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">se.16</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co">## Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">se.16</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/singlet/man/RunNMF.html" class="external-link">RunNMF</a></span><span class="op">(</span><span class="va">se.16</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 8um</span></span>
<span><span class="va">se.8</span> <span class="op">&lt;-</span> <span class="va">se.8</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span><span class="va">se.8</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="co">## Set seed for reproducibility</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">42</span><span class="op">)</span></span>
<span><span class="va">se.8</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/singlet/man/RunNMF.html" class="external-link">RunNMF</a></span><span class="op">(</span><span class="va">se.8</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h2>
<p>We have done a few changes to <code>semla</code>’s plotting functions
to properly represent this new type of data. The capture areas of a
Visium slide are circles, and we represent them as such when using
<code><a href="../reference/visualize-features.html">MapFeatures()</a></code>, <code><a href="../reference/visualize-labels.html">MapLabels()</a></code> or
<code><a href="../reference/visualize-multiple-features.html">MapMultipleFeatures()</a></code>. However, in VisiumHD the capture
areas are no longer circles, but rather squares. Thus, we have added a
new argument for all of the aforementioned functions named
<code>shape =</code>. This new argument allows for 3 different
options:</p>
<ul>
<li><p><code>shape = "point"</code>. This would be the default argument
that the functions will take. It will represent the capture areas as
circles.</p></li>
<li><p><code>shape = "tile"</code>. One of the new options, it will
represent the capture areas as squares. Unless specified otherwise in
the argument <code>spot_side</code>, this shape will represent the exact
dimensions of the capture areas in respect to the tissue.</p></li>
<li><p><code>shape = "raster"</code>. The other new plotting option.
Like <code>tile</code>, it will represent the capture areas as squares.
The main peculiarity of this argument is that it is a faster
implementation of <code>tile</code>, which could be useful when plotting
at high resolutions. However, this argument does not accept the HE
image, so one should set <code>image_use = NULL</code> when using this
argument.</p></li>
</ul>
<p>These arguments are compatible with all of the other arguments
already present in the plotting functions, like <code>crop_area</code>
and such. As a side note, even though these options make sense with
VisiumHD data, you can also use them with normal Visium data in order to
represent the capture areas as squares instead of the real circles.</p>
<p>With this information in mind, let’s take a look at the results of
our analysis, starting with Seurat’s clustering:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se.16</span>,</span>
<span>               column_name <span class="op">=</span> <span class="st">"seurat_clusters"</span>, </span>
<span>               override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               shape <span class="op">=</span> <span class="st">"tile"</span>, </span>
<span>               image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>               label_by <span class="op">=</span> <span class="st">"resolution"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<div class="float">
<img src="visiumHD/clusters_16.jpg" alt="1. Seurat clusters identified at 16ums."><div class="figcaption">1. Seurat clusters identified at 16ums.</div>
</div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-labels.html">MapLabels</a></span><span class="op">(</span><span class="va">se.8</span>,</span>
<span>               column_name <span class="op">=</span> <span class="st">"seurat_clusters"</span>, </span>
<span>               override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               shape <span class="op">=</span> <span class="st">"tile"</span>, </span>
<span>               image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>               label_by <span class="op">=</span> <span class="st">"resolution"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<div class="float">
<img src="visiumHD/clusters_8.jpg" alt="2. Seurat clusters identified at 8ums."><div class="figcaption">2. Seurat clusters identified at 8ums.</div>
</div>
<p>Next, let’s take a quick look at the spatial distribution of some of
the NMF factors computed in each resolution:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se.16</span>,</span>
<span>                 features <span class="op">=</span> <span class="st">"NMF_3"</span>, </span>
<span>                 label_by <span class="op">=</span> <span class="st">"resolution"</span>,</span>
<span>                 image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                 override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 shape <span class="op">=</span> <span class="st">"tile"</span>, </span>
<span>                 scale_alpha <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkblue"</span>, <span class="st">"cyan"</span>, <span class="st">"yellow"</span>, <span class="st">"red"</span>, <span class="st">"darkred"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<div class="float">
<img src="visiumHD/NMF_16.jpg" alt="1. NMF factors identified at 16ums."><div class="figcaption">1. NMF factors identified at 16ums.</div>
</div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/visualize-features.html">MapFeatures</a></span><span class="op">(</span><span class="va">se.8</span>,</span>
<span>                 features <span class="op">=</span> <span class="st">"NMF_2"</span>, </span>
<span>                 label_by <span class="op">=</span> <span class="st">"resolution"</span>,</span>
<span>                 image_use <span class="op">=</span> <span class="st">"raw"</span>, </span>
<span>                 override_plot_dims <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 shape <span class="op">=</span> <span class="st">"tile"</span>, </span>
<span>                 scale_alpha <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkblue"</span>, <span class="st">"cyan"</span>, <span class="st">"yellow"</span>, <span class="st">"red"</span>, <span class="st">"darkred"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<div class="float">
<img src="visiumHD/NMF_8.jpg" alt="2. NMF factors identified at 8ums."><div class="figcaption">2. NMF factors identified at 8ums.</div>
</div>
<p><br></p>
<hr>
<details open><summary><strong>Package versions</strong>
</summary><ul>
<li>
<code>semla</code>: 1.1.6</li>
</ul></details><details><summary><strong>Session info</strong>
</summary><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.0 (2024-04-24)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Stockholm</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices datasets  utils     methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] stringr_1.5.1      semla_1.1.6        ggplot2_3.5.1      dplyr_1.1.4       </span></span>
<span><span class="co">## [5] SeuratObject_4.1.4 Seurat_4.3.0.1    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3     rstudioapi_0.16.0      jsonlite_1.8.9        </span></span>
<span><span class="co">##   [4] magrittr_2.0.3         magick_2.8.4           spatstat.utils_3.1-0  </span></span>
<span><span class="co">##   [7] farver_2.1.2           rmarkdown_2.28         fs_1.6.4              </span></span>
<span><span class="co">##  [10] ragg_1.3.3             vctrs_0.6.5            ROCR_1.0-11           </span></span>
<span><span class="co">##  [13] spatstat.explore_3.3-2 forcats_1.0.0          htmltools_0.5.8.1     </span></span>
<span><span class="co">##  [16] sass_0.4.9             sctransform_0.4.1      parallelly_1.38.0     </span></span>
<span><span class="co">##  [19] KernSmooth_2.23-24     bslib_0.8.0            htmlwidgets_1.6.4     </span></span>
<span><span class="co">##  [22] desc_1.4.3             ica_1.0-3              plyr_1.8.9            </span></span>
<span><span class="co">##  [25] plotly_4.10.4          zoo_1.8-12             cachem_1.1.0          </span></span>
<span><span class="co">##  [28] igraph_2.0.3           mime_0.12              lifecycle_1.0.4       </span></span>
<span><span class="co">##  [31] pkgconfig_2.0.3        Matrix_1.7-0           R6_2.5.1              </span></span>
<span><span class="co">##  [34] fastmap_1.2.0          fitdistrplus_1.2-1     future_1.34.0         </span></span>
<span><span class="co">##  [37] shiny_1.9.1            digest_0.6.37          colorspace_2.1-1      </span></span>
<span><span class="co">##  [40] patchwork_1.3.0        tensor_1.5             irlba_2.3.5.1         </span></span>
<span><span class="co">##  [43] textshaping_0.4.0      progressr_0.14.0       fansi_1.0.6           </span></span>
<span><span class="co">##  [46] spatstat.sparse_3.1-0  httr_1.4.7             polyclip_1.10-7       </span></span>
<span><span class="co">##  [49] abind_1.4-8            compiler_4.4.0         bit64_4.0.5           </span></span>
<span><span class="co">##  [52] withr_3.0.1            highr_0.11             MASS_7.3-60.2         </span></span>
<span><span class="co">##  [55] tools_4.4.0            lmtest_0.9-40          httpuv_1.6.15         </span></span>
<span><span class="co">##  [58] future.apply_1.11.2    goftest_1.2-3          glue_1.7.0            </span></span>
<span><span class="co">##  [61] dbscan_1.2-0           nlme_3.1-164           promises_1.3.0        </span></span>
<span><span class="co">##  [64] grid_4.4.0             Rtsne_0.17             cluster_2.1.6         </span></span>
<span><span class="co">##  [67] reshape2_1.4.4         generics_0.1.3         hdf5r_1.3.11          </span></span>
<span><span class="co">##  [70] gtable_0.3.5           spatstat.data_3.1-2    tidyr_1.3.1           </span></span>
<span><span class="co">##  [73] data.table_1.16.0      sp_2.1-4               utf8_1.2.4            </span></span>
<span><span class="co">##  [76] spatstat.geom_3.3-2    RcppAnnoy_0.0.22       ggrepel_0.9.6         </span></span>
<span><span class="co">##  [79] RANN_2.6.2             pillar_1.9.0           spam_2.10-0           </span></span>
<span><span class="co">##  [82] later_1.3.2            splines_4.4.0          lattice_0.22-6        </span></span>
<span><span class="co">##  [85] bit_4.0.5              renv_1.0.2             survival_3.6-4        </span></span>
<span><span class="co">##  [88] deldir_2.0-4           tidyselect_1.2.1       miniUI_0.1.1.1        </span></span>
<span><span class="co">##  [91] pbapply_1.7-2          knitr_1.48             gridExtra_2.3         </span></span>
<span><span class="co">##  [94] scattermore_1.2        xfun_0.47              matrixStats_1.4.1     </span></span>
<span><span class="co">##  [97] stringi_1.8.4          lazyeval_0.2.2         yaml_2.3.10           </span></span>
<span><span class="co">## [100] evaluate_0.24.0        codetools_0.2-20       tibble_3.2.1          </span></span>
<span><span class="co">## [103] BiocManager_1.30.25    cli_3.6.3              uwot_0.2.2            </span></span>
<span><span class="co">## [106] arrow_17.0.0.1         xtable_1.8-4           reticulate_1.39.0     </span></span>
<span><span class="co">## [109] systemfonts_1.1.0      munsell_0.5.1          jquerylib_0.1.4       </span></span>
<span><span class="co">## [112] Rcpp_1.0.13            globals_0.16.3         spatstat.random_3.3-1 </span></span>
<span><span class="co">## [115] zeallot_0.1.0          png_0.1-8              spatstat.univar_3.0-1 </span></span>
<span><span class="co">## [118] parallel_4.4.0         assertthat_0.2.1       pkgdown_2.1.0         </span></span>
<span><span class="co">## [121] dotCall64_1.1-1        listenv_0.9.1          viridisLite_0.4.2     </span></span>
<span><span class="co">## [124] scales_1.3.0           ggridges_0.5.6         leiden_0.4.3.1        </span></span>
<span><span class="co">## [127] purrr_1.0.2            rlang_1.1.4            cowplot_1.1.3         </span></span>
<span><span class="co">## [130] shinyjs_2.1.0</span></span></code></pre>
</details><p><br></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://github.com/jemorlanes" class="external-link">Javier Escudero Morlanes</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
