<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Mixed cell type prediction with NNLS (experimental) — RunMixedNNLS • semla</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Mixed cell type prediction with NNLS (experimental) — RunMixedNNLS"><meta name="description" content="This function can be used to obtain proportion estimates for a mixture of cell types
and unknown factors. The input spatial expression matrix is first deconvolved using
Non-negative Matrix Factorization (NMF) with the rank set to number of cell types + k
additional factors. The estimated factor expression profiles are are then compared with
the cell type expression profiles by computing pairwise correlation scores.
Factor expression profiles with the lowest correlation scores are kept and combined
with the cell type expression profiles to predict proportions directly from the spatial
data using Non-negative Least Squares (NNLS)."><meta property="og:description" content="This function can be used to obtain proportion estimates for a mixture of cell types
and unknown factors. The input spatial expression matrix is first deconvolved using
Non-negative Matrix Factorization (NMF) with the rank set to number of cell types + k
additional factors. The estimated factor expression profiles are are then compared with
the cell type expression profiles by computing pairwise correlation scores.
Factor expression profiles with the lowest correlation scores are kept and combined
with the cell type expression profiles to predict proportions directly from the spatial
data using Non-negative Least Squares (NNLS)."><meta property="og:image" content="https://ludvigla.github.io/semla/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.1.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/semla.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><h6 class="dropdown-header" data-toc-skip>Intro</h6></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Get started</a></li>
    <li><a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6></li>
    <li><a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a></li>
    <li><a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a></li>
    <li><a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a></li>
    <li><a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Spatial methods</h6></li>
    <li><a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a></li>
    <li><a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a></li>
    <li><a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Analysis</h6></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a></li>
    <li><a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a></li>
    <li><a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Image processing</h6></li>
    <li><a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a></li>
    <li><a class="dropdown-item" href="../articles/mask_images.html">Image masking</a></li>
    <li><a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/create_object.html">Create a Staffli object</a></li>
    <li><a class="dropdown-item" href="../articles/compare_cell_type_mapping_NNLS.html">Cell type mapping comparison</a></li>
    <li><a class="dropdown-item" href="../articles/cell_type_mapping_benchmark.html">Cell type mapping benchmark</a></li>
    <li><a class="dropdown-item" href="../articles/slide-seq.html">Slide-Seq data</a></li>
    <li><a class="dropdown-item" href="../articles/IF_data.html">IF data</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub"><span class="fa fa-github"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Mixed cell type prediction with NNLS (experimental)</h1>

      <div class="d-none name"><code>celltype-prediction-mixed.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function can be used to obtain proportion estimates for a mixture of cell types
and unknown factors. The input spatial expression matrix is first deconvolved using
Non-negative Matrix Factorization (NMF) with the rank set to number of cell types + k
additional factors. The estimated factor expression profiles are are then compared with
the cell type expression profiles by computing pairwise correlation scores.
Factor expression profiles with the lowest correlation scores are kept and combined
with the cell type expression profiles to predict proportions directly from the spatial
data using Non-negative Least Squares (NNLS).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RunMixedNNLS</span><span class="op">(</span><span class="va">object</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Default S3 method</span></span>
<span><span class="fu">RunMixedNNLS</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">singlecell_matrix</span>,</span>
<span>  <span class="va">groups</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  nCells_per_group <span class="op">=</span> <span class="fl">50L</span>,</span>
<span>  minCells_per_celltype <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  min_prop <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  L1 <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1337L</span>,</span>
<span>  return_expression_profiles <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S3 method for class 'Seurat'</span></span>
<span><span class="fu">RunMixedNNLS</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">singlecell_object</span>,</span>
<span>  groups <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  singlecell_assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  spatial_assay <span class="op">=</span> <span class="st">"Spatial"</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"data"</span>,</span>
<span>  min_prop <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  nCells_per_group <span class="op">=</span> <span class="fl">50L</span>,</span>
<span>  minCells_per_celltype <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"celltypepropsmixed"</span>,</span>
<span>  dimred_name <span class="op">=</span> <span class="st">"nnlsmixed"</span>,</span>
<span>  return_as_dimred <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  L1 <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1337L</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>A matrix-like object with 10x Visium data</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Arguments passed to other methods</p></dd>


<dt id="arg-singlecell-matrix">singlecell_matrix<a class="anchor" aria-label="anchor" href="#arg-singlecell-matrix"></a></dt>
<dd><p>A matrix-like object with scRNA-seq data</p></dd>


<dt id="arg-groups">groups<a class="anchor" aria-label="anchor" href="#arg-groups"></a></dt>
<dd><p>A character vector of length <code>ncol(singlecell_matrix)</code>
with cell type labels. For <code>Seurat</code> objects, one can omit groups in which case
the current identity is used instead. Alternatively, a string can be provided
specifying a column to use from the <code>singlecell_object</code> meta.data slot.</p></dd>


<dt id="arg-k">k<a class="anchor" aria-label="anchor" href="#arg-k"></a></dt>
<dd><p>An integer specifying the number of factors to compute</p></dd>


<dt id="arg-ncells-per-group">nCells_per_group<a class="anchor" aria-label="anchor" href="#arg-ncells-per-group"></a></dt>
<dd><p>Each cell type is down sampled to this number before
computing cell type expression profiles. This is mainly used to speed up
computation. Note that if a cell type has fewer than <code>nCells_per_group</code>
cells, the cells will be sampled with replacement to obtain <code>nCells_per_group</code>
cells.</p></dd>


<dt id="arg-mincells-per-celltype">minCells_per_celltype<a class="anchor" aria-label="anchor" href="#arg-mincells-per-celltype"></a></dt>
<dd><p>Minimum number of cells allowed per cell type</p></dd>


<dt id="arg-min-prop">min_prop<a class="anchor" aria-label="anchor" href="#arg-min-prop"></a></dt>
<dd><p>Minimum proportion allowed. Any proportion values lower than
this threshold will be removed and the remaining proportions will be rescaled.</p></dd>


<dt id="arg-l-">L1<a class="anchor" aria-label="anchor" href="#arg-l-"></a></dt>
<dd><p>lasso penalty</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>An integer to set seed for reproducibility</p></dd>


<dt id="arg-return-expression-profiles">return_expression_profiles<a class="anchor" aria-label="anchor" href="#arg-return-expression-profiles"></a></dt>
<dd><p>Logical specifying if the expression profile matrix
should also be returned</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Print messages</p></dd>


<dt id="arg-singlecell-object">singlecell_object<a class="anchor" aria-label="anchor" href="#arg-singlecell-object"></a></dt>
<dd><p>A <code>Seurat</code> object with single-cell gene expression data</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>Features to use for computation. If no features are
provided, the intersection of variable features between sc_object and st_object
will be used. Note that the intersect can be quite small, in which case you can
increase the number of variable features by rerunning <code>FindVariableFeatures</code>
and increase the number with the <code>nfeatures</code> argument.</p></dd>


<dt id="arg-singlecell-assay">singlecell_assay<a class="anchor" aria-label="anchor" href="#arg-singlecell-assay"></a></dt>
<dd><p>Assay in single cell data object to use for deconvolution</p></dd>


<dt id="arg-spatial-assay">spatial_assay<a class="anchor" aria-label="anchor" href="#arg-spatial-assay"></a></dt>
<dd><p>Assay in Visium data object to use for deconvolution</p></dd>


<dt id="arg-slot">slot<a class="anchor" aria-label="anchor" href="#arg-slot"></a></dt>
<dd><p>Name of slot in <code>singlecell_assay</code> and <code>spatial_assay</code> to use for
deconvolution</p></dd>


<dt id="arg-assay-name">assay_name<a class="anchor" aria-label="anchor" href="#arg-assay-name"></a></dt>
<dd><p>Sets the name of the returned <code>Assay</code> object. Only active
if return.as.dimred = FALSE.</p></dd>


<dt id="arg-dimred-name">dimred_name<a class="anchor" aria-label="anchor" href="#arg-dimred-name"></a></dt>
<dd><p>Sets the name of  the returned <code>DimReduc</code> object. Only
active if return.as.dimred = TRUE.</p></dd>


<dt id="arg-return-as-dimred">return_as_dimred<a class="anchor" aria-label="anchor" href="#arg-return-as-dimred"></a></dt>
<dd><p>By default, the results are returned as an <code>Assay</code>
named "celltypeprops", where each feature is named after the cell type IDs.
Alternatively, you can return the results as a <code>DimReduc</code> object. This will
also provide the cell gene loadings for each cell type. However, the proportions
will instead be named "factor_1", "factor_2", ...</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An object with mixed cell type and factor proportion estimates</p>
    </div>
    <div class="section level2">
    <h2 id="default-method">default method<a class="anchor" aria-label="anchor" href="#default-method"></a></h2>


<p>Input <code>object</code> is a matrix-like object with 10x Visium data. The function returns
a matrix with estimated proportions of dimensions (nCellTypes + k) x nSpots,
where nCellTypes is the number of cell types, k is the number of factors and nSpots is the number of spots.
If <code>return_expression_profiles=TRUE</code>, the returned object will be a list with estimated
proportions <code>prop</code> and the cell type/factor expression profile matrix <code>W</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="seurat-method">Seurat method<a class="anchor" aria-label="anchor" href="#seurat-method"></a></h2>


<p>Input <code>object</code> is a <code>Seurat</code> object with 10x Visium data. The function returns
the <code>Seurat</code> object with either a new <code>Assay</code> or <code>DimReduc</code> object containing
estimated proportions for cell types and additional factors.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

