<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function serves as a wrapper for LoadAndMergeMatrices and
LoadSpatialCoordinates to load spaceranger output files and create
a Seurat object. The spatial information, i.e. images and spot coordinates, are
stored inside the tools slot of the Seurat object in Staffli object."><title>Read spaceranger output files — ReadVisiumData • semla</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Read spaceranger output files — ReadVisiumData"><meta property="og:description" content="This function serves as a wrapper for LoadAndMergeMatrices and
LoadSpatialCoordinates to load spaceranger output files and create
a Seurat object. The spatial information, i.e. images and spot coordinates, are
stored inside the tools slot of the Seurat object in Staffli object."><meta property="og:image" content="https://ludvigla.github.io/semla/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">semla</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/semla.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-tutorials">Tutorials</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-tutorials">
    <h6 class="dropdown-header" data-toc-skip>Intro</h6>
    <a class="dropdown-item" href="../articles/getting_started.html">Get started</a>
    <a class="dropdown-item" href="../articles/subset_and_merge.html">Subset/merge</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Interactive</h6>
    <a class="dropdown-item" href="../articles/feature_viewer.html">Feature viewer</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial visualization</h6>
    <a class="dropdown-item" href="../articles/numeric_features.html">Map numeric features</a>
    <a class="dropdown-item" href="../articles/categorical_features.html">Map categorical features</a>
    <a class="dropdown-item" href="../articles/advanced_visualization.html">Advanced visualization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Spatial methods</h6>
    <a class="dropdown-item" href="../articles/radial_distances.html">Radial distances</a>
    <a class="dropdown-item" href="../articles/region_neighbors.html">Region neighbors</a>
    <a class="dropdown-item" href="../articles/spatial_statistics_for_labelled_spots.html">Statistics for labelled spots</a>
    <a class="dropdown-item" href="../articles/spatial_autocorrelation.html">Spatial autocorrelation</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Analysis</h6>
    <a class="dropdown-item" href="../articles/cell_type_mapping_with_NNLS.html">Cell type mapping</a>
    <a class="dropdown-item" href="../articles/digital_unrolling.html">Digital unrolling</a>
    <a class="dropdown-item" href="../articles/NNMF.html">Non-negative Matrix Factorization</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Image processing</h6>
    <a class="dropdown-item" href="../articles/images_and_coordinates.html">Images and spot coordinates</a>
    <a class="dropdown-item" href="../articles/mask_images.html">Image masking</a>
    <a class="dropdown-item" href="../articles/image_alignment.html">Image alignment</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/ludvigla/semla" aria-label="GitHub">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Read spaceranger output files</h1>
      
      <div class="d-none name"><code>ReadVisiumData.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function serves as a wrapper for <code><a href="LoadAndMergeMatrices.html">LoadAndMergeMatrices</a></code> and
<code><a href="LoadSpatialCoordinates.html">LoadSpatialCoordinates</a></code> to load spaceranger output files and create
a Seurat object. The spatial information, i.e. images and spot coordinates, are
stored inside the tools slot of the <code>Seurat</code> object in <code>Staffli</code> object.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ReadVisiumData</span><span class="op">(</span></span>
<span>  <span class="va">infoTable</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"Spatial"</span>,</span>
<span>  remove_spots_outside_HE <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  remove_spots_outside_tissue <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>infoTable</dt>
<dd><p>A <code>data.frame</code> or <code>tbl</code> with paths to spaceranger output files</p></dd>


<dt>assay</dt>
<dd><p>Assay name (default = "Spatial")</p></dd>


<dt>remove_spots_outside_HE</dt>
<dd><p>Should spots outside the H&amp;E be removed? This option
can be useful for CytAssist data when the H&amp;E image only cover a smaller part of the
entire tissue section.</p></dd>


<dt>remove_spots_outside_tissue</dt>
<dd><p>Should spots outside the tissue be removed?</p></dd>


<dt>verbose</dt>
<dd><p>Print messages</p></dd>


<dt>...</dt>
<dd><p>Parameters passed to <code><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A <code><a href="https://satijalab.org/seurat/reference/Seurat-package.html" class="external-link">Seurat</a></code> object with additional spatial information stored in
a <code>Staffli</code> object</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>ReadVisiumData</code> takes a <code>data.frame</code> like table as input that should hold
certain spaceranger output file paths. The table should consist of four columns:
"samples", "imgs", "spotfiles" and "json".</p>
<ul><li><p>"samples" : file paths to expression matrices, e.g. <code>filtered_bc_matrix.h5</code></p></li>
<li><p>"imgs" : file paths to images, e.g. <code>tissue_hires_image.jpg</code></p></li>
<li><p>"spotfiles" : file paths to spot coordinate CSV files <code>tissue_positions_list.csv</code></p></li>
<li><p>"json" : file paths to scale factor JSON files, e.g. <code>scalefactors_json.json</code>. 
   It is also possible to provide custom scale factors (more info below).</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="custom-scale-factors">Custom scale factors<a class="anchor" aria-label="anchor" href="#custom-scale-factors"></a></h2>
    

<p>You can provide an additional column named 'scalefactor' with custom scale factor values.
The values should be between 0 to 1, where a 1 would correspond to the original H&amp;E image
used as input for spaceranger. For instance, if your image was scaled from a height of 30,000
pixels to a height of 3,000 pixels (with the same aspect ratio), the scalefactor would be 0.1.</p>
    </div>
    <div class="section level2">
    <h2 id="load-data-outside-tissue-section">Load data outside tissue section<a class="anchor" aria-label="anchor" href="#load-data-outside-tissue-section"></a></h2>
    

<p>Sometimes it can be useful to load data for all spots in a 10x Visium dataset, if you
need to explore transcripts captured outside of the tissue. In this case, you can
provide paths to the <code>raw_feature_bc_matrix.h5</code> files in the spaceranger output folders
and set <code>remove_spots_outside_tissue = FALSE</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="filter-data">Filter data<a class="anchor" aria-label="anchor" href="#filter-data"></a></h2>
    

<p>If you want to filter out spots and features, you can pass the <code>min.cells</code> and
<code>min.features</code> parameters (see <code><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></code> for more details);
however, it is recommended to use the <code><a href="subset.html">SubsetSTData</a></code> function for filtering
after the object has been created.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other pre-process: 
<code><a href="LoadAndMergeMatrices.html">LoadAndMergeMatrices</a>()</code>,
<code><a href="LoadSpatialCoordinates.html">LoadSpatialCoordinates</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Assemble spaceranger output files</span></span></span>
<span class="r-in"><span><span class="va">samples</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span>paths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          <span class="st">"/*/filtered_feature_bc_matrix.h5"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">imgs</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span>paths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          <span class="st">"/*/spatial/tissue_hires_image.jpg"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">spotfiles</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span>paths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          <span class="st">"/*/spatial/tissue_positions_list.csv"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">json</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span>paths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"semla"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                          <span class="st">"/*/spatial/scalefactors_json.json"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create a tibble/data.frame with file paths</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">infoTable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="va">samples</span>, <span class="va">imgs</span>, <span class="va">spotfiles</span>, <span class="va">json</span>, sample_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mousebrain"</span>, <span class="st">"mousecolon"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create Seurat object</span></span></span>
<span class="r-in"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">ReadVisiumData</span><span class="op">(</span>infoTable <span class="op">=</span> <span class="va">infoTable</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ── <span style="font-weight: bold;">Reading 10x Visium data</span> ──</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Loading matrices:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> →   Finished loading expression matrix 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> →   Finished loading expression matrix 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> <span style="color: #FF5555;">There are only 188 gene shared across all matrices:</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> → <span style="color: #FF5555;">  Are you sure that the matrices share the same gene IDs?</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> → <span style="color: #FF5555;">  Are the datasets from the same species?</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Merging matrices:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> There are <span style="color: #5555FF;">188</span> features and <span style="color: #FF55FF;">5164</span> spots in the merged matrix.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Loading coordinates:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> →   Finished loading coordinates for sample 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> →   Finished loading coordinates for sample 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Collected coordinates for <span style="color: #FF55FF;">5164</span> spots.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> ── Creating `Seurat` object </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Expression matrices and coordinates are compatible</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Created `Seurat` object</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Created `Staffli` object</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Returning a `Seurat` object with <span style="color: #5555FF;">188</span> features and <span style="color: #FF55FF;">5164</span> spots</span>
<span class="r-in"><span><span class="va">se</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> An object of class Seurat </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 188 features across 5164 samples within 1 assay </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Active assay: Spatial (188 features, 0 variable features)</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by <a href="https://github.com/ludvigla" class="external-link">Ludvig Larsson</a>, <a href="https://github.com/lfranzen" class="external-link">Lovisa Franzen</a>, <a href="https://www.spatialresearch.org/" class="external-link"><img src="reference/figures/sr-logo.png" height="36" alt="spatialresearch"></a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

